// tdesign-vue-next/lib/radio/radio-group-props.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

var props = {
  allowUncheck: Boolean,
  direction: {
    type: String,
    "default": "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  disabled: {
    type: Boolean,
    "default": void 0
  },
  name: {
    type: String,
    "default": ""
  },
  options: {
    type: Array
  },
  readonly: {
    type: Boolean,
    "default": void 0
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  theme: {
    type: String,
    "default": "radio",
    validator: function validator(val) {
      if (!val) return true;
      return ["radio", "button"].includes(val);
    }
  },
  value: {
    type: [String, Number, Boolean],
    "default": void 0
  },
  modelValue: {
    type: [String, Number, Boolean],
    "default": void 0
  },
  defaultValue: {
    type: [String, Number, Boolean]
  },
  variant: {
    type: String,
    "default": "outline",
    validator: function validator(val) {
      if (!val) return true;
      return ["outline", "primary-filled", "default-filled"].includes(val);
    }
  },
  onChange: Function
};

export { props as default };
