// tdesign-vue-next/lib/date-picker/DateRangePicker.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { defineComponent, computed, ref, watch, createVNode } from 'vue';
import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import dayjs from 'dayjs';
import { isArray, isFunction } from 'lodash-es';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-651bdc55.js';
import { u as usePrefixClass } from '../_chunks/dep-59f77e78.js';
import { u as useDisabled } from '../_chunks/dep-ed0885e7.js';
import '../_chunks/dep-204ae706.js';
import { u as useReadonly } from '../_chunks/dep-0d255418.js';
import '@babel/runtime/helpers/defineProperty';
import dateRangePickerProps from './date-range-picker-props.js';
import { RangeInputPopup } from '../range-input/index.js';
import TRangePanel from './components/panel/RangePanel.js';
import { useRange } from './hooks/useRange.js';
import { g as getDefaultFormat, f as formatDate, a as formatTime, i as initYearMonthTime, p as parseToDayjs, b as isValidDate } from '../_chunks/dep-f54e73f7.js';
import { g as getRangeBounds, s as subtractMonth, a as addMonth, e as extractTimeObj } from '../_chunks/dep-a7cf6cc7.js';
import { dateCorrection, triggerMap } from './utils/index.js';
import '../_chunks/dep-76473034.js';
import '../config-provider/hooks/useConfig.js';
import '../config-provider/utils/context.js';
import '../_chunks/dep-5c1cdd0b.js';
import '@babel/runtime/helpers/createClass';
import '@babel/runtime/helpers/classCallCheck';
import '../range-input/range-input.js';
import 'tdesign-icons-vue-next';
import '../input/index.js';
import '../input/input.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../input/props.js';
import '../_chunks/dep-eff3e90f.js';
import '../_chunks/dep-a7922f61.js';
import '../_chunks/dep-69502ff8.js';
import '../_chunks/dep-7daa5fe4.js';
import '../_chunks/dep-3d495891.js';
import '../input/hooks/useInput.js';
import '../form/constants/index.js';
import '../_chunks/dep-f32cc39c.js';
import '../input/hooks/useLengthLimit.js';
import '../_chunks/dep-f2301d1c.js';
import '../_chunks/dep-a9edb4d8.js';
import '../input/hooks/useInputEventHandler.js';
import '../input/hooks/useInputWidth.js';
import '../_chunks/dep-f5c5752f.js';
import '../input/input-group.js';
import '../input/input-group-props.js';
import '../_chunks/dep-5dfdb698.js';
import '../range-input/props.js';
import '../_chunks/dep-ff412569.js';
import '../range-input/range-input-popup.js';
import '../popup/index.js';
import '../popup/popup.js';
import '@popperjs/core';
import '../_chunks/dep-d15b7e74.js';
import '../popup/container.js';
import '../popup/props.js';
import '../range-input/range-input-popup-props.js';
import '../select-input/hooks/useOverlayInnerStyle.js';
import './components/panel/PanelContent.js';
import './components/base/Header.js';
import '../pagination/index.js';
import '../pagination/pagination.js';
import '../_chunks/dep-c6879e67.js';
import '../input-number/index.js';
import '../input-number/input-number.js';
import '../button/index.js';
import '../button/button.js';
import '../loading/index.js';
import '../loading/directive.js';
import '../loading/plugin.js';
import '../_chunks/dep-d32d8561.js';
import '../loading/icon/gradient.js';
import '../loading/props.js';
import '../button/props.js';
import '../_chunks/dep-16e13fa9.js';
import '../input-number/props.js';
import '../input-number/hooks/useInputNumber.js';
import '../_chunks/dep-b21bd6d2.js';
import '../_chunks/dep-75be08e7.js';
import '../select/index.js';
import '../select/select.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '../common-components/fake-arrow.js';
import '../select-input/index.js';
import '../select-input/select-input.js';
import '../select-input/props.js';
import '../select-input/hooks/useMultiple.js';
import '../tag-input/index.js';
import '../tag-input/tag-input.js';
import '../tag-input/props.js';
import '../tag-input/hooks/useDragSorter.js';
import '../tag-input/hooks/useHover.js';
import '../tag-input/hooks/useTagScroll.js';
import '../tag-input/hooks/useTagList.js';
import '../tag/index.js';
import '../tag/tag.js';
import 'tinycolor2';
import '../tag/props.js';
import '../tag/check-tag.js';
import '../tag/check-tag-props.js';
import '../_chunks/dep-8d513880.js';
import '../tag/check-tag-group.js';
import '../tag/check-tag-group-props.js';
import '../select-input/hooks/useSingle.js';
import '../select/components/select-panel.js';
import '../select/option.js';
import '../select/option-props.js';
import '../checkbox/index.js';
import '../checkbox/checkbox.js';
import '../checkbox/props.js';
import '../checkbox/constants/index.js';
import '../checkbox/hooks/useCheckboxLazyLoad.js';
import '../_chunks/dep-2cabe3fe.js';
import '../checkbox/hooks/useKeyboardEvent.js';
import '../checkbox/group.js';
import '../checkbox/checkbox-group-props.js';
import '../_chunks/dep-d4579130.js';
import '../select/utils/index.js';
import '../select/constants/index.js';
import '../select/option-group.js';
import '../select/option-group-props.js';
import '../select/props.js';
import '../select/hooks/usePanelVirtualScroll.js';
import '../select/hooks/useKeyboardControl.js';
import '../select/hooks/useSelectOptions.js';
import '../input-adornment/index.js';
import '../input-adornment/input-adornment.js';
import '../input-adornment/props.js';
import '../pagination/props.js';
import '../pagination/hooks/useMoreAction.js';
import '../pagination/hooks/usePaginationClasses.js';
import '../pagination/pagination-mini.js';
import '../pagination/pagination-mini-props.js';
import '../tooltip/index.js';
import '../tooltip/tooltip.js';
import '../tooltip/props.js';
import '../tooltip/utils/index.js';
import './hooks/useSelectRange.js';
import './components/base/Table.js';
import './components/base/Cell.js';
import '../time-picker/index.js';
import '../time-picker/time-picker.js';
import '../_chunks/dep-7a2c5955.js';
import '../time-picker/panel/time-picker-panel.js';
import '../_chunks/dep-ea86b40b.js';
import '../time-picker/props.js';
import '../_chunks/dep-c817a169.js';
import '../_chunks/dep-3331fba2.js';
import '../time-picker/time-range-picker.js';
import '../time-picker/time-range-picker-props.js';
import './components/panel/ExtraContent.js';
import './components/base/Footer.js';
import './hooks/useDisableDate.js';
import './hooks/useTableData.js';
import './hooks/useRangeValue.js';
import '@babel/runtime/helpers/objectDestructuringEmpty';
import '@babel/runtime/helpers/extends';

var _DateRangePicker = defineComponent({
  name: "TDateRangePicker",
  props: dateRangePickerProps,
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var COMPONENT_NAME = usePrefixClass("date-range-picker");
    var _useRange = useRange(props2),
      inputValue = _useRange.inputValue,
      popupVisible = _useRange.popupVisible,
      rangeInputProps = _useRange.rangeInputProps,
      popupProps = _useRange.popupProps,
      value = _useRange.value,
      year = _useRange.year,
      month = _useRange.month,
      time = _useRange.time,
      cacheValue = _useRange.cacheValue,
      activeIndex = _useRange.activeIndex,
      isHoverCell = _useRange.isHoverCell,
      isFirstValueSelected = _useRange.isFirstValueSelected,
      onChange = _useRange.onChange;
    var isDisabled = useDisabled();
    var isReadOnly = useReadonly();
    var formatRef = computed(function () {
      return getDefaultFormat({
        mode: props2.mode,
        enableTimePicker: props2.enableTimePicker,
        format: props2.format,
        valueType: props2.valueType
      });
    });
    var isSelected = ref(false);
    watch(popupVisible, function (visible) {
      if (visible) {
        if (isArray(props2.disabled)) activeIndex.value = props2.disabled[0] ? 1 : 0;
        isSelected.value = false;
        cacheValue.value = formatDate(value.value || [], {
          format: formatRef.value.valueType,
          targetFormat: formatRef.value.format
        });
        time.value = formatTime(value.value || [dayjs().format(formatRef.value.timeFormat), dayjs().format(formatRef.value.timeFormat)], formatRef.value.format, formatRef.value.timeFormat, props2.defaultTime);
        if (!value.value.length) {
          if (props2.range && isArray(props2.range) || props2.panelActiveDate) {
            var _startRangeBounds$min, _startRangeBounds$min2, _startRangeBounds$max, _startRangeBounds$min3, _startRangeBounds$min4, _startRangeBounds$max2, _endRangeBounds$min$g, _endRangeBounds$min, _endRangeBounds$max, _endRangeBounds$min$g2, _endRangeBounds$min2, _endRangeBounds$max2, _startPanelActiveDate, _startPanelActiveDate2, _startPanelActiveDate3, _startPanelActiveDate4, _endPanelActiveDate$y, _endPanelActiveDate, _endPanelActiveDate2, _endPanelActiveDate3;
            var startRange = props2.range;
            var endRange = props2.range;
            if (isArray(props2.range)) {
              var _props2$range = _slicedToArray(props2.range, 2),
                first = _props2$range[0],
                second = _props2$range[1];
              if (isArray(first) || isFunction(first) || isArray(second) || isFunction(second)) {
                startRange = first;
                endRange = second;
              }
            }
            var startRangeBounds = getRangeBounds(startRange);
            var endRangeBounds = getRangeBounds(endRange);
            var startYearFromRange = (_startRangeBounds$min = (_startRangeBounds$min2 = startRangeBounds.min) === null || _startRangeBounds$min2 === void 0 ? void 0 : _startRangeBounds$min2.getFullYear()) !== null && _startRangeBounds$min !== void 0 ? _startRangeBounds$min : (_startRangeBounds$max = startRangeBounds.max) === null || _startRangeBounds$max === void 0 ? void 0 : _startRangeBounds$max.getFullYear();
            var startMonthFromRange = (_startRangeBounds$min3 = (_startRangeBounds$min4 = startRangeBounds.min) === null || _startRangeBounds$min4 === void 0 ? void 0 : _startRangeBounds$min4.getMonth()) !== null && _startRangeBounds$min3 !== void 0 ? _startRangeBounds$min3 : (_startRangeBounds$max2 = startRangeBounds.max) === null || _startRangeBounds$max2 === void 0 ? void 0 : _startRangeBounds$max2.getMonth();
            var endYearFromRange = (_endRangeBounds$min$g = (_endRangeBounds$min = endRangeBounds.min) === null || _endRangeBounds$min === void 0 ? void 0 : _endRangeBounds$min.getFullYear()) !== null && _endRangeBounds$min$g !== void 0 ? _endRangeBounds$min$g : (_endRangeBounds$max = endRangeBounds.max) === null || _endRangeBounds$max === void 0 ? void 0 : _endRangeBounds$max.getFullYear();
            var endMonthFromRange = (_endRangeBounds$min$g2 = (_endRangeBounds$min2 = endRangeBounds.min) === null || _endRangeBounds$min2 === void 0 ? void 0 : _endRangeBounds$min2.getMonth()) !== null && _endRangeBounds$min$g2 !== void 0 ? _endRangeBounds$min$g2 : (_endRangeBounds$max2 = endRangeBounds.max) === null || _endRangeBounds$max2 === void 0 ? void 0 : _endRangeBounds$max2.getMonth();
            var startPanelActiveDate = props2.panelActiveDate;
            var endPanelActiveDate = props2.panelActiveDate;
            if (isArray(props2.panelActiveDate)) {
              var _props2$panelActiveDa = _slicedToArray(props2.panelActiveDate, 2);
              startPanelActiveDate = _props2$panelActiveDa[0];
              endPanelActiveDate = _props2$panelActiveDa[1];
            }
            var leftYear = (_startPanelActiveDate = (_startPanelActiveDate2 = startPanelActiveDate) === null || _startPanelActiveDate2 === void 0 ? void 0 : _startPanelActiveDate2.year) !== null && _startPanelActiveDate !== void 0 ? _startPanelActiveDate : startYearFromRange;
            var leftMonth = (_startPanelActiveDate3 = startPanelActiveDate) !== null && _startPanelActiveDate3 !== void 0 && _startPanelActiveDate3.month ? Number((_startPanelActiveDate4 = startPanelActiveDate) === null || _startPanelActiveDate4 === void 0 ? void 0 : _startPanelActiveDate4.month) - 1 : startMonthFromRange;
            var rightYear = (_endPanelActiveDate$y = (_endPanelActiveDate = endPanelActiveDate) === null || _endPanelActiveDate === void 0 ? void 0 : _endPanelActiveDate.year) !== null && _endPanelActiveDate$y !== void 0 ? _endPanelActiveDate$y : endYearFromRange;
            var rightMonth = (_endPanelActiveDate2 = endPanelActiveDate) !== null && _endPanelActiveDate2 !== void 0 && _endPanelActiveDate2.month ? Number((_endPanelActiveDate3 = endPanelActiveDate) === null || _endPanelActiveDate3 === void 0 ? void 0 : _endPanelActiveDate3.month) - 1 : endMonthFromRange;
            var _initYearMonthTime = initYearMonthTime({
                value: value.value,
                mode: props2.mode,
                format: formatRef.value.format,
                enableTimePicker: props2.enableTimePicker
              }),
              defaultYear = _initYearMonthTime.year,
              defaultMonth = _initYearMonthTime.month;
            year.value = [leftYear !== null && leftYear !== void 0 ? leftYear : defaultYear[0], rightYear !== null && rightYear !== void 0 ? rightYear : defaultYear[1]];
            month.value = [leftMonth !== null && leftMonth !== void 0 ? leftMonth : defaultMonth[0], rightMonth !== null && rightMonth !== void 0 ? rightMonth : defaultMonth[1]];
            if (year.value[0] === year.value[1] && month.value[0] === month.value[1] && !props2.enableTimePicker) {
              if (startRange === endRange) {
                if (props2.mode === "year") year.value[1] += 10;else if (props2.mode === "month" || props2.mode === "quarter") year.value[1] += 1;else month.value[1] += 1;
              }
            }
          } else {
            var _initYearMonthTime2 = initYearMonthTime({
                value: value.value,
                mode: props2.mode,
                format: formatRef.value.format,
                enableTimePicker: props2.enableTimePicker
              }),
              _defaultYear = _initYearMonthTime2.year,
              _defaultMonth = _initYearMonthTime2.month;
            year.value = _defaultYear;
            month.value = _defaultMonth;
          }
        } else if (value.value.length === 2 && !props2.enableTimePicker) {
          var nextMonth = value.value.map(function (v) {
            return parseToDayjs(v, formatRef.value.format).month();
          });
          year.value = value.value.map(function (v) {
            return parseToDayjs(v, formatRef.value.valueType).year();
          });
          if (year.value[0] === year.value[1] && nextMonth[0] === nextMonth[1]) {
            nextMonth[0] === 11 ? nextMonth[0] -= 1 : nextMonth[1] += 1;
          }
          month.value = nextMonth;
          if ((props2.mode === "month" || props2.mode === "quarter") && year.value[0] === year.value[1]) {
            year.value = [year.value[0], year.value[0] + 1];
          }
        } else {
          year.value = value.value.map(function (v) {
            return parseToDayjs(v, formatRef.value.format).year();
          });
          if (year.value.length === 1) year.value = [year.value[0], year.value[0]];
          month.value = value.value.map(function (v) {
            return parseToDayjs(v, formatRef.value.format).month();
          });
          if (month.value.length === 1) month.value = [month.value[0], Math.min(month.value[0] + 1, 11)];
        }
      } else {
        activeIndex.value = 0;
        isHoverCell.value = false;
        isFirstValueSelected.value = false;
        if (props2.needConfirm) {
          inputValue.value = formatDate(value.value, {
            format: formatRef.value.valueType,
            targetFormat: formatRef.value.format
          });
        } else {
          confirmValueChange();
        }
      }
    });
    function onCellMouseEnter(date) {
      isHoverCell.value = true;
      var nextValue = _toConsumableArray(inputValue.value);
      nextValue[activeIndex.value] = formatDate(date, {
        format: formatRef.value.format
      });
      inputValue.value = nextValue;
    }
    function onCellMouseLeave() {
      isHoverCell.value = false;
      inputValue.value = cacheValue.value;
    }
    function onCellClick(date, _ref2) {
      var _props2$onPick;
      var e = _ref2.e;
      (_props2$onPick = props2.onPick) === null || _props2$onPick === void 0 || _props2$onPick.call(props2, date, {
        e: e,
        partial: activeIndex.value ? "end" : "start"
      });
      isHoverCell.value = false;
      isSelected.value = true;
      var nextValue = _toConsumableArray(inputValue.value);
      nextValue[activeIndex.value] = formatDate(date, {
        format: formatRef.value.format
      });
      cacheValue.value = nextValue;
      inputValue.value = nextValue;
      if (props2.enableTimePicker) return;
      var notValidIndex = nextValue.findIndex(function (v) {
        return !v || !isValidDate(v, formatRef.value.format);
      });
      if (notValidIndex === -1 && nextValue.length === 2) {
        if (!isFirstValueSelected.value && parseToDayjs(nextValue[0], formatRef.value.format).isAfter(parseToDayjs(nextValue[1], formatRef.value.format))) {
          nextValue[activeIndex.value ? 0 : 1] = "";
          cacheValue.value = nextValue;
          inputValue.value = nextValue;
        } else {
          onChange === null || onChange === void 0 || onChange(formatDate(nextValue, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
            autoSwap: true,
            defaultTime: props2.defaultTime
          }), {
            dayjsValue: nextValue.map(function (v, i) {
              var _props2$defaultTime;
              return parseToDayjs(v, formatRef.value.format, void 0, void 0, (_props2$defaultTime = props2.defaultTime) === null || _props2$defaultTime === void 0 ? void 0 : _props2$defaultTime[i]);
            }),
            trigger: "pick"
          });
        }
      }
      if (Array.isArray(props2.disabled)) return;
      if (!isFirstValueSelected.value) {
        var nextIndex = notValidIndex;
        if (nextIndex === -1) nextIndex = activeIndex.value ? 0 : 1;
        activeIndex.value = nextIndex;
        isFirstValueSelected.value = nextValue.some(function (v) {
          return !!v;
        });
      } else {
        popupVisible.value = false;
      }
    }
    function onJumperClick(_ref3) {
      var trigger = _ref3.trigger,
        partial = _ref3.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      var monthCountMap = {
        date: 1,
        week: 1,
        month: 12,
        quarter: 12,
        year: 120
      };
      var monthCount = monthCountMap[props2.mode] || 0;
      var current = new Date(year.value[partialIndex], month.value[partialIndex]);
      var next = null;
      if (trigger === "prev") {
        next = subtractMonth(current, monthCount);
      } else if (trigger === "current") {
        next = new Date();
      } else if (trigger === "next") {
        next = addMonth(current, monthCount);
      }
      var nextYear = _toConsumableArray(year.value);
      nextYear[partialIndex] = next.getFullYear();
      var nextMonth = _toConsumableArray(month.value);
      nextMonth[partialIndex] = next.getMonth();
      var onlyYearSelect = ["year", "quarter", "month"].includes(props2.mode);
      var correctedDate = dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
      nextYear = correctedDate.nextYear;
      nextMonth = correctedDate.nextMonth;
      var yearChanged = year.value[partialIndex] !== nextYear[partialIndex];
      var monthChanged = month.value[partialIndex] !== nextMonth[partialIndex];
      year.value = nextYear;
      month.value = nextMonth;
      if (yearChanged) {
        var _props2$onYearChange;
        (_props2$onYearChange = props2.onYearChange) === null || _props2$onYearChange === void 0 || _props2$onYearChange.call(props2, {
          partial: partial,
          year: nextYear[partialIndex],
          date: value.value.map(function (v) {
            return dayjs(v).toDate();
          }),
          trigger: trigger === "current" ? "today" : "year-".concat(triggerMap[trigger])
        });
      }
      if (monthChanged) {
        var _props2$onMonthChange;
        (_props2$onMonthChange = props2.onMonthChange) === null || _props2$onMonthChange === void 0 || _props2$onMonthChange.call(props2, {
          partial: partial,
          month: nextMonth[partialIndex],
          date: value.value.map(function (v) {
            return dayjs(v).toDate();
          }),
          trigger: trigger === "current" ? "today" : "month-".concat(triggerMap[trigger])
        });
      }
    }
    function onTimePickerChange(val) {
      var _extractTimeObj = extractTimeObj(val),
        hours = _extractTimeObj.hours,
        minutes = _extractTimeObj.minutes,
        seconds = _extractTimeObj.seconds,
        milliseconds = _extractTimeObj.milliseconds,
        meridiem = _extractTimeObj.meridiem;
      var nextInputValue = _toConsumableArray(inputValue.value);
      var changedInputValue = nextInputValue[activeIndex.value];
      var currentDate = !dayjs(changedInputValue, formatRef.value.format).isValid() ? dayjs().year(year.value[activeIndex.value]).month(month.value[activeIndex.value]) : dayjs(changedInputValue, formatRef.value.format);
      var nextHours = hours;
      if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
      if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
      var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
      nextInputValue[activeIndex.value] = nextDate;
      var nextTime = _toConsumableArray(time.value);
      nextTime[activeIndex.value] = val;
      time.value = nextTime;
      isSelected.value = true;
      inputValue.value = formatDate(nextInputValue, {
        format: formatRef.value.format
      });
      cacheValue.value = formatDate(nextInputValue, {
        format: formatRef.value.format
      });
    }
    var confirmValueChange = function confirmValueChange(e) {
      var nextValue = _toConsumableArray(inputValue.value);
      var notValidIndex = nextValue.findIndex(function (v) {
        return !v || !isValidDate(v, formatRef.value.format);
      });
      if (notValidIndex === -1 && nextValue.length === 2) {
        if (!isFirstValueSelected.value && parseToDayjs(nextValue[0], formatRef.value.format).isAfter(parseToDayjs(nextValue[1], formatRef.value.format))) {
          nextValue[activeIndex.value ? 0 : 1] = "";
          cacheValue.value = nextValue;
          inputValue.value = nextValue;
        } else {
          var _props2$onConfirm;
          var formattedValue = formatDate(nextValue, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
            autoSwap: true,
            defaultTime: props2.defaultTime
          });
          var isSame = Array.isArray(formattedValue) && formattedValue.length === 2 && formattedValue[0] != null && formattedValue[1] != null && value.value[0] != null && value.value[1] != null && dayjs(formattedValue[0]).valueOf() === dayjs(value.value[0]).valueOf() && dayjs(formattedValue[1]).valueOf() === dayjs(value.value[1]).valueOf();
          props2 === null || props2 === void 0 || (_props2$onConfirm = props2.onConfirm) === null || _props2$onConfirm === void 0 || _props2$onConfirm.call(props2, {
            date: nextValue.map(function (v) {
              return dayjs(v).toDate();
            }),
            e: e || null,
            partial: activeIndex.value ? "end" : "start"
          });
          if (!isSame) {
            onChange === null || onChange === void 0 || onChange(formattedValue, {
              dayjsValue: nextValue.map(function (v, i) {
                var _props2$defaultTime2;
                return parseToDayjs(v, formatRef.value.format, void 0, void 0, (_props2$defaultTime2 = props2.defaultTime) === null || _props2$defaultTime2 === void 0 ? void 0 : _props2$defaultTime2[i]);
              }),
              trigger: "confirm"
            });
          }
        }
      }
    };
    function onConfirmClick(_ref4) {
      var e = _ref4.e;
      confirmValueChange(e);
      var nextValue = _toConsumableArray(inputValue.value);
      var notValidIndex = nextValue.findIndex(function (v) {
        return !v || !isValidDate(v, formatRef.value.format);
      });
      if (Array.isArray(props2.disabled)) {
        popupVisible.value = false;
      } else if (!isFirstValueSelected.value) {
        var nextIndex = notValidIndex;
        if (nextIndex === -1) nextIndex = activeIndex.value ? 0 : 1;
        activeIndex.value = nextIndex;
        isFirstValueSelected.value = nextValue.some(function (v) {
          return !!v;
        });
      } else if (nextValue.length === 2) {
        popupVisible.value = false;
      }
    }
    function onPresetClick(preset, context) {
      var presetValue = preset;
      if (isFunction(preset)) {
        presetValue = preset();
      }
      if (!isArray(presetValue)) {
        console.error("preset: ".concat(preset, " \u9884\u8BBE\u503C\u5FC5\u987B\u662F\u6570\u7EC4!"));
      } else {
        var _props2$onPresetClick;
        onChange === null || onChange === void 0 || onChange(formatDate(presetValue, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.valueType,
          autoSwap: true,
          defaultTime: props2.defaultTime
        }), {
          dayjsValue: presetValue.map(function (p, i) {
            var _props2$defaultTime3;
            return parseToDayjs(p, formatRef.value.format, void 0, void 0, (_props2$defaultTime3 = props2.defaultTime) === null || _props2$defaultTime3 === void 0 ? void 0 : _props2$defaultTime3[i]);
          }),
          trigger: "preset"
        });
        inputValue.value = formatDate(presetValue, {
          format: formatRef.value.format
        });
        popupVisible.value = false;
        (_props2$onPresetClick = props2.onPresetClick) === null || _props2$onPresetClick === void 0 || _props2$onPresetClick.call(props2, context);
      }
    }
    function onYearChange(nextVal, _ref5) {
      var _props2$onYearChange2;
      var partial = _ref5.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      if (props2.enableTimePicker) partialIndex = activeIndex.value;
      var nextYear = _toConsumableArray(year.value);
      var nextMonth = _toConsumableArray(month.value);
      nextYear[partialIndex] = nextVal;
      var onlyYearSelect = ["year", "quarter", "month"].includes(props2.mode);
      var correctedDate = dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
      nextYear = correctedDate.nextYear;
      nextMonth = correctedDate.nextMonth;
      year.value = nextYear;
      if (!onlyYearSelect) month.value = nextMonth;
      (_props2$onYearChange2 = props2.onYearChange) === null || _props2$onYearChange2 === void 0 || _props2$onYearChange2.call(props2, {
        partial: partial,
        year: nextYear[partialIndex],
        date: value.value.map(function (v) {
          return dayjs(v).toDate();
        }),
        trigger: "year-select"
      });
    }
    function onMonthChange(nextVal, _ref6) {
      var _props2$onMonthChange2;
      var partial = _ref6.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      if (props2.enableTimePicker) partialIndex = activeIndex.value;
      var nextMonth = _toConsumableArray(month.value);
      nextMonth[partialIndex] = nextVal;
      if (year.value[0] === year.value[1]) {
        if (partialIndex === 0) {
          if (nextMonth[1] <= nextMonth[0]) {
            nextMonth[1] = nextMonth[0] + 1;
            if (nextMonth[1] === 12) {
              var _year$value, _year$value2;
              nextMonth[1] = 0;
              year.value = [(_year$value = year.value) === null || _year$value === void 0 ? void 0 : _year$value[0], ((_year$value2 = year.value) === null || _year$value2 === void 0 ? void 0 : _year$value2[1]) + 1];
            }
          }
        }
        if (partialIndex === 1) {
          nextMonth[0] = Math.min(nextMonth[0], nextMonth[1]);
          if (nextMonth[0] >= nextMonth[1]) {
            nextMonth[0] -= 1;
            if (nextMonth[0] === -1) {
              var _year$value3, _year$value4;
              nextMonth[0] = 11;
              year.value = [((_year$value3 = year.value) === null || _year$value3 === void 0 ? void 0 : _year$value3[0]) - 1, (_year$value4 = year.value) === null || _year$value4 === void 0 ? void 0 : _year$value4[1]];
            }
          }
        }
      }
      month.value = nextMonth;
      (_props2$onMonthChange2 = props2.onMonthChange) === null || _props2$onMonthChange2 === void 0 || _props2$onMonthChange2.call(props2, {
        partial: partial,
        month: nextMonth[partialIndex],
        date: value.value.map(function (v) {
          return dayjs(v).toDate();
        }),
        trigger: "month-select"
      });
    }
    var panelProps = computed(function () {
      return {
        hoverValue: isHoverCell.value ? inputValue.value : [],
        value: isSelected.value ? cacheValue.value : value.value,
        isFirstValueSelected: isFirstValueSelected.value,
        activeIndex: activeIndex.value,
        year: year.value,
        month: month.value,
        format: formatRef.value.format,
        mode: props2.mode,
        presets: props2.presets,
        time: time.value,
        disableDate: props2.disableDate,
        disabled: isDisabled.value,
        disableTime: props2.disableTime,
        firstDayOfWeek: props2.firstDayOfWeek,
        timePickerProps: props2.timePickerProps,
        enableTimePicker: props2.enableTimePicker,
        presetsPlacement: props2.presetsPlacement,
        popupVisible: popupVisible.value,
        range: props2.range,
        panelPreselection: props2.panelPreselection,
        cancelRangeSelectLimit: props2.cancelRangeSelectLimit,
        needConfirm: props2.needConfirm,
        onCellClick: onCellClick,
        onCellMouseEnter: onCellMouseEnter,
        onCellMouseLeave: onCellMouseLeave,
        onJumperClick: onJumperClick,
        onConfirmClick: onConfirmClick,
        onPresetClick: onPresetClick,
        onYearChange: onYearChange,
        onMonthChange: onMonthChange,
        onTimePickerChange: onTimePickerChange,
        cell: props2.cell
      };
    });
    return function () {
      return createVNode("div", {
        "class": COMPONENT_NAME.value
      }, [createVNode(RangeInputPopup, {
        "readonly": isReadOnly.value,
        "disabled": isDisabled.value,
        "label": props2.label,
        "status": props2.status,
        "tips": props2.tips || slots.tips,
        "inputValue": inputValue.value,
        "popupProps": popupProps.value,
        "rangeInputProps": rangeInputProps.value,
        "popupVisible": popupVisible.value,
        "panel": function panel() {
          return createVNode(TRangePanel, panelProps.value, slots);
        }
      }, null)]);
    };
  }
});

export { _DateRangePicker as default };
