// tdesign-vue-next/esm/date-picker/components/panel/SinglePanel.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { defineComponent, computed, createVNode } from 'vue';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'lodash-es';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/typeof';
import '../../../_chunks/dep-651bdc55.js';
import { u as usePrefixClass } from '../../../_chunks/dep-59f77e78.js';
import '@babel/runtime/helpers/slicedToArray';
import '../../../_chunks/dep-204ae706.js';
import TPanelContent from './PanelContent.js';
import TExtraContent from './ExtraContent.js';
import { g as getDefaultFormat, p as parseToDayjs } from '../../../_chunks/dep-f54e73f7.js';
import { useDisableDate } from '../../hooks/useDisableDate.js';
import 'tdesign-icons-vue-next';
import 'dayjs';
import { useTableData } from '../../hooks/useTableData.js';
import '../../../_chunks/dep-a7cf6cc7.js';
import { useConfig } from '../../../config-provider/hooks/useConfig.js';
import '../../../_chunks/dep-76473034.js';
import '../../../config-provider/utils/context.js';
import '../../../_chunks/dep-5c1cdd0b.js';
import '@babel/runtime/helpers/createClass';
import '@babel/runtime/helpers/classCallCheck';
import '../base/Header.js';
import '../../../pagination/index.js';
import '../../../pagination/pagination.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../../../_chunks/dep-eff3e90f.js';
import '../../../_chunks/dep-a7922f61.js';
import '../../../_chunks/dep-69502ff8.js';
import '../../../_chunks/dep-c6879e67.js';
import '../../../_chunks/dep-3d495891.js';
import '../../../_chunks/dep-f32cc39c.js';
import '../../../input-number/index.js';
import '../../../input-number/input-number.js';
import '../../../button/index.js';
import '../../../button/button.js';
import '../../../loading/index.js';
import '../../../loading/directive.js';
import '../../../loading/plugin.js';
import '../../../_chunks/dep-d32d8561.js';
import '../../../loading/icon/gradient.js';
import '../../../_chunks/dep-d15b7e74.js';
import '../../../_chunks/dep-a9edb4d8.js';
import '../../../_chunks/dep-7daa5fe4.js';
import '../../../loading/props.js';
import '../../../_chunks/dep-5dfdb698.js';
import 'tdesign-vue-next/esm/common/style/web/components/loading/_index.less';
import '../../../button/props.js';
import '../../../_chunks/dep-ed0885e7.js';
import '../../../_chunks/dep-16e13fa9.js';
import 'tdesign-vue-next/esm/common/style/web/components/button/_index.less';
import '../../../input/index.js';
import '../../../input/input.js';
import '../../../input/props.js';
import '../../../_chunks/dep-0d255418.js';
import '../../../input/hooks/useInput.js';
import '../../../form/constants/index.js';
import '../../../input/hooks/useLengthLimit.js';
import '../../../_chunks/dep-f2301d1c.js';
import '../../../input/hooks/useInputEventHandler.js';
import '../../../input/hooks/useInputWidth.js';
import '../../../_chunks/dep-f5c5752f.js';
import '../../../input/input-group.js';
import '../../../input/input-group-props.js';
import 'tdesign-vue-next/esm/common/style/web/components/input/_index.less';
import '../../../input-number/props.js';
import '../../../_chunks/dep-ff412569.js';
import '../../../input-number/hooks/useInputNumber.js';
import '../../../_chunks/dep-b21bd6d2.js';
import '../../../_chunks/dep-75be08e7.js';
import 'tdesign-vue-next/esm/common/style/web/components/input-number/_index.less';
import '../../../select/index.js';
import '../../../select/select.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '../../../common-components/fake-arrow.js';
import '../../../select-input/index.js';
import '../../../select-input/select-input.js';
import '../../../popup/index.js';
import '../../../popup/popup.js';
import '@popperjs/core';
import '../../../popup/container.js';
import '../../../popup/props.js';
import 'tdesign-vue-next/esm/common/style/web/components/popup/_index.less';
import '../../../select-input/props.js';
import '../../../select-input/hooks/useMultiple.js';
import '../../../tag-input/index.js';
import '../../../tag-input/tag-input.js';
import '../../../tag-input/props.js';
import '../../../tag-input/hooks/useDragSorter.js';
import '../../../tag-input/hooks/useHover.js';
import '../../../tag-input/hooks/useTagScroll.js';
import '../../../tag-input/hooks/useTagList.js';
import '../../../tag/index.js';
import '../../../tag/tag.js';
import 'tinycolor2';
import '../../../tag/props.js';
import '../../../tag/check-tag.js';
import '../../../tag/check-tag-props.js';
import '../../../_chunks/dep-8d513880.js';
import '../../../tag/check-tag-group.js';
import '../../../tag/check-tag-group-props.js';
import 'tdesign-vue-next/esm/common/style/web/components/tag/_index.less';
import 'tdesign-vue-next/esm/common/style/web/components/tag-input/_index.less';
import '../../../select-input/hooks/useOverlayInnerStyle.js';
import '../../../select-input/hooks/useSingle.js';
import 'tdesign-vue-next/esm/common/style/web/components/select-input/_index.less';
import '../../../select/components/select-panel.js';
import '../../../select/option.js';
import '../../../select/option-props.js';
import '../../../checkbox/index.js';
import '../../../checkbox/checkbox.js';
import '../../../checkbox/props.js';
import '../../../checkbox/constants/index.js';
import '../../../checkbox/hooks/useCheckboxLazyLoad.js';
import '../../../_chunks/dep-2cabe3fe.js';
import '../../../checkbox/hooks/useKeyboardEvent.js';
import '../../../checkbox/group.js';
import '../../../checkbox/checkbox-group-props.js';
import '../../../_chunks/dep-d4579130.js';
import 'tdesign-vue-next/esm/common/style/web/components/checkbox/_index.less';
import '../../../select/utils/index.js';
import '../../../select/constants/index.js';
import '../../../select/option-group.js';
import '../../../select/option-group-props.js';
import '../../../select/props.js';
import '../../../select/hooks/usePanelVirtualScroll.js';
import '../../../select/hooks/useKeyboardControl.js';
import '../../../select/hooks/useSelectOptions.js';
import 'tdesign-vue-next/esm/common/style/web/components/select/_index.less';
import '../../../input-adornment/index.js';
import '../../../input-adornment/input-adornment.js';
import '../../../input-adornment/props.js';
import 'tdesign-vue-next/esm/common/style/web/components/input-adornment/_index.less';
import '../../../pagination/props.js';
import '../../../pagination/hooks/useMoreAction.js';
import '../../../pagination/hooks/usePaginationClasses.js';
import '../../../pagination/pagination-mini.js';
import '../../../pagination/pagination-mini-props.js';
import '../../../tooltip/index.js';
import '../../../tooltip/tooltip.js';
import '../../../tooltip/props.js';
import '../../../tooltip/utils/index.js';
import 'tdesign-vue-next/esm/common/style/web/components/tooltip/_index.less';
import 'tdesign-vue-next/esm/common/style/web/components/pagination/_index.less';
import '../../hooks/useSelectRange.js';
import '../base/Table.js';
import '../base/Cell.js';
import '../../../time-picker/index.js';
import '../../../time-picker/time-picker.js';
import '../../../_chunks/dep-7a2c5955.js';
import '../../../time-picker/panel/time-picker-panel.js';
import '../../../_chunks/dep-ea86b40b.js';
import '../../../time-picker/props.js';
import '../../../_chunks/dep-c817a169.js';
import '../../../_chunks/dep-3331fba2.js';
import '../../../time-picker/time-range-picker.js';
import '../../../range-input/index.js';
import '../../../range-input/range-input.js';
import '../../../range-input/props.js';
import '../../../range-input/range-input-popup.js';
import '../../../range-input/range-input-popup-props.js';
import 'tdesign-vue-next/esm/common/style/web/components/range-input/_index.less';
import '../../../time-picker/time-range-picker-props.js';
import 'tdesign-vue-next/esm/common/style/web/components/time-picker/_index.less';
import '../../utils/index.js';
import '../base/Footer.js';
import '@babel/runtime/helpers/objectDestructuringEmpty';
import '@babel/runtime/helpers/extends';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TSinglePanel = defineComponent({
  name: "TSinglePanel",
  props: {
    disableDate: [Object, Array, Function],
    mode: {
      type: String,
      "default": "date"
    },
    format: String,
    presetsPlacement: {
      type: String,
      "default": "bottom"
    },
    value: [String, Number, Array, Date],
    timePickerProps: Object,
    presets: Object,
    enableTimePicker: Boolean,
    firstDayOfWeek: Number,
    year: Number,
    month: Number,
    time: String,
    popupVisible: Boolean,
    multiple: Boolean,
    needConfirm: Boolean,
    defaultTime: [String, Array],
    range: [Array, Function],
    onPanelClick: Function,
    onCellClick: Function,
    onCellMouseEnter: Function,
    onCellMouseLeave: Function,
    onJumperClick: Function,
    onConfirmClick: Function,
    onPresetClick: Function,
    onYearChange: Function,
    onMonthChange: Function,
    onTimePickerChange: Function,
    disableTime: Function,
    cell: Function
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots;
    var COMPONENT_NAME = usePrefixClass("date-picker__panel");
    var _useConfig = useConfig("datePicker"),
      globalConfig = _useConfig.globalConfig;
    var format = computed(function () {
      var _getDefaultFormat;
      return (_getDefaultFormat = getDefaultFormat({
        mode: props.mode,
        format: props.format,
        enableTimePicker: props.enableTimePicker
      })) === null || _getDefaultFormat === void 0 ? void 0 : _getDefaultFormat.format;
    });
    var disableDateOptions = computed(function () {
      return useDisableDate({
        format: format.value,
        mode: props.mode,
        disableDate: props.disableDate
      });
    });
    var tableData = computed(function () {
      return useTableData(_objectSpread({
        year: props.year,
        month: props.month,
        mode: props.mode,
        range: props.range,
        start: props.value ? parseToDayjs(props.multiple ? props.value[0] : props.value, format.value).toDate() : void 0,
        firstDayOfWeek: props.firstDayOfWeek || globalConfig.value.firstDayOfWeek,
        multiple: props.multiple,
        value: props.value
      }, disableDateOptions.value));
    });
    var panelContentProps = computed(function () {
      return {
        format: format.value,
        value: props.value,
        mode: props.mode,
        year: props.year,
        month: props.month,
        range: props.range,
        firstDayOfWeek: props.firstDayOfWeek || globalConfig.value.firstDayOfWeek,
        tableData: tableData.value,
        popupVisible: props.popupVisible,
        multiple: props.multiple,
        enableTimePicker: props.enableTimePicker,
        timePickerProps: props.timePickerProps,
        time: props.time,
        onMonthChange: props.onMonthChange,
        onYearChange: props.onYearChange,
        onJumperClick: props.onJumperClick,
        onCellClick: props.onCellClick,
        onCellMouseEnter: props.onCellMouseEnter,
        onCellMouseLeave: props.onCellMouseLeave,
        onTimePickerChange: props.onTimePickerChange,
        defaultTime: props.defaultTime,
        disableTime: props.disableTime,
        cell: props.cell
      };
    });
    var extraProps = computed(function () {
      return {
        presets: props.presets,
        enableTimePicker: props.enableTimePicker,
        presetsPlacement: props.presetsPlacement,
        onPresetClick: props.onPresetClick,
        onConfirmClick: props.onConfirmClick,
        selectedValue: props.value,
        needConfirm: props.needConfirm
      };
    });
    return function () {
      return createVNode("div", {
        "class": [COMPONENT_NAME.value, _defineProperty({}, "".concat(COMPONENT_NAME.value, "--direction-row"), ["left", "right"].includes(props.presetsPlacement))],
        "onClick": function onClick(e) {
          var _props$onPanelClick;
          return (_props$onPanelClick = props.onPanelClick) === null || _props$onPanelClick === void 0 ? void 0 : _props$onPanelClick.call(props, {
            e: e
          });
        }
      }, [["top", "left"].includes(props.presetsPlacement) ? createVNode(TExtraContent, extraProps.value, slots.presets ? {
        presets: slots.presets
      } : {}) : null, createVNode(TPanelContent, panelContentProps.value, slots), ["bottom", "right"].includes(props.presetsPlacement) ? createVNode(TExtraContent, extraProps.value, slots.presets ? {
        presets: slots.presets
      } : {}) : null]);
    };
  }
});

export { TSinglePanel as default };
