// tdesign-vue-next/cjs/dialog/dialog-card.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var dialog_dialogCardProps = require('./dialog-card-props.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-c7472ccd.js');
var index$1 = require('../_chunks/dep-0e56b1da.js');
var index = require('../_chunks/dep-b56e7ff5.js');
var index$2 = require('../_chunks/dep-125f71ff.js');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-d1a4892a.js');
var dialog_hooks_useAction = require('./hooks/useAction.js');
var dialog_props = require('./props.js');
var dialog_utils_index = require('./utils/index.js');
var configProvider_hooks_useConfig = require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-bf2e7501.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-37af9215.js');
require('../_chunks/dep-e2205a6f.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-d1d374b4.js');
require('../_chunks/dep-6f4a5c31.js');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('dayjs');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-2e9fed7b.js');
require('../_chunks/dep-a324ad6a.js');
require('../_chunks/dep-06e53606.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../button/index.js');
require('../button/button.js');
require('../loading/index.js');
require('../_chunks/dep-548039eb.js');
require('../loading/plugin.js');
require('../_chunks/dep-4a5eca86.js');
require('../loading/icon/gradient.js');
require('../_chunks/dep-fd403354.js');
require('../_chunks/dep-585d18dc.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-8308792f.js');
require('../_chunks/dep-d5e8d527.js');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-8e4f83bf.js');
require('../_chunks/dep-6c498eb2.js');
require('../loading/props.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('../_chunks/dep-f8f462e0.js');
require('../_chunks/dep-47af4fff.js');
require('../_chunks/dep-ea9fd5a2.js');
require('../_chunks/dep-a8f427e5.js');
require('../_chunks/dep-9997364c.js');
require('../_chunks/dep-1e2d078f.js');
require('../_chunks/dep-7d888cd2.js');
require('../_chunks/dep-ffd72e3b.js');
require('../_chunks/dep-6ed4388e.js');
require('../_chunks/dep-742210e1.js');
require('../button/props.js');
require('../_chunks/dep-a8de5365.js');
require('../_chunks/dep-850526b5.js');
require('../_chunks/dep-a90757f2.js');
require('../_chunks/dep-d3b20f18.js');
require('../_chunks/dep-3d9f41ae.js');
require('../_chunks/dep-117093bb.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _DialogCard = Vue.defineComponent({
  name: "TDialogCard",
  directives: {
    draggable: function draggable(el, binding) {
      if (el && binding && binding.value) {
        dialog_utils_index.initDragEvent(el);
      }
    }
  },
  props: _objectSpread(_objectSpread({}, dialog_props["default"]), dialog_dialogCardProps["default"]),
  setup: function setup(props, _ref) {
    var expose = _ref.expose;
    var rootRef = Vue.ref(null);
    var COMPONENT_NAME = index.usePrefixClass("dialog");
    var classPrefix = index.usePrefixClass();
    var renderTNodeJSX = index$1.useTNodeJSX();
    var renderContent = index$1.useContent();
    var _useConfig = configProvider_hooks_useConfig.useConfig("dialog"),
      globalConfig = _useConfig.globalConfig;
    var _useGlobalIcon = index$2.useGlobalIcon({
        CloseIcon: tdesignIconsVueNext.CloseIcon,
        InfoCircleFilledIcon: tdesignIconsVueNext.InfoCircleFilledIcon,
        CheckCircleFilledIcon: tdesignIconsVueNext.CheckCircleFilledIcon,
        ErrorCircleFilledIcon: tdesignIconsVueNext.ErrorCircleFilledIcon
      }),
      CloseIcon = _useGlobalIcon.CloseIcon,
      InfoCircleFilledIcon = _useGlobalIcon.InfoCircleFilledIcon,
      CheckCircleFilledIcon = _useGlobalIcon.CheckCircleFilledIcon,
      ErrorCircleFilledIcon = _useGlobalIcon.ErrorCircleFilledIcon;
    var _toRefs = Vue.toRefs(props),
      cancelBtn = _toRefs.cancelBtn,
      confirmBtn = _toRefs.confirmBtn,
      confirmLoading = _toRefs.confirmLoading;
    var confirmBtnAction = function confirmBtnAction(e) {
      var _props$onConfirm;
      return (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 ? void 0 : _props$onConfirm.call(props, {
        e: e
      });
    };
    var cancelBtnAction = function cancelBtnAction(e) {
      var _props$onCancel;
      return (_props$onCancel = props.onCancel) === null || _props$onCancel === void 0 ? void 0 : _props$onCancel.call(props, {
        e: e
      });
    };
    var _useAction = dialog_hooks_useAction.useAction({
        confirmBtnAction: confirmBtnAction,
        cancelBtnAction: cancelBtnAction
      }),
      getConfirmBtn = _useAction.getConfirmBtn,
      getCancelBtn = _useAction.getCancelBtn;
    var isModeLess = Vue.computed(function () {
      return props.mode === "modeless";
    });
    var isDraggableMode = Vue.computed(function () {
      return ["modal", "modeless"].includes(props.mode);
    });
    var isFullScreen = Vue.computed(function () {
      return props.mode === "full-screen";
    });
    var closeBtnAction = function closeBtnAction(e) {
      var _props$onCloseBtnClic;
      return props === null || props === void 0 || (_props$onCloseBtnClic = props.onCloseBtnClick) === null || _props$onCloseBtnClic === void 0 ? void 0 : _props$onCloseBtnClic.call(props, {
        e: e
      });
    };
    var onStopDown = function onStopDown(e) {
      if (isDraggableMode.value && props !== null && props !== void 0 && props.draggable) e.stopPropagation();
    };
    var resetPosition = function resetPosition() {
      if (!rootRef.value && isModeLess.value && props.draggable) return;
      rootRef.value.style.position = "relative";
      rootRef.value.style.left = "unset";
      rootRef.value.style.top = "unset";
    };
    expose({
      $el: rootRef,
      resetPosition: resetPosition
    });
    var dialogClass = Vue.computed(function () {
      var dialogClass2 = ["".concat(COMPONENT_NAME.value), "".concat(COMPONENT_NAME.value, "__modal-").concat(props.theme), isDraggableMode.value && props.draggable && "".concat(COMPONENT_NAME.value, "--draggable"), props.dialogClassName];
      if (isFullScreen.value) {
        dialogClass2.push("".concat(COMPONENT_NAME.value, "__fullscreen"));
      } else {
        dialogClass2.push("".concat(COMPONENT_NAME.value, "--default"), "".concat(COMPONENT_NAME.value, "--").concat(props.placement));
      }
      return dialogClass2;
    });
    var dialogStyle = Vue.computed(function () {
      return !isFullScreen.value ? _objectSpread({
        width: dialog_utils_index.getCSSValue(props.width)
      }, props.dialogStyle) : _objectSpread({}, props.dialogStyle);
    });
    var renderCard = function renderCard() {
      var confirmBtnLoading = Vue.computed(function () {
        var _confirmBtn$value;
        return ((_confirmBtn$value = confirmBtn.value) === null || _confirmBtn$value === void 0 ? void 0 : _confirmBtn$value.loading) || confirmLoading.value;
      });
      var defaultFooter = Vue.createVNode("div", null, [getCancelBtn({
        cancelBtn: cancelBtn.value,
        globalCancel: globalConfig.value.cancel,
        className: "".concat(COMPONENT_NAME.value, "__cancel")
      }), getConfirmBtn({
        theme: props === null || props === void 0 ? void 0 : props.theme,
        confirmBtn: confirmBtn.value,
        globalConfirm: globalConfig.value.confirm,
        globalConfirmBtnTheme: globalConfig.value.confirmBtnTheme,
        className: "".concat(COMPONENT_NAME.value, "__confirm"),
        confirmLoading: confirmBtnLoading.value
      })]);
      var footerContent = renderTNodeJSX("footer", defaultFooter);
      var renderHeader = function renderHeader() {
        var _renderTNodeJSX;
        var header = (_renderTNodeJSX = renderTNodeJSX("header", Vue.createVNode("h5", {
          "class": "title"
        }, null))) !== null && _renderTNodeJSX !== void 0 ? _renderTNodeJSX : false;
        var headerClassName = isFullScreen.value ? ["".concat(COMPONENT_NAME.value, "__header"), "".concat(COMPONENT_NAME.value, "__header--fullscreen")] : "".concat(COMPONENT_NAME.value, "__header");
        var closeClassName = isFullScreen.value ? ["".concat(COMPONENT_NAME.value, "__close"), "".concat(COMPONENT_NAME.value, "__close--fullscreen")] : "".concat(COMPONENT_NAME.value, "__close");
        var getIcon = function getIcon() {
          var icon = {
            info: Vue.createVNode(InfoCircleFilledIcon, {
              "class": "".concat(classPrefix.value, "-is-info")
            }, null),
            warning: Vue.createVNode(ErrorCircleFilledIcon, {
              "class": "".concat(classPrefix.value, "-is-warning")
            }, null),
            danger: Vue.createVNode(ErrorCircleFilledIcon, {
              "class": "".concat(classPrefix.value, "-is-error")
            }, null),
            success: Vue.createVNode(CheckCircleFilledIcon, {
              "class": "".concat(classPrefix.value, "-is-success")
            }, null)
          };
          return icon[props === null || props === void 0 ? void 0 : props.theme];
        };
        return (header || (props === null || props === void 0 ? void 0 : props.closeBtn)) && Vue.createVNode("div", {
          "class": headerClassName
        }, [Vue.createVNode("div", {
          "class": "".concat(COMPONENT_NAME.value, "__header-content")
        }, [getIcon(), header]), props !== null && props !== void 0 && props.closeBtn ? Vue.createVNode("span", {
          "class": closeClassName,
          "onClick": closeBtnAction,
          "onMousedown": onStopDown
        }, [renderTNodeJSX("closeBtn", Vue.createVNode(CloseIcon, null, null))]) : null]);
      };
      var renderBody = function renderBody() {
        var body = renderContent("default", "body");
        var bodyClassName = (props === null || props === void 0 ? void 0 : props.theme) === "default" ? ["".concat(COMPONENT_NAME.value, "__body")] : ["".concat(COMPONENT_NAME.value, "__body__icon")];
        if (isFullScreen.value && footerContent) {
          bodyClassName.push("".concat(COMPONENT_NAME.value, "__body--fullscreen"));
        } else if (isFullScreen.value) {
          bodyClassName.push("".concat(COMPONENT_NAME.value, "__body--fullscreen--without-footer"));
        }
        return Vue.createVNode("div", {
          "class": bodyClassName,
          "onMousedown": onStopDown
        }, [body]);
      };
      var renderFooter = function renderFooter() {
        var footerClassName = isFullScreen.value ? ["".concat(COMPONENT_NAME.value, "__footer"), "".concat(COMPONENT_NAME.value, "__footer--fullscreen")] : "".concat(COMPONENT_NAME.value, "__footer");
        return footerContent && Vue.createVNode("div", {
          "class": footerClassName,
          "onMousedown": onStopDown
        }, [footerContent]);
      };
      return Vue.createVNode(Vue.Fragment, null, [renderHeader(), renderBody(), !!props.footer && renderFooter()]);
    };
    return function () {
      return Vue.withDirectives(Vue.createVNode("div", {
        "key": "dialog",
        "ref": rootRef,
        "class": dialogClass.value,
        "style": dialogStyle.value
      }, [renderCard()]), [[Vue.resolveDirective("draggable"), isDraggableMode.value && props.draggable]]);
    };
  }
});

exports["default"] = _DialogCard;
