// tdesign-vue-next/es/color-picker/components/panel/header.mjs
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { isVNode, defineComponent, ref, watch, createVNode } from 'vue';
import { _ as _defineProperty } from '../../../_chunks/dep-edde6136.mjs';
import { c as COLOR_MODES } from '../../../_chunks/dep-8d10c4a9.mjs';
import '../../../_chunks/dep-d6195387.mjs';
import { RadioGroup, RadioButton } from '../../../radio/index.mjs';
import props from '../../color-picker-panel-props.mjs';
import { useBaseClassName } from '../../hooks/index.mjs';
import { useConfig } from '../../../config-provider/hooks/useConfig.mjs';
import '../../../_chunks/dep-cb344d2d.mjs';
import 'lodash-es';
import '../../../_chunks/dep-fe2215b7.mjs';
import '../../../_chunks/dep-c13a05a2.mjs';
import '../../../_chunks/dep-478d8d87.mjs';
import '../../../_chunks/dep-80ab80bc.mjs';
import '../../../_chunks/dep-1d48345c.mjs';
import '../../../_chunks/dep-39047c40.mjs';
import '../../../_chunks/dep-4f3de704.mjs';
import '../../../_chunks/dep-15b76bbf.mjs';
import '../../../config-provider/utils/context.mjs';
import '../../../_chunks/dep-fa2e9d13.mjs';
import '../../../_chunks/dep-c22f10cd.mjs';
import '../../../_chunks/dep-2512dba6.mjs';
import '../../../_chunks/dep-61161446.mjs';
import '../../../_chunks/dep-671c8553.mjs';
import '../../../radio/radio.mjs';
import '../../../radio/props.mjs';
import '../../../radio/constants/index.mjs';
import '../../../_chunks/dep-58da668e.mjs';
import '../../../_chunks/dep-15e8aadf.mjs';
import '../../../radio/group.mjs';
import '../../../_chunks/dep-bcb347a6.mjs';
import '../../../radio/radio-group-props.mjs';
import '../../../radio/radio-button.mjs';
import '../../../radio/hooks/index.mjs';
import '../../../radio/hooks/useKeyboard.mjs';
import '../../../watermark/hooks/index.mjs';
import './style/css.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var PanelHeader = defineComponent({
  name: "PanelHeader",
  props: _objectSpread(_objectSpread({}, props), {}, {
    mode: {
      type: String,
      "default": "color"
    },
    onModeChange: {
      type: Function,
      "default": function _default() {
        return function () {};
      }
    }
  }),
  setup: function setup(props2) {
    var _useConfig = useConfig("colorPicker"),
      globalConfig = _useConfig.globalConfig;
    var baseClassName = useBaseClassName();
    var modeValue = ref(props2.mode);
    var handleModeChange = function handleModeChange(v) {
      return props2.onModeChange(v);
    };
    watch(function () {
      return props2.mode;
    }, function (v) {
      return modeValue.value = v;
    });
    return function () {
      var _slot2;
      var _props2$colorModes;
      if (((_props2$colorModes = props2.colorModes) === null || _props2$colorModes === void 0 ? void 0 : _props2$colorModes.length) === 1) {
        return null;
      }
      return createVNode("div", {
        "class": "".concat(baseClassName.value, "__head")
      }, [createVNode("div", {
        "class": "".concat(baseClassName.value, "__mode")
      }, [createVNode(RadioGroup, {
        "variant": "default-filled",
        "size": "small",
        "modelValue": modeValue.value,
        "onUpdate:modelValue": function onUpdateModelValue($event) {
          return modeValue.value = $event;
        },
        "onChange": handleModeChange
      }, _isSlot(_slot2 = Object.keys(COLOR_MODES).map(function (key) {
        var _slot;
        return createVNode(RadioButton, {
          "key": key,
          "value": key
        }, _isSlot(_slot = Reflect.get(globalConfig.value, COLOR_MODES[key])) ? _slot : {
          "default": function _default() {
            return [_slot];
          }
        });
      })) ? _slot2 : {
        "default": function _default() {
          return [_slot2];
        }
      })])]);
    };
  }
});

export { PanelHeader as default };
