// tdesign-vue-next/es/watermark/props.mjs
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

var props = {
  alpha: {
    type: Number,
    "default": 1
  },
  content: {
    type: [String, Function]
  },
  "default": {
    type: [String, Function]
  },
  height: {
    type: Number
  },
  isRepeat: {
    type: Boolean,
    "default": true
  },
  layout: {
    type: String,
    "default": "rectangular",
    validator: function validator(val) {
      if (!val) return true;
      return ["rectangular", "hexagonal"].includes(val);
    }
  },
  lineSpace: {
    type: Number,
    "default": 16
  },
  movable: Boolean,
  moveInterval: {
    type: Number,
    "default": 3e3
  },
  offset: {
    type: Array
  },
  removable: {
    type: Boolean,
    "default": true
  },
  rotate: {
    type: Number,
    "default": -22
  },
  watermarkContent: {
    type: [Object, Array]
  },
  width: {
    type: Number
  },
  x: {
    type: Number
  },
  y: {
    type: Number
  },
  zIndex: {
    type: Number
  }
};

export { props as default };
