// tdesign-vue-next/cjs/table/components/filter-controller.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _typeof = require('@babel/runtime/helpers/typeof');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var popup_index = require('../../popup/index.js');
var checkbox_index = require('../../checkbox/index.js');
var radio_index = require('../../radio/index.js');
var input_index = require('../../input/index.js');
var button_index = require('../../button/index.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../../_chunks/dep-c7472ccd.js');
var index = require('../../_chunks/dep-0e56b1da.js');
var configProvider_hooks_useConfig = require('../../config-provider/hooks/useConfig.js');
var index$1 = require('../../_chunks/dep-125f71ff.js');
require('@babel/runtime/helpers/slicedToArray');
require('../../_chunks/dep-d1a4892a.js');
var isEmpty = require('../../_chunks/dep-954dbe59.js');
var isFunction = require('../../_chunks/dep-37af9215.js');
require('../../popup/popup.js');
require('@popperjs/core');
require('../../_chunks/dep-eb214be8.js');
require('../../_chunks/dep-11e6da14.js');
require('../../_chunks/dep-e2205a6f.js');
require('../../_chunks/dep-d1d374b4.js');
require('../../_chunks/dep-6f4a5c31.js');
require('../../_chunks/dep-6c498eb2.js');
require('../../_chunks/dep-b56e7ff5.js');
require('../../_chunks/dep-fd403354.js');
require('../../_chunks/dep-d149369c.js');
require('../../_chunks/dep-09b19404.js');
require('../../_chunks/dep-445d3f65.js');
require('../../_chunks/dep-2333b13f.js');
require('../../_chunks/dep-5729cb14.js');
require('../../popup/container.js');
require('../../popup/props.js');
require('../../_chunks/dep-6344c74f.js');
require('../../_chunks/dep-bf2e7501.js');
require('../../_chunks/dep-c644d54f.js');
require('../../_chunks/dep-07f46433.js');
require('../../_chunks/dep-0b08f67c.js');
require('dayjs');
require('../../_chunks/dep-eb1f700b.js');
require('../../_chunks/dep-9399a4e2.js');
require('../../_chunks/dep-58ff00b9.js');
require('../../_chunks/dep-0c9948d8.js');
require('../../_chunks/dep-2e9fed7b.js');
require('../../_chunks/dep-a324ad6a.js');
require('../../_chunks/dep-06e53606.js');
require('../../_chunks/dep-bde8abb8.js');
require('../../_chunks/dep-41cc84a3.js');
require('../../_chunks/dep-c51f4345.js');
require('../../_chunks/dep-e2c29461.js');
require('../../_chunks/dep-52d3718c.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../../_chunks/dep-d8ab5fb5.js');
require('../../_chunks/dep-cde052f0.js');
require('../../_chunks/dep-a3501e24.js');
require('../../_chunks/dep-c8d40779.js');
require('../../_chunks/dep-388cb46f.js');
require('../../_chunks/dep-89d65b78.js');
require('../../_chunks/dep-dc9d7b38.js');
require('../../_chunks/dep-742210e1.js');
require('../../checkbox/checkbox.js');
require('../../checkbox/props.js');
require('../../_chunks/dep-a8de5365.js');
require('../../_chunks/dep-850526b5.js');
require('../../_chunks/dep-4725a846.js');
require('../../_chunks/dep-a90757f2.js');
require('../../checkbox/constants/index.js');
require('../../checkbox/hooks/useCheckboxLazyLoad.js');
require('../../_chunks/dep-b4842612.js');
require('../../checkbox/hooks/useKeyboardEvent.js');
require('../../_chunks/dep-1177d395.js');
require('../../checkbox/group.js');
require('../../checkbox/checkbox-group-props.js');
require('../../_chunks/dep-e53cd0a5.js');
require('../../_chunks/dep-2bfa66aa.js');
require('../../_chunks/dep-ea9fd5a2.js');
require('../../_chunks/dep-c2cca927.js');
require('../../_chunks/dep-5e1fbdf8.js');
require('../../radio/radio.js');
require('../../radio/props.js');
require('../../radio/constants/index.js');
require('../../_chunks/dep-585d18dc.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../../_chunks/dep-8308792f.js');
require('../../_chunks/dep-d5e8d527.js');
require('../../_chunks/dep-8e4f83bf.js');
require('../../_chunks/dep-d3b20f18.js');
require('../../_chunks/dep-7d888cd2.js');
require('../../_chunks/dep-3d9f41ae.js');
require('../../_chunks/dep-117093bb.js');
require('../../radio/group.js');
require('@babel/runtime/helpers/asyncToGenerator');
require('@babel/runtime/regenerator');
require('../../radio/radio-group-props.js');
require('../../radio/radio-button.js');
require('../../radio/hooks/useKeyboard.js');
require('../../watermark/hooks/index.js');
require('../../_chunks/dep-ffe05cb1.js');
require('../../_chunks/dep-f8f462e0.js');
require('../../_chunks/dep-47af4fff.js');
require('../../_chunks/dep-a8f427e5.js');
require('../../_chunks/dep-e70e28c7.js');
require('../../_chunks/dep-880baab0.js');
require('../../input/input.js');
require('../../input/props.js');
require('../../input/hooks/useInput.js');
require('../../form/constants/index.js');
require('../../input/hooks/useLengthLimit.js');
require('../../_chunks/dep-f7deb88a.js');
require('../../input/hooks/useInputEventHandler.js');
require('../../input/hooks/useInputWidth.js');
require('../../input/input-group.js');
require('../../input/input-group-props.js');
require('../../button/button.js');
require('../../loading/index.js');
require('../../_chunks/dep-548039eb.js');
require('../../loading/plugin.js');
require('../../_chunks/dep-4a5eca86.js');
require('../../loading/icon/gradient.js');
require('../../loading/props.js');
require('../../_chunks/dep-9997364c.js');
require('../../_chunks/dep-1e2d078f.js');
require('../../_chunks/dep-ffd72e3b.js');
require('../../_chunks/dep-6ed4388e.js');
require('../../button/props.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TableFilterController = Vue.defineComponent({
  name: "TableFilterController",
  props: {
    locale: Object,
    column: Object,
    colIndex: Number,
    tFilterValue: Object,
    innerFilterValue: Object,
    tableFilterClasses: Object,
    isFocusClass: String,
    primaryTableElement: {},
    popupProps: Object,
    attach: [String, Function],
    onVisibleChange: Function,
    onInnerFilterChange: Function,
    filterIcon: [Function]
  },
  emits: ["inner-filter-change", "reset", "confirm"],
  setup: function setup(props, context) {
    var triggerElementRef = Vue.ref(null);
    var renderTNode = index.useTNodeDefault();
    var _useConfig = configProvider_hooks_useConfig.useConfig("table", props.locale),
      t = _useConfig.t,
      globalConfig = _useConfig.globalConfig;
    var _useGlobalIcon = index$1.useGlobalIcon({
        FilterIcon: tdesignIconsVueNext.FilterIcon
      }),
      FilterIcon = _useGlobalIcon.FilterIcon;
    var filterPopupVisible = Vue.ref(false);
    var onFilterPopupVisibleChange = function onFilterPopupVisibleChange(visible) {
      var _props$onVisibleChang;
      filterPopupVisible.value = visible;
      (_props$onVisibleChang = props.onVisibleChange) === null || _props$onVisibleChang === void 0 || _props$onVisibleChang.call(props, visible);
    };
    var renderComponent = function renderComponent(column, filterComponentProps, component) {
      if (!component) return null;
      var isVueComponent = !!component.setup;
      if (isFunction.isFunction(column.filter.component) && !isVueComponent) {
        return column.filter.component(function (v, b) {
          var tProps = _typeof__default["default"](b) === "object" && "attrs" in b ? b.attrs : {};
          return Vue.h(v, {
            props: _objectSpread(_objectSpread({}, filterComponentProps), tProps)
          });
        });
      }
      var filter = column.filter || {};
      return Vue.createVNode(component, Vue.mergeProps({
        "class": filter.classNames,
        "style": filter.style
      }, filter.attrs, filterComponentProps), null);
    };
    var getFilterContent = function getFilterContent(column) {
      var _column$filter, _column$filter2, _column$filter3;
      var types = ["single", "multiple", "input"];
      if ((_column$filter = column.filter) !== null && _column$filter !== void 0 && _column$filter.type && !types.includes(column.filter.type)) {
        console.error("TDesign Table Error: column.filter.type must be the following: ".concat(JSON.stringify(types)));
        return;
      }
      var _props$innerFilterVal = props.innerFilterValue,
        innerFilterValue = _props$innerFilterVal === void 0 ? {} : _props$innerFilterVal;
      var component = {
        single: radio_index.RadioGroup,
        multiple: checkbox_index.CheckboxGroup,
        input: input_index.Input
      }[column.filter.type] || column.filter.component;
      if (!component && !column.filter.component) return;
      var filterComponentProps = _objectSpread(_objectSpread({
        options: ["single", "multiple"].includes(column.filter.type) ? (_column$filter2 = column.filter) === null || _column$filter2 === void 0 ? void 0 : _column$filter2.list : void 0
      }, (_column$filter3 = column.filter) === null || _column$filter3 === void 0 ? void 0 : _column$filter3.props), {}, {
        onChange: function onChange(val, ctx) {
          var _column$filter$props, _column$filter4;
          context.emit("inner-filter-change", val, column);
          if ((_column$filter$props = column.filter.props) !== null && _column$filter$props !== void 0 && _column$filter$props.onChange) {
            var _column$filter$props$, _column$filter$props2;
            (_column$filter$props$ = (_column$filter$props2 = column.filter.props).onChange) === null || _column$filter$props$ === void 0 || _column$filter$props$.call(_column$filter$props2, val, ctx);
          }
          if ((_column$filter4 = column.filter) !== null && _column$filter4 !== void 0 && (_column$filter4 = _column$filter4.confirmEvents) !== null && _column$filter4 !== void 0 && _column$filter4.includes("onChange")) {
            filterPopupVisible.value = false;
          }
        }
      });
      if (column.colKey && innerFilterValue && column.colKey in innerFilterValue) {
        filterComponentProps.value = innerFilterValue === null || innerFilterValue === void 0 ? void 0 : innerFilterValue[column.colKey];
      }
      if (column.filter.confirmEvents) {
        column.filter.confirmEvents.forEach(function (event) {
          if (event === "onChange") return;
          filterComponentProps[event] = function () {
            context.emit("confirm", column);
            filterPopupVisible.value = false;
          };
        });
      }
      return Vue.createVNode("div", {
        "class": props.tableFilterClasses.contentInner
      }, [renderComponent(column, filterComponentProps, component)]);
    };
    var getBottomButtons = function getBottomButtons(column) {
      if (!column.filter.showConfirmAndReset) return;
      return Vue.createVNode("div", {
        "class": props.tableFilterClasses.bottomButtons
      }, [Vue.createVNode(button_index.Button, {
        "theme": "default",
        "size": "small",
        "onClick": function onClick() {
          context.emit("reset", column);
          filterPopupVisible.value = false;
        }
      }, {
        "default": function _default() {
          return [globalConfig.value.resetText];
        }
      }), Vue.createVNode(button_index.Button, {
        "theme": "primary",
        "size": "small",
        "onClick": function onClick() {
          context.emit("confirm", column);
          filterPopupVisible.value = false;
        }
      }, {
        "default": function _default() {
          return [globalConfig.value.confirmText];
        }
      })]);
    };
    var getContent = function getContent() {
      return Vue.createVNode("div", {
        "class": props.tableFilterClasses.popupContent
      }, [getFilterContent(props.column), getBottomButtons(props.column)]);
    };
    return function () {
      var _props$tFilterValue;
      if (!props.column.filter || props.column.filter && !Object.keys(props.column.filter).length) return null;
      var defaultFilterIcon = t(globalConfig.value.filterIcon) || Vue.createVNode(FilterIcon, null, null);
      var filterValue = (_props$tFilterValue = props.tFilterValue) === null || _props$tFilterValue === void 0 ? void 0 : _props$tFilterValue[props.column.colKey];
      var isObjectTrue = _typeof__default["default"](filterValue) === "object" && !isEmpty.isEmpty(filterValue);
      var isValueExist = ![null, void 0, ""].includes(filterValue) && _typeof__default["default"](filterValue) !== "object";
      return Vue.createVNode(popup_index.Popup, Vue.mergeProps({
        "attach": props.attach || (props.primaryTableElement ? function () {
          return props.primaryTableElement;
        } : void 0),
        "visible": filterPopupVisible.value,
        "destroyOnClose": true,
        "trigger": "click",
        "placement": "bottom-right",
        "showArrow": true,
        "overlayClassName": props.tableFilterClasses.popup,
        "onVisibleChange": function onVisibleChange(val) {
          return onFilterPopupVisibleChange(val);
        },
        "class": [props.tableFilterClasses.icon, _defineProperty__default["default"]({}, props.isFocusClass, isObjectTrue || isValueExist)],
        "content": getContent
      }, props.popupProps), {
        "default": function _default() {
          return [Vue.createVNode("div", {
            "ref": triggerElementRef
          }, [renderTNode("filterIcon", {
            defaultNode: defaultFilterIcon,
            params: {
              col: props.column,
              colIndex: props.colIndex
            }
          })])];
        }
      });
    };
  }
});

exports["default"] = TableFilterController;
