// tdesign-vue-next/esm/table/hooks/usePagination.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { toRefs, ref, watch, createVNode, mergeProps } from 'vue';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'lodash-es';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/typeof';
import '../../_chunks/dep-651bdc55.js';
import { useConfig } from '../../config-provider/hooks/useConfig.js';
import '@babel/runtime/helpers/slicedToArray';
import '../../_chunks/dep-204ae706.js';
import { Pagination } from '../../pagination/index.js';
import { getLocalPaginationPageData } from '../utils/index.js';
import '../../_chunks/dep-76473034.js';
import '../../config-provider/utils/context.js';
import '../../_chunks/dep-5c1cdd0b.js';
import 'dayjs';
import '@babel/runtime/helpers/createClass';
import '@babel/runtime/helpers/classCallCheck';
import '../../pagination/pagination.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import 'tdesign-icons-vue-next';
import '../../_chunks/dep-eff3e90f.js';
import '../../_chunks/dep-a7922f61.js';
import '../../_chunks/dep-69502ff8.js';
import '../../_chunks/dep-59f77e78.js';
import '../../_chunks/dep-c6879e67.js';
import '../../_chunks/dep-3d495891.js';
import '../../_chunks/dep-f32cc39c.js';
import '../../input-number/index.js';
import '../../input-number/input-number.js';
import '../../button/index.js';
import '../../button/button.js';
import '../../loading/index.js';
import '../../loading/directive.js';
import '../../loading/plugin.js';
import '../../_chunks/dep-d32d8561.js';
import '../../loading/icon/gradient.js';
import '../../_chunks/dep-d15b7e74.js';
import '../../_chunks/dep-a9edb4d8.js';
import '../../_chunks/dep-7daa5fe4.js';
import '../../loading/props.js';
import '../../_chunks/dep-5dfdb698.js';
import 'tdesign-vue-next/esm/common/style/web/components/loading/_index.less';
import '../../button/props.js';
import '../../_chunks/dep-ed0885e7.js';
import '../../_chunks/dep-16e13fa9.js';
import 'tdesign-vue-next/esm/common/style/web/components/button/_index.less';
import '../../input/index.js';
import '../../input/input.js';
import '../../input/props.js';
import '../../_chunks/dep-0d255418.js';
import '../../input/hooks/useInput.js';
import '../../form/constants/index.js';
import '../../input/hooks/useLengthLimit.js';
import '../../_chunks/dep-f2301d1c.js';
import '../../input/hooks/useInputEventHandler.js';
import '../../input/hooks/useInputWidth.js';
import '../../_chunks/dep-f5c5752f.js';
import '../../input/input-group.js';
import '../../input/input-group-props.js';
import 'tdesign-vue-next/esm/common/style/web/components/input/_index.less';
import '../../input-number/props.js';
import '../../_chunks/dep-ff412569.js';
import '../../input-number/hooks/useInputNumber.js';
import '../../_chunks/dep-b21bd6d2.js';
import '../../_chunks/dep-75be08e7.js';
import 'tdesign-vue-next/esm/common/style/web/components/input-number/_index.less';
import '../../select/index.js';
import '../../select/select.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '../../common-components/fake-arrow.js';
import '../../select-input/index.js';
import '../../select-input/select-input.js';
import '../../popup/index.js';
import '../../popup/popup.js';
import '@popperjs/core';
import '../../popup/container.js';
import '../../popup/props.js';
import 'tdesign-vue-next/esm/common/style/web/components/popup/_index.less';
import '../../select-input/props.js';
import '../../select-input/hooks/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/tag-input.js';
import '../../tag-input/props.js';
import '../../tag-input/hooks/useDragSorter.js';
import '../../tag-input/hooks/useHover.js';
import '../../tag-input/hooks/useTagScroll.js';
import '../../tag-input/hooks/useTagList.js';
import '../../tag/index.js';
import '../../tag/tag.js';
import 'tinycolor2';
import '../../tag/props.js';
import '../../tag/check-tag.js';
import '../../tag/check-tag-props.js';
import '../../_chunks/dep-8d513880.js';
import '../../tag/check-tag-group.js';
import '../../tag/check-tag-group-props.js';
import 'tdesign-vue-next/esm/common/style/web/components/tag/_index.less';
import 'tdesign-vue-next/esm/common/style/web/components/tag-input/_index.less';
import '../../select-input/hooks/useOverlayInnerStyle.js';
import '../../select-input/hooks/useSingle.js';
import 'tdesign-vue-next/esm/common/style/web/components/select-input/_index.less';
import '../../select/components/select-panel.js';
import '../../select/option.js';
import '../../select/option-props.js';
import '../../checkbox/index.js';
import '../../checkbox/checkbox.js';
import '../../checkbox/props.js';
import '../../checkbox/constants/index.js';
import '../../checkbox/hooks/useCheckboxLazyLoad.js';
import '../../_chunks/dep-2cabe3fe.js';
import '../../checkbox/hooks/useKeyboardEvent.js';
import '../../checkbox/group.js';
import '../../checkbox/checkbox-group-props.js';
import '../../_chunks/dep-d4579130.js';
import 'tdesign-vue-next/esm/common/style/web/components/checkbox/_index.less';
import '../../select/utils/index.js';
import '../../select/constants/index.js';
import '../../select/option-group.js';
import '../../select/option-group-props.js';
import '../../select/props.js';
import '../../select/hooks/usePanelVirtualScroll.js';
import '../../select/hooks/useKeyboardControl.js';
import '../../select/hooks/useSelectOptions.js';
import 'tdesign-vue-next/esm/common/style/web/components/select/_index.less';
import '../../input-adornment/index.js';
import '../../input-adornment/input-adornment.js';
import '../../input-adornment/props.js';
import 'tdesign-vue-next/esm/common/style/web/components/input-adornment/_index.less';
import '../../pagination/props.js';
import '../../pagination/hooks/useMoreAction.js';
import '../../pagination/hooks/usePaginationClasses.js';
import '../../pagination/pagination-mini.js';
import '../../pagination/pagination-mini-props.js';
import '../../tooltip/index.js';
import '../../tooltip/tooltip.js';
import '../../tooltip/props.js';
import '../../tooltip/utils/index.js';
import 'tdesign-vue-next/esm/common/style/web/components/tooltip/_index.less';
import 'tdesign-vue-next/esm/common/style/web/components/pagination/_index.less';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function usePagination(props, context, tableContentRef) {
  var _toRefs = toRefs(props),
    pagination = _toRefs.pagination,
    data = _toRefs.data,
    disableDataPage = _toRefs.disableDataPage;
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var innerPagination = ref(props.pagination);
  var dataSource = ref([]);
  var isPaginateData = ref(false);
  var updateDataSourceAndPaginate = function updateDataSourceAndPaginate() {
    var current = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    var pageSize = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 10;
    var data2 = props.data;
    var t = Boolean(!disableDataPage.value && data2.length > pageSize);
    isPaginateData.value = t;
    dataSource.value = getLocalPaginationPageData(data2, {
      current: current,
      pageSize: pageSize
    }, disableDataPage.value);
  };
  watch([data, function () {
    return data.value.length;
  }, function () {
    var _pagination$value;
    return (_pagination$value = pagination.value) === null || _pagination$value === void 0 ? void 0 : _pagination$value.current;
  }, function () {
    var _pagination$value2;
    return (_pagination$value2 = pagination.value) === null || _pagination$value2 === void 0 ? void 0 : _pagination$value2.pageSize;
  }, disableDataPage], function () {
    if (!pagination.value || !pagination.value.current) return;
    var _pagination$value3 = pagination.value,
      current = _pagination$value3.current,
      pageSize = _pagination$value3.pageSize;
    var newPageInfo = {
      current: current,
      pageSize: pageSize
    };
    innerPagination.value = newPageInfo;
    updateDataSourceAndPaginate(pagination.value.current, pagination.value.pageSize || pagination.value.defaultPageSize);
  }, {
    immediate: true
  });
  watch([data], function () {
    var _innerPagination$valu, _innerPagination$valu2;
    if (!pagination.value || !pagination.value.defaultCurrent) return;
    var isControlled = Boolean(pagination.value.current);
    if (isControlled) return;
    updateDataSourceAndPaginate((_innerPagination$valu = innerPagination.value.current) !== null && _innerPagination$valu !== void 0 ? _innerPagination$valu : pagination.value.defaultCurrent, (_innerPagination$valu2 = innerPagination.value.pageSize) !== null && _innerPagination$valu2 !== void 0 ? _innerPagination$valu2 : pagination.value.defaultPageSize);
  }, {
    immediate: true
  });
  var renderPagination = function renderPagination() {
    if (!props.pagination) return null;
    var paginationProps = _objectSpread({}, props.pagination);
    delete paginationProps.onChange;
    return createVNode("div", {
      "class": "".concat(classPrefix.value, "-table__pagination")
    }, [createVNode(Pagination, mergeProps({
      "size": props.size === "large" ? null : props.size
    }, paginationProps, {
      "onChange": function onChange(pageInfo) {
        var _props$pagination, _props$pagination$onC, _props$onPageChange;
        (_props$pagination = props.pagination) === null || _props$pagination === void 0 || (_props$pagination$onC = _props$pagination.onChange) === null || _props$pagination$onC === void 0 || _props$pagination$onC.call(_props$pagination, pageInfo);
        innerPagination.value = pageInfo;
        updateDataSourceAndPaginate(pageInfo.current, pageInfo.pageSize);
        (_props$onPageChange = props.onPageChange) === null || _props$onPageChange === void 0 || _props$onPageChange.call(props, pageInfo, dataSource.value);
        var ref2 = tableContentRef.value;
        if (ref2.scrollTo) {
          ref2.scrollTo({
            top: 0,
            left: 0
          });
        } else {
          ref2.scrollTop = 0;
          ref2.scrollLeft = 0;
        }
      }
    }), {
      totalContent: context.slots.totalContent
    })]);
  };
  return {
    isPaginateData: isPaginateData,
    dataSource: dataSource,
    innerPagination: innerPagination,
    renderPagination: renderPagination
  };
}

export { usePagination as default };
