// tdesign-vue-next/es/date-picker/components/panel/PanelContent.mjs
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-edde6136.mjs';
import { defineComponent, createVNode } from 'vue';
import { isFunction, isArray } from 'lodash-es';
import { u as usePrefixClass } from '../../../_chunks/dep-d6195387.mjs';
import TDateHeader from '../base/Header.mjs';
import TDateTable from '../base/Table.mjs';
import { TimePickerPanel } from '../../../time-picker/index.mjs';
import { g as getDefaultFormat } from '../../../_chunks/dep-53e8af8e.mjs';
import { parseToDateTime } from '../../utils/index.mjs';
import '../../../_chunks/dep-cb344d2d.mjs';
import '../../../_chunks/dep-fe2215b7.mjs';
import '../../../_chunks/dep-c13a05a2.mjs';
import '../../../_chunks/dep-478d8d87.mjs';
import '../../../_chunks/dep-80ab80bc.mjs';
import '../../../config-provider/hooks/useConfig.mjs';
import '../../../config-provider/utils/context.mjs';
import '../../../_chunks/dep-fa2e9d13.mjs';
import '../../../_chunks/dep-c22f10cd.mjs';
import '../../../_chunks/dep-2512dba6.mjs';
import '../../../_chunks/dep-39047c40.mjs';
import '../../../_chunks/dep-1d48345c.mjs';
import '../../../_chunks/dep-4f3de704.mjs';
import '../../../_chunks/dep-15b76bbf.mjs';
import '../../../_chunks/dep-61161446.mjs';
import '../../../_chunks/dep-671c8553.mjs';
import '../../../pagination/index.mjs';
import '../../../pagination/pagination.mjs';
import '../../../_chunks/dep-15e8aadf.mjs';
import 'tdesign-icons-vue-next';
import '../../../input-number/index.mjs';
import '../../../input-number/input-number.mjs';
import '../../../button/index.mjs';
import '../../../button/button.mjs';
import '../../../loading/index.mjs';
import '../../../loading/directive.mjs';
import '../../../loading/plugin.mjs';
import '../../../loading/loading.mjs';
import '../../../loading/icon/gradient.mjs';
import '../../../_chunks/dep-58da668e.mjs';
import '../../../loading/props.mjs';
import './style/css.mjs';
import '../../../button/props.mjs';
import '../../../input/index.mjs';
import '../../../input/input.mjs';
import '../../../input/props.mjs';
import '../../../input/hooks/index.mjs';
import '../../../input/hooks/useInput.mjs';
import '../../../form/constants/index.mjs';
import '../../../input/hooks/useLengthLimit.mjs';
import '../../../_chunks/dep-c7fa2bca.mjs';
import '../../../input/hooks/useInputEventHandler.mjs';
import '../../../input/hooks/useInputWidth.mjs';
import '../../../input/input-group.mjs';
import '../../../input/input-group-props.mjs';
import '../../../input-number/props.mjs';
import '../../../input-number/hooks/useInputNumber.mjs';
import '../../../_chunks/dep-71bbcf0c.mjs';
import '../../../_chunks/dep-f3ae51ae.mjs';
import '../../../select/index.mjs';
import '../../../select/select.mjs';
import '../../../_chunks/dep-bcb347a6.mjs';
import '../../../common-components/fake-arrow.mjs';
import '../../../select-input/index.mjs';
import '../../../select-input/select-input.mjs';
import '../../../popup/index.mjs';
import '../../../popup/popup.mjs';
import '@popperjs/core';
import '../../../popup/container.mjs';
import '../../../popup/props.mjs';
import '../../../select-input/props.mjs';
import '../../../select-input/hooks/index.mjs';
import '../../../select-input/hooks/useMultiple.mjs';
import '../../../tag-input/index.mjs';
import '../../../tag-input/tag-input.mjs';
import '../../../tag-input/props.mjs';
import '../../../tag-input/hooks/index.mjs';
import '../../../tag-input/hooks/useDragSorter.mjs';
import '../../../tag-input/hooks/useHover.mjs';
import '../../../tag-input/hooks/useTagScroll.mjs';
import '../../../tag-input/hooks/useTagList.mjs';
import '../../../tag/index.mjs';
import '../../../tag/tag.mjs';
import '../../../_chunks/dep-eaa90005.mjs';
import '../../../tag/props.mjs';
import '../../../tag/check-tag.mjs';
import '../../../tag/check-tag-props.mjs';
import '../../../tag/check-tag-group.mjs';
import '../../../tag/check-tag-group-props.mjs';
import '../../../select-input/hooks/useOverlayInnerStyle.mjs';
import '../../../select-input/hooks/useSingle.mjs';
import '../../../select/components/select-panel.mjs';
import '../../../select/option.mjs';
import '../../../select/option-props.mjs';
import '../../../checkbox/index.mjs';
import '../../../checkbox/checkbox.mjs';
import '../../../checkbox/props.mjs';
import '../../../checkbox/constants/index.mjs';
import '../../../checkbox/hooks/useCheckboxLazyLoad.mjs';
import '../../../checkbox/hooks/useKeyboardEvent.mjs';
import '../../../checkbox/group.mjs';
import '../../../checkbox/checkbox-group-props.mjs';
import '../../../select/utils/index.mjs';
import '../../../select/constants/index.mjs';
import '../../../select/option-group.mjs';
import '../../../select/option-group-props.mjs';
import '../../../select/props.mjs';
import '../../../select/hooks/index.mjs';
import '../../../select/hooks/useKeyboardControl.mjs';
import '../../../select/hooks/usePanelVirtualScroll.mjs';
import '../../../select/hooks/useSelectOptions.mjs';
import '../../../input-adornment/index.mjs';
import '../../../input-adornment/input-adornment.mjs';
import '../../../input-adornment/props.mjs';
import '../../../pagination/props.mjs';
import '../../../pagination/hooks/index.mjs';
import '../../../pagination/hooks/useMoreAction.mjs';
import '../../../pagination/hooks/usePaginationClasses.mjs';
import '../../../pagination/pagination-mini.mjs';
import '../../../pagination/pagination-mini-props.mjs';
import '../../../tooltip/index.mjs';
import '../../../tooltip/tooltip.mjs';
import '../../../tooltip/props.mjs';
import '../../../tooltip/utils/index.mjs';
import '../../hooks/index.mjs';
import '../../hooks/useDisableDate.mjs';
import '../../../_chunks/dep-66bd7218.mjs';
import '../../../_chunks/dep-25210ce3.mjs';
import '../../../_chunks/dep-65ec4207.mjs';
import '../../hooks/useRange.mjs';
import '../../hooks/useRangeValue.mjs';
import '../../hooks/useSingle.mjs';
import '../../hooks/useSingleValue.mjs';
import '../../hooks/useTableData.mjs';
import '../../hooks/useSelectRange.mjs';
import '../base/Cell.mjs';
import '../../../time-picker/time-picker.mjs';
import '../../../time-picker/panel/time-picker-panel.mjs';
import '../../../_chunks/dep-951ab8eb.mjs';
import '../../../time-picker/props.mjs';
import '../../../_chunks/dep-613ac2d6.mjs';
import '../../../time-picker/time-range-picker.mjs';
import '../../../range-input/index.mjs';
import '../../../range-input/range-input.mjs';
import '../../../range-input/props.mjs';
import '../../../range-input/range-input-popup.mjs';
import '../../../range-input/range-input-popup-props.mjs';
import '../../../time-picker/time-range-picker-props.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TPanelContent = defineComponent({
  name: "TPanelContent",
  props: {
    mode: String,
    format: String,
    enableTimePicker: Boolean,
    timePickerProps: {
      type: Object,
      "default": function _default() {
        return {};
      }
    },
    year: Number,
    month: Number,
    range: [Array, Function],
    tableData: Array,
    time: String,
    multiple: Boolean,
    firstDayOfWeek: Number,
    partial: String,
    popupVisible: Boolean,
    onYearChange: Function,
    onMonthChange: Function,
    onJumperClick: Function,
    onCellMouseEnter: Function,
    onCellClick: Function,
    onCellMouseLeave: Function,
    onTimePickerChange: Function,
    value: [String, Number, Array, Date],
    internalYear: Array,
    disableTime: Function,
    defaultTime: [String, Array],
    cell: {
      type: Function
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots;
    var COMPONENT_NAME = usePrefixClass("date-picker__panel");
    var _getDefaultFormat = getDefaultFormat({
        mode: props.mode,
        format: props.format,
        enableTimePicker: props.enableTimePicker
      }),
      timeFormat = _getDefaultFormat.timeFormat;
    var disableTimeOptions = function disableTimeOptions() {
      if (!isFunction(props.disableTime)) {
        return {};
      }
      var startValue = isArray(props.value) ? props.value[0] : props.value;
      var endValue = isArray(props.value) ? props.value[1] : props.value;
      var isRangePicker = props.partial;
      if (!isRangePicker) {
        return props.disableTime(parseToDateTime(startValue, props.format));
      }
      return props.disableTime([parseToDateTime(startValue, props.format), parseToDateTime(endValue, props.format)], {
        partial: props.partial
      });
    };
    var defaultTimeValue = "00:00:00";
    return function () {
      return createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-content")
      }, [createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-").concat(props.mode)
      }, [createVNode(TDateHeader, {
        "mode": props.mode,
        "year": props.year,
        "month": props.month,
        "range": props.range,
        "internalYear": props.internalYear,
        "partial": props.partial,
        "onMonthChange": function onMonthChange(val) {
          var _props$onMonthChange;
          return (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 ? void 0 : _props$onMonthChange.call(props, val, {
            partial: props.partial
          });
        },
        "onYearChange": function onYearChange(val) {
          var _props$onYearChange;
          return (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 ? void 0 : _props$onYearChange.call(props, val, {
            partial: props.partial
          });
        },
        "onJumperClick": function onJumperClick(_ref2) {
          var _props$onJumperClick;
          var trigger = _ref2.trigger;
          return (_props$onJumperClick = props.onJumperClick) === null || _props$onJumperClick === void 0 ? void 0 : _props$onJumperClick.call(props, {
            trigger: trigger,
            partial: props.partial
          });
        }
      }, null), createVNode(TDateTable, {
        "mode": props.mode,
        "data": props.tableData,
        "time": props.time,
        "value": props.value,
        "format": props.format,
        "multiple": props.multiple,
        "firstDayOfWeek": props.firstDayOfWeek,
        "onCellClick": function onCellClick(date, _ref3) {
          var _props$onCellClick;
          var e = _ref3.e;
          return (_props$onCellClick = props.onCellClick) === null || _props$onCellClick === void 0 ? void 0 : _props$onCellClick.call(props, date, {
            e: e,
            partial: props.partial
          });
        },
        "onCellMouseEnter": function onCellMouseEnter(date) {
          var _props$onCellMouseEnt;
          return (_props$onCellMouseEnt = props.onCellMouseEnter) === null || _props$onCellMouseEnt === void 0 ? void 0 : _props$onCellMouseEnt.call(props, date, {
            partial: props.partial
          });
        },
        "onCellMouseLeave": props.onCellMouseLeave,
        "cell": props.cell
      }, slots)]), props.enableTimePicker && createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-time")
      }, [createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-time-viewer")
      }, [props.time || defaultTimeValue]), createVNode(TimePickerPanel, _objectSpread({
        key: props.partial,
        isShowPanel: props.popupVisible,
        format: timeFormat,
        value: props.time || defaultTimeValue,
        onChange: props.onTimePickerChange,
        disableTime: disableTimeOptions
      }, props.timePickerProps), null)])]);
    };
  }
});

export { TPanelContent as default };
