// tdesign-vue-next/cjs/time-picker/panel/time-picker-panel.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var dayjs = require('dayjs');
var customParseFormat = require('../../_chunks/dep-a7068a83.js');
var timePicker_panel_props = require('../../_chunks/dep-83ab8a79.js');
var timePicker_panel_singlePanel = require('../../_chunks/dep-5fee93c0.js');
var button_button = require('../../button/button.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('../../_chunks/dep-c7472ccd.js');
var index = require('../../_chunks/dep-b56e7ff5.js');
require('@babel/runtime/helpers/slicedToArray');
var log = require('../../_chunks/dep-f7deb88a.js');
require('../../_chunks/dep-d1a4892a.js');
var isUndefined = require('../../_chunks/dep-5e1fbdf8.js');
var configProvider_hooks_useConfig = require('../../config-provider/hooks/useConfig.js');
require('../../_chunks/dep-bf2e7501.js');
require('../props.js');
require('../../_chunks/dep-850526b5.js');
require('../../_chunks/dep-e2205a6f.js');
require('../../_chunks/dep-6f4a5c31.js');
require('../../_chunks/dep-e01a4fce.js');
require('../../_chunks/dep-6c498eb2.js');
require('../../_chunks/dep-cde052f0.js');
require('../../_chunks/dep-c644d54f.js');
require('../../_chunks/dep-5729cb14.js');
require('../../_chunks/dep-07f46433.js');
require('../../_chunks/dep-0b08f67c.js');
require('../../_chunks/dep-eb1f700b.js');
require('../../_chunks/dep-9399a4e2.js');
require('../../_chunks/dep-58ff00b9.js');
require('../../_chunks/dep-37af9215.js');
require('../../_chunks/dep-0c9948d8.js');
require('../../_chunks/dep-2e9fed7b.js');
require('../../_chunks/dep-d1d374b4.js');
require('../../_chunks/dep-a324ad6a.js');
require('../../_chunks/dep-06e53606.js');
require('../../_chunks/dep-d8ab5fb5.js');
require('../../_chunks/dep-2333b13f.js');
require('../../_chunks/dep-89d65b78.js');
require('../../_chunks/dep-dc9d7b38.js');
require('../../_chunks/dep-6ed4388e.js');
require('../../loading/index.js');
require('../../_chunks/dep-548039eb.js');
require('../../loading/plugin.js');
require('../../_chunks/dep-4a5eca86.js');
require('../../loading/icon/gradient.js');
require('../../_chunks/dep-fd403354.js');
require('../../_chunks/dep-585d18dc.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../../_chunks/dep-8308792f.js');
require('../../_chunks/dep-d5e8d527.js');
require('../../_chunks/dep-bde8abb8.js');
require('../../_chunks/dep-8e4f83bf.js');
require('../../_chunks/dep-eb214be8.js');
require('../../_chunks/dep-11e6da14.js');
require('../../_chunks/dep-0e56b1da.js');
require('../../_chunks/dep-c8d40779.js');
require('../../_chunks/dep-388cb46f.js');
require('../../_chunks/dep-445d3f65.js');
require('../../_chunks/dep-09b19404.js');
require('../../loading/props.js');
require('../../_chunks/dep-41cc84a3.js');
require('../../_chunks/dep-c51f4345.js');
require('../../_chunks/dep-e2c29461.js');
require('../../_chunks/dep-52d3718c.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../../_chunks/dep-f8f462e0.js');
require('../../_chunks/dep-47af4fff.js');
require('../../_chunks/dep-ea9fd5a2.js');
require('../../_chunks/dep-a8f427e5.js');
require('../../_chunks/dep-9997364c.js');
require('../../_chunks/dep-1e2d078f.js');
require('../../_chunks/dep-7d888cd2.js');
require('../../_chunks/dep-ffd72e3b.js');
require('../../_chunks/dep-742210e1.js');
require('../../button/props.js');
require('../../_chunks/dep-a8de5365.js');
require('../../_chunks/dep-a90757f2.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
dayjs__default["default"].extend(customParseFormat.customParseFormat);
var _TimePickerPanel = Vue.defineComponent({
  name: "TTimePickerPanel",
  props: _objectSpread(_objectSpread({}, timePicker_panel_props.panelProps()), {}, {
    isFooterDisplay: Boolean,
    handleConfirmClick: Function,
    onChange: Function,
    position: String
  }),
  setup: function setup(props) {
    var _useConfig = configProvider_hooks_useConfig.useConfig("timePicker"),
      globalConfig = _useConfig.globalConfig;
    var COMPONENT_NAME = index.usePrefixClass("time-picker__panel");
    var _toRefs = Vue.toRefs(props),
      steps = _toRefs.steps,
      isFooterDisplay = _toRefs.isFooterDisplay,
      isShowPanel = _toRefs.isShowPanel;
    var triggerScroll = Vue.ref(false);
    var panelRef = Vue.ref();
    var showNowTimeBtn = Vue.computed(function () {
      return !!steps.value.filter(function (step) {
        return Number(step) > 1;
      }).length;
    });
    var defaultValue = Vue.computed(function () {
      var isStepsSet = showNowTimeBtn.value;
      var formattedValue = dayjs__default["default"](props.value, props.format);
      if (props.value && formattedValue.isValid()) {
        return formattedValue.format(props.format);
      }
      if (isStepsSet) {
        return dayjs__default["default"]().hour(0).minute(0).second(0).format(props.format);
      }
      return dayjs__default["default"]().hour(0).minute(0).second(0).format(props.format);
    });
    var panelColUpdate = function panelColUpdate() {
      Vue.nextTick(function () {
        triggerScroll.value = true;
      });
    };
    var resetTriggerScroll = function resetTriggerScroll() {
      triggerScroll.value = false;
    };
    var handlePresetClick = function handlePresetClick(presetValue) {
      var presetVal = typeof presetValue === "function" ? presetValue() : presetValue;
      if (typeof props.activeIndex === "number") {
        if (Array.isArray(presetVal)) {
          var _props$onChange;
          (_props$onChange = props.onChange) === null || _props$onChange === void 0 || _props$onChange.call(props, presetVal[props.activeIndex]);
        } else {
          log.log.error("TimePicker", "preset: ".concat(props.presets, " \u9884\u8BBE\u503C\u5FC5\u987B\u662F\u6570\u7EC4!"));
        }
      } else {
        var _props$onChange2;
        (_props$onChange2 = props.onChange) === null || _props$onChange2 === void 0 || _props$onChange2.call(props, presetVal);
      }
    };
    var renderFooter = function renderFooter() {
      if (!isUndefined.isUndefined(props.presets)) return Object.keys(props.presets || []).map(function (key) {
        return Vue.createVNode(button_button["default"], {
          "key": key,
          "theme": "primary",
          "size": "small",
          "variant": "text",
          "onClick": function onClick() {
            return handlePresetClick(props.presets[key]);
          }
        }, _isSlot(key) ? key : {
          "default": function _default() {
            return [key];
          }
        });
      });
      return !showNowTimeBtn.value ? Vue.createVNode(button_button["default"], {
        "theme": "primary",
        "variant": "text",
        "size": "small",
        "onClick": function onClick() {
          var _props$onChange3;
          return (_props$onChange3 = props.onChange) === null || _props$onChange3 === void 0 ? void 0 : _props$onChange3.call(props, dayjs__default["default"]().format(props.format));
        }
      }, {
        "default": function _default() {
          return [globalConfig.value.now];
        }
      }) : null;
    };
    Vue.onMounted(function () {
      panelColUpdate();
    });
    Vue.watch(function () {
      return isShowPanel.value;
    }, function () {
      panelColUpdate();
    });
    return function () {
      return Vue.createVNode("div", {
        "class": COMPONENT_NAME.value
      }, [Vue.createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-section-body")
      }, [Vue.createVNode(timePicker_panel_singlePanel.SinglePanel, Vue.mergeProps(props, {
        "ref": panelRef,
        "format": props.format || timePicker_panel_props.DEFAULT_FORMAT,
        "steps": props.steps || timePicker_panel_props.DEFAULT_STEPS,
        "value": dayjs__default["default"](props.value, props.format).isValid() ? props.value : defaultValue.value,
        "triggerScroll": triggerScroll.value,
        "onChange": props.onChange,
        "resetTriggerScroll": resetTriggerScroll,
        "isShowPanel": props.isShowPanel
      }), null)]), isFooterDisplay.value ? Vue.createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-section-footer")
      }, [Vue.createVNode(button_button["default"], {
        "theme": "primary",
        "variant": "base",
        "disabled": !props.value,
        "onClick": function onClick(e) {
          var _props$handleConfirmC;
          return (_props$handleConfirmC = props.handleConfirmClick) === null || _props$handleConfirmC === void 0 ? void 0 : _props$handleConfirmC.call(props, e, defaultValue.value);
        },
        "size": "small"
      }, {
        "default": function _default() {
          return [globalConfig.value.confirm];
        }
      }), renderFooter()]) : null]);
    };
  }
});

exports["default"] = _TimePickerPanel;
