// tdesign-vue-next/lib/_chunks/dep-69502ff8.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { getCurrentInstance, isVNode, Fragment } from 'vue';
import { isObject, isString, isArray } from 'lodash-es';

function getDefaultNode(options) {
  var defaultNode;
  if (isObject(options) && "defaultNode" in options) {
    defaultNode = options.defaultNode;
  } else if (isVNode(options) || isString(options)) {
    defaultNode = options;
  }
  return defaultNode;
}
function getChildren(content) {
  var childList = [];
  var _innerGetChildren = function innerGetChildren(content2) {
    if (!isArray(content2)) return;
    content2.forEach(function (item) {
      if (item.children && isArray(item.children)) {
        if (item.type !== Fragment) return;
        _innerGetChildren(item.children);
      } else {
        childList.push(item);
      }
    });
    return childList;
  };
  return _innerGetChildren(content);
}
function getParams(options) {
  return isObject(options) && "params" in options ? options.params : {};
}
function getSlotFirst(options) {
  return isObject(options) && "slotFirst" in options ? options.slotFirst : false;
}
function isPropsUsed(propsName) {
  var _getCurrentInstance = getCurrentInstance(),
    vnode = _getCurrentInstance.vnode,
    slots = _getCurrentInstance.slots;
  var vProps = vnode.props || {};
  if (Object.prototype.hasOwnProperty.call(vProps, propsName)) return true;
  if (slots && Object.prototype.hasOwnProperty.call(slots, propsName)) return true;
  return false;
}

export { getParams as a, getDefaultNode as b, getSlotFirst as c, getChildren as g, isPropsUsed as i };
