// tdesign-vue-next/es/tree/hooks/useTreeState.mjs
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { useVModel } from '../utils/adapt.mjs';
import { ref, toRefs } from 'vue';
import '../../_chunks/dep-d6195387.mjs';
import 'lodash-es';
import '../../_chunks/dep-fe2215b7.mjs';
import '../../_chunks/dep-c13a05a2.mjs';
import '../../_chunks/dep-cb344d2d.mjs';
import '../../_chunks/dep-478d8d87.mjs';
import '../../_chunks/dep-80ab80bc.mjs';
import '../../config-provider/hooks/useConfig.mjs';
import '../../config-provider/utils/context.mjs';
import '../../_chunks/dep-fa2e9d13.mjs';
import '../../_chunks/dep-c22f10cd.mjs';
import '../../_chunks/dep-2512dba6.mjs';
import '../../_chunks/dep-39047c40.mjs';
import '../../_chunks/dep-1d48345c.mjs';
import '../../_chunks/dep-4f3de704.mjs';
import '../../_chunks/dep-15b76bbf.mjs';
import '../../_chunks/dep-61161446.mjs';
import '../../_chunks/dep-edde6136.mjs';
import '../../_chunks/dep-671c8553.mjs';
import 'tdesign-icons-vue-next';
import '../../checkbox/index.mjs';
import '../../checkbox/checkbox.mjs';
import '../../checkbox/props.mjs';
import '../../checkbox/constants/index.mjs';
import '../../checkbox/hooks/useCheckboxLazyLoad.mjs';
import '../../checkbox/hooks/useKeyboardEvent.mjs';
import '../../checkbox/group.mjs';
import '../../checkbox/checkbox-group-props.mjs';
import './style/css.mjs';
import '../../loading/index.mjs';
import '../../loading/directive.mjs';
import '../../loading/plugin.mjs';
import '../../loading/loading.mjs';
import '../../loading/icon/gradient.mjs';
import '../../_chunks/dep-58da668e.mjs';
import '../../_chunks/dep-15e8aadf.mjs';
import '../../loading/props.mjs';
import '../../_chunks/dep-3ed33ceb.mjs';
import '../../_chunks/dep-bcb347a6.mjs';
import '../../_chunks/dep-c7fa2bca.mjs';

function useTreeState(props, context) {
  var treeContentRef = ref();
  var nodes = ref([]);
  var allNodes = ref([]);
  var isScrolling = ref(false);
  var refProps = toRefs(props);
  var vmValue = useVModel(props, refProps, "value", "defaultValue", "onChange", "change");
  var vmActived = useVModel(props, refProps, "actived", "defaultActived", "onActive", "active");
  var vmExpanded = useVModel(props, refProps, "expanded", "defaultExpanded", "onExpand", "expand");
  function setStore(store) {
    state.store = store;
    state.scope.store = store;
    allNodes.value = store.getNodes();
  }
  var state = {
    props: props,
    context: context,
    store: null,
    treeContentRef: treeContentRef,
    nodes: nodes,
    allNodes: allNodes,
    isScrolling: isScrolling,
    mouseEvent: null,
    virtualConfig: null,
    scope: {
      store: null,
      treeContentRef: treeContentRef,
      treeProps: props,
      scopedSlots: {},
      virtualConfig: null
    },
    setStore: setStore,
    refProps: refProps,
    vmValue: vmValue,
    vmActived: vmActived,
    vmExpanded: vmExpanded
  };
  return {
    state: state
  };
}

export { useTreeState as default };
