// tdesign-vue-next/cjs/transfer/transfer.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var transfer_components_transferList = require('./components/transfer-list.js');
var transfer_components_transferOperations = require('./components/transfer-operations.js');
var transfer_utils_index = require('./utils/index.js');
var transfer_props = require('./props.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-c7472ccd.js');
var index$1 = require('../_chunks/dep-b56e7ff5.js');
var index$3 = require('../_chunks/dep-86694091.js');
var index = require('../_chunks/dep-a8de5365.js');
require('../_chunks/dep-d1a4892a.js');
var index$2 = require('../_chunks/dep-d149369c.js');
var isFunction = require('../_chunks/dep-37af9215.js');
var pick = require('../_chunks/dep-a66df673.js');
var isEqual = require('../_chunks/dep-f8f462e0.js');
require('../pagination/index.js');
require('../pagination/pagination.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('tdesign-icons-vue-next');
require('../_chunks/dep-0e56b1da.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-e2205a6f.js');
require('../_chunks/dep-d1d374b4.js');
require('../_chunks/dep-6f4a5c31.js');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../_chunks/dep-125f71ff.js');
require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('../_chunks/dep-bf2e7501.js');
require('dayjs');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-2e9fed7b.js');
require('../_chunks/dep-a324ad6a.js');
require('../_chunks/dep-06e53606.js');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('../input-number/index.js');
require('../input-number/input-number.js');
require('../button/index.js');
require('../button/button.js');
require('../loading/index.js');
require('../_chunks/dep-548039eb.js');
require('../loading/plugin.js');
require('../_chunks/dep-4a5eca86.js');
require('../loading/icon/gradient.js');
require('../_chunks/dep-fd403354.js');
require('../_chunks/dep-585d18dc.js');
require('../_chunks/dep-8308792f.js');
require('../_chunks/dep-d5e8d527.js');
require('../_chunks/dep-8e4f83bf.js');
require('../_chunks/dep-6c498eb2.js');
require('../loading/props.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../_chunks/dep-9997364c.js');
require('../_chunks/dep-47af4fff.js');
require('../_chunks/dep-ea9fd5a2.js');
require('../_chunks/dep-a8f427e5.js');
require('../_chunks/dep-1e2d078f.js');
require('../_chunks/dep-7d888cd2.js');
require('../_chunks/dep-ffd72e3b.js');
require('../_chunks/dep-6ed4388e.js');
require('../_chunks/dep-742210e1.js');
require('../button/props.js');
require('../_chunks/dep-a90757f2.js');
require('../_chunks/dep-850526b5.js');
require('../input/index.js');
require('../input/input.js');
require('../input/props.js');
require('../_chunks/dep-4725a846.js');
require('../input/hooks/useInput.js');
require('../form/constants/index.js');
require('../input/hooks/useLengthLimit.js');
require('../_chunks/dep-f7deb88a.js');
require('../_chunks/dep-880baab0.js');
require('../input/hooks/useInputEventHandler.js');
require('../input/hooks/useInputWidth.js');
require('../_chunks/dep-6344c74f.js');
require('../input/input-group.js');
require('../input/input-group-props.js');
require('../input-number/props.js');
require('../_chunks/dep-6f95225b.js');
require('../input-number/hooks/useInputNumber.js');
require('../_chunks/dep-3688812b.js');
require('../_chunks/dep-77efa7eb.js');
require('../_chunks/dep-cf727d09.js');
require('../select/index.js');
require('../select/select.js');
require('@babel/runtime/helpers/asyncToGenerator');
require('@babel/runtime/regenerator');
require('../common-components/fake-arrow.js');
require('../select-input/index.js');
require('../select-input/select-input.js');
require('../popup/index.js');
require('../popup/popup.js');
require('@popperjs/core');
require('../popup/container.js');
require('../popup/props.js');
require('../_chunks/dep-d8ab5fb5.js');
require('../_chunks/dep-cde052f0.js');
require('../_chunks/dep-a3501e24.js');
require('../select-input/props.js');
require('../select-input/hooks/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/tag-input.js');
require('../tag-input/props.js');
require('../tag-input/hooks/useDragSorter.js');
require('../tag-input/hooks/useHover.js');
require('../tag-input/hooks/useTagScroll.js');
require('../tag-input/hooks/useTagList.js');
require('../tag/index.js');
require('../tag/tag.js');
require('tinycolor2');
require('../tag/props.js');
require('../tag/check-tag.js');
require('../tag/check-tag-props.js');
require('../_chunks/dep-1177d395.js');
require('../tag/check-tag-group.js');
require('../tag/check-tag-group-props.js');
require('../select-input/hooks/useOverlayInnerStyle.js');
require('../select-input/hooks/useSingle.js');
require('../_chunks/dep-f3b67279.js');
require('../_chunks/dep-117093bb.js');
require('../select/components/select-panel.js');
require('../select/option.js');
require('../select/option-props.js');
require('../checkbox/index.js');
require('../checkbox/checkbox.js');
require('../checkbox/props.js');
require('../checkbox/constants/index.js');
require('../checkbox/hooks/useCheckboxLazyLoad.js');
require('../_chunks/dep-b4842612.js');
require('../checkbox/hooks/useKeyboardEvent.js');
require('../checkbox/group.js');
require('../checkbox/checkbox-group-props.js');
require('../_chunks/dep-e53cd0a5.js');
require('../_chunks/dep-2bfa66aa.js');
require('../_chunks/dep-c2cca927.js');
require('../_chunks/dep-5e1fbdf8.js');
require('../select/utils/index.js');
require('../select/constants/index.js');
require('../select/option-group.js');
require('../select/option-group-props.js');
require('../select/props.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../_chunks/dep-d3b20f18.js');
require('../_chunks/dep-3d9f41ae.js');
require('../select/hooks/useKeyboardControl.js');
require('../select/hooks/useSelectOptions.js');
require('../_chunks/dep-5c150838.js');
require('../input-adornment/index.js');
require('../input-adornment/input-adornment.js');
require('../input-adornment/props.js');
require('../pagination/props.js');
require('../pagination/hooks/useMoreAction.js');
require('../pagination/hooks/usePaginationClasses.js');
require('../pagination/pagination-mini.js');
require('../pagination/pagination-mini-props.js');
require('../tooltip/index.js');
require('../tooltip/tooltip.js');
require('../tooltip/props.js');
require('../tooltip/utils/index.js');
require('./components/transfer-search.js');
require('./hooks/useDragSort.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
var _Transfer = Vue.defineComponent({
  name: transfer_utils_index.TRANSFER_NAME,
  props: transfer_props["default"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var disabled = index.useDisabled();
    var classPrefix = index$1.usePrefixClass();
    var _toRefs = Vue.toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue,
      checked = _toRefs.checked;
    var _useVModel = index$2.useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      innerValue = _useVModel2[0],
      setInnerValue = _useVModel2[1];
    var _useDefaultValue = index$3.useDefaultValue(checked, props2.defaultChecked, props2.onCheckedChange, "checked"),
      _useDefaultValue2 = _slicedToArray__default["default"](_useDefaultValue, 1),
      innerChecked = _useDefaultValue2[0];
    var valueList = Vue.computed(function () {
      return innerValue.value;
    });
    var isTreeMode = Vue.computed(function () {
      var treeSlot = slots.tree;
      return isFunction.isFunction(treeSlot);
    });
    var leftButtonDisabled = Vue.computed(function () {
      return props2.direction === "right";
    });
    var rightButtonDisabled = Vue.computed(function () {
      return props2.direction === "left";
    });
    var transferData = Vue.computed(function () {
      return transfer_utils_index.getTransferData(props2.data, props2.keys, isTreeMode.value);
    });
    var sourceList = Vue.computed(function () {
      return transfer_utils_index.filterTransferData(transferData.value, valueList.value, false, isTreeMode.value);
    });
    var targetList = Vue.computed(function () {
      return transfer_utils_index.filterTransferData(transferData.value, valueList.value, true, isTreeMode.value);
    });
    var checkedValue = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"]({}, transfer_utils_index.SOURCE, transfer_utils_index.getDataValues(sourceList.value, innerChecked.value, {
        isTreeMode: isTreeMode.value
      })), transfer_utils_index.TARGET, transfer_utils_index.getDataValues(targetList.value, innerChecked.value, {
        isTreeMode: isTreeMode.value
      }));
    });
    var hasFooter = Vue.computed(function () {
      return !!slots.footer || !!props2.footer;
    });
    var showPagination = Vue.computed(function () {
      return !!props2.pagination && !slots.content;
    });
    var showSearch = Vue.computed(function () {
      return !!props2.search;
    });
    var footerOption = Vue.computed(function () {
      var footer = props2.footer || "";
      return transfer_utils_index.getTransferListOption(footer);
    });
    var emptyOption = Vue.computed(function () {
      return transfer_utils_index.getTransferListOption(props2.empty);
    });
    var searchOption = Vue.computed(function () {
      return transfer_utils_index.getTransferListOption(props2.search);
    });
    var checkAllOption = Vue.computed(function () {
      return transfer_utils_index.getTransferListOption(props2.showCheckAll);
    });
    var disabledOption = Vue.computed(function () {
      return transfer_utils_index.getTransferListOption(disabled.value);
    });
    var titleOption = Vue.computed(function () {
      return transfer_utils_index.getTransferListOption(props2.title);
    });
    var paginationOption = Vue.computed(function () {
      return transfer_utils_index.getTransferListOption(props2.pagination);
    });
    var handleCheckedChange = function handleCheckedChange(val, listType) {
      var _props2$onCheckedChan;
      var sourceChecked = listType === transfer_utils_index.SOURCE ? val : checkedValue.value[transfer_utils_index.SOURCE];
      var targetChecked = listType === transfer_utils_index.TARGET ? val : checkedValue.value[transfer_utils_index.TARGET];
      var checked2 = [].concat(_toConsumableArray__default["default"](sourceChecked), _toConsumableArray__default["default"](targetChecked));
      var event = {
        checked: checked2,
        sourceChecked: sourceChecked,
        targetChecked: targetChecked,
        type: listType
      };
      innerChecked.value = checked2;
      (_props2$onCheckedChan = props2.onCheckedChange) === null || _props2$onCheckedChan === void 0 || _props2$onCheckedChan.call(props2, event);
    };
    var transferTo = function transferTo(toDirection) {
      var oldTargetValue = JSON.parse(JSON.stringify(valueList.value));
      var newTargetValue;
      var selfCheckedValue = toDirection === transfer_utils_index.TARGET ? checkedValue.value[transfer_utils_index.SOURCE] : checkedValue.value[transfer_utils_index.TARGET];
      if (toDirection === transfer_utils_index.SOURCE) {
        newTargetValue = oldTargetValue.filter(function (v) {
          return !selfCheckedValue.some(function (cv) {
            return isEqual.isEqual(cv, v);
          });
        });
      } else if (props2.targetSort === "original") {
        var remainValue = transferData.value.reduce(function (acc, data) {
          if (oldTargetValue.some(function (v) {
            return isEqual.isEqual(v, data.value);
          }) && data.disabled) {
            return acc.concat(data.value);
          }
          return acc;
        }, []);
        newTargetValue = transfer_utils_index.getDataValues(transferData.value, oldTargetValue.concat(selfCheckedValue), {
          isTreeMode: isTreeMode.value,
          remainValue: remainValue
        });
      } else if (props2.targetSort === "unshift") {
        newTargetValue = selfCheckedValue.concat(oldTargetValue);
      } else {
        newTargetValue = oldTargetValue.concat(selfCheckedValue);
      }
      handleCheckedChange([], toDirection === transfer_utils_index.SOURCE ? transfer_utils_index.TARGET : transfer_utils_index.SOURCE);
      var params = {
        type: toDirection,
        movedValue: selfCheckedValue
      };
      setInnerValue(newTargetValue, params);
    };
    var transferToRight = function transferToRight() {
      transferTo(transfer_utils_index.TARGET);
    };
    var transferToLeft = function transferToLeft() {
      transferTo(transfer_utils_index.SOURCE);
    };
    var handleScroll = function handleScroll(e, listType) {
      var _props2$onScroll;
      var target = e.target;
      var bottomDistance = target.scrollHeight - target.scrollTop - target.clientHeight;
      var event = {
        e: e,
        bottomDistance: bottomDistance,
        type: listType
      };
      (_props2$onScroll = props2.onScroll) === null || _props2$onScroll === void 0 || _props2$onScroll.call(props2, event);
    };
    var handleSearch = function handleSearch(e) {
      var _props2$onSearch;
      (_props2$onSearch = props2.onSearch) === null || _props2$onSearch === void 0 || _props2$onSearch.call(props2, e);
    };
    var handlePageChange = function handlePageChange(pageInfo, listType) {
      var _props2$onPageChange;
      (_props2$onPageChange = props2.onPageChange) === null || _props2$onPageChange === void 0 || _props2$onPageChange.call(props2, pageInfo, {
        type: listType
      });
    };
    var handleDataChange = function handleDataChange(data, movedValue) {
      setInnerValue(data, {
        type: transfer_utils_index.TARGET,
        movedValue: movedValue
      });
    };
    var renderTransferList = function renderTransferList(listType) {
      var scopedSlots = pick.pick(slots, ["title", "empty", "footer", "operation", "transferItem", "default", "tree"]);
      return Vue.createVNode(transfer_components_transferList["default"], {
        "checkboxProps": props2.checkboxProps,
        "transferItem": props2.transferItem,
        "list-type": listType,
        "title": titleOption.value[listType],
        "data-source": listType === transfer_utils_index.SOURCE ? sourceList.value : targetList.value,
        "checked-value": checkedValue.value[listType],
        "disabled": disabledOption.value[listType],
        "search": searchOption.value[listType],
        "pagination": paginationOption.value[listType],
        "check-all": checkAllOption.value[listType],
        "footer": footerOption.value[listType],
        "empty": emptyOption.value[listType],
        "onCheckedChange": function onCheckedChange($event) {
          return handleCheckedChange($event, listType);
        },
        "onScroll": function onScroll($event) {
          return handleScroll($event, listType);
        },
        "onSearch": handleSearch,
        "onPageChange": function onPageChange($event) {
          return handlePageChange($event, listType);
        },
        "isTreeMode": isTreeMode.value,
        "onDataChange": handleDataChange,
        "currentValue": valueList.value,
        "draggable": props2.targetDraggable && listType === transfer_utils_index.TARGET
      }, _isSlot(scopedSlots) ? scopedSlots : {
        "default": function _default() {
          return [scopedSlots];
        }
      });
    };
    return function () {
      return Vue.createVNode("div", {
        "class": ["".concat(classPrefix.value, "-transfer"), showSearch.value ? "".concat(classPrefix.value, "-transfer__search") : "", hasFooter.value ? "".concat(classPrefix.value, "-transfer__footer") : "", showPagination.value ? "".concat(classPrefix.value, "-transfer__pagination") : "", isTreeMode.value ? "".concat(classPrefix.value, "-transfer--with-tree") : ""]
      }, [renderTransferList(transfer_utils_index.SOURCE), Vue.createVNode(transfer_components_transferOperations["default"], {
        "leftDisabled": disabledOption.value[transfer_utils_index.TARGET] || leftButtonDisabled.value || checkedValue.value[transfer_utils_index.TARGET].length === 0,
        "rightDisabled": disabledOption.value[transfer_utils_index.SOURCE] || rightButtonDisabled.value || checkedValue.value[transfer_utils_index.SOURCE].length === 0,
        "operation": props2.operation,
        "onMoveToRight": transferToRight,
        "onMoveToLeft": transferToLeft
      }, {
        operation: slots.operation
      }), renderTransferList(transfer_utils_index.TARGET)]);
    };
  }
});

exports["default"] = _Transfer;
