// tdesign-vue-next/cjs/table/hooks/useFilter.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _typeof = require('@babel/runtime/helpers/typeof');
var table_hooks_useClassName = require('./useClassName.js');
var button_index = require('../../button/index.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../../_chunks/dep-c7472ccd.js');
var index = require('../../_chunks/dep-0e56b1da.js');
var configProvider_hooks_useConfig = require('../../config-provider/hooks/useConfig.js');
var index$1 = require('../../_chunks/dep-86694091.js');
require('../../_chunks/dep-d1a4892a.js');
var table_components_filterController = require('../components/filter-controller.js');
var utils = require('../../_chunks/dep-f0cbacdb.js');
var table_hooks_useTableHeader = require('./useTableHeader.js');
var isFunction = require('../../_chunks/dep-37af9215.js');
require('../../_chunks/dep-b56e7ff5.js');
require('../../_chunks/dep-bf2e7501.js');
require('../../_chunks/dep-07f46433.js');
require('../../_chunks/dep-0b08f67c.js');
require('dayjs');
require('../../_chunks/dep-eb1f700b.js');
require('../../_chunks/dep-9399a4e2.js');
require('../../_chunks/dep-58ff00b9.js');
require('../../_chunks/dep-e2205a6f.js');
require('../../_chunks/dep-c644d54f.js');
require('../../_chunks/dep-0c9948d8.js');
require('../../_chunks/dep-2e9fed7b.js');
require('../../_chunks/dep-6f4a5c31.js');
require('../../_chunks/dep-d1d374b4.js');
require('../../_chunks/dep-a324ad6a.js');
require('../../_chunks/dep-06e53606.js');
require('../../_chunks/dep-bde8abb8.js');
require('../../_chunks/dep-41cc84a3.js');
require('../../_chunks/dep-c51f4345.js');
require('../../_chunks/dep-e2c29461.js');
require('../../_chunks/dep-52d3718c.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../../button/button.js');
require('../../loading/index.js');
require('../../_chunks/dep-548039eb.js');
require('../../loading/plugin.js');
require('../../_chunks/dep-4a5eca86.js');
require('../../loading/icon/gradient.js');
require('../../_chunks/dep-fd403354.js');
require('../../_chunks/dep-585d18dc.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../../_chunks/dep-8308792f.js');
require('../../_chunks/dep-d5e8d527.js');
require('../../_chunks/dep-8e4f83bf.js');
require('../../_chunks/dep-eb214be8.js');
require('../../_chunks/dep-11e6da14.js');
require('../../_chunks/dep-6c498eb2.js');
require('../../loading/props.js');
require('../../_chunks/dep-c8d40779.js');
require('../../_chunks/dep-388cb46f.js');
require('../../_chunks/dep-2333b13f.js');
require('../../_chunks/dep-5729cb14.js');
require('../../_chunks/dep-89d65b78.js');
require('../../_chunks/dep-dc9d7b38.js');
require('../../_chunks/dep-445d3f65.js');
require('../../_chunks/dep-09b19404.js');
require('../../_chunks/dep-f8f462e0.js');
require('../../_chunks/dep-47af4fff.js');
require('../../_chunks/dep-ea9fd5a2.js');
require('../../_chunks/dep-a8f427e5.js');
require('../../_chunks/dep-9997364c.js');
require('../../_chunks/dep-1e2d078f.js');
require('../../_chunks/dep-7d888cd2.js');
require('../../_chunks/dep-ffd72e3b.js');
require('../../_chunks/dep-6ed4388e.js');
require('../../_chunks/dep-742210e1.js');
require('../../button/props.js');
require('../../_chunks/dep-a8de5365.js');
require('../../_chunks/dep-850526b5.js');
require('../../_chunks/dep-a90757f2.js');
require('tdesign-icons-vue-next');
require('../../popup/index.js');
require('../../popup/popup.js');
require('@popperjs/core');
require('../../_chunks/dep-d149369c.js');
require('../../popup/container.js');
require('../../popup/props.js');
require('../../_chunks/dep-6344c74f.js');
require('../../_chunks/dep-d8ab5fb5.js');
require('../../_chunks/dep-cde052f0.js');
require('../../_chunks/dep-a3501e24.js');
require('../../checkbox/index.js');
require('../../checkbox/checkbox.js');
require('../../checkbox/props.js');
require('../../_chunks/dep-4725a846.js');
require('../../checkbox/constants/index.js');
require('../../checkbox/hooks/useCheckboxLazyLoad.js');
require('../../_chunks/dep-b4842612.js');
require('../../checkbox/hooks/useKeyboardEvent.js');
require('../../_chunks/dep-1177d395.js');
require('../../checkbox/group.js');
require('../../checkbox/checkbox-group-props.js');
require('../../_chunks/dep-e53cd0a5.js');
require('../../_chunks/dep-2bfa66aa.js');
require('../../_chunks/dep-c2cca927.js');
require('../../_chunks/dep-5e1fbdf8.js');
require('../../radio/index.js');
require('../../radio/radio.js');
require('../../radio/props.js');
require('../../radio/constants/index.js');
require('../../_chunks/dep-d3b20f18.js');
require('../../_chunks/dep-3d9f41ae.js');
require('../../_chunks/dep-117093bb.js');
require('../../radio/group.js');
require('@babel/runtime/helpers/asyncToGenerator');
require('@babel/runtime/regenerator');
require('../../radio/radio-group-props.js');
require('../../radio/radio-button.js');
require('../../radio/hooks/useKeyboard.js');
require('../../watermark/hooks/index.js');
require('../../_chunks/dep-ffe05cb1.js');
require('../../_chunks/dep-e70e28c7.js');
require('../../_chunks/dep-880baab0.js');
require('../../input/index.js');
require('../../input/input.js');
require('../../input/props.js');
require('../../_chunks/dep-125f71ff.js');
require('../../input/hooks/useInput.js');
require('../../form/constants/index.js');
require('../../input/hooks/useLengthLimit.js');
require('../../_chunks/dep-f7deb88a.js');
require('../../input/hooks/useInputEventHandler.js');
require('../../input/hooks/useInputWidth.js');
require('../../input/input-group.js');
require('../../input/input-group-props.js');
require('../../_chunks/dep-954dbe59.js');
require('../../_chunks/dep-0797927a.js');
require('../../_chunks/dep-cf727d09.js');
require('../../_chunks/dep-e492b603.js');
require('../../_chunks/dep-c8284192.js');
require('./useMultiHeader.js');
require('../components/ellipsis.js');
require('../../tooltip/index.js');
require('../../tooltip/tooltip.js');
require('../../tooltip/props.js');
require('../../tooltip/utils/index.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function isFilterValueExist(value) {
  var isArrayTrue = value instanceof Array && value.length;
  var isObject = _typeof__default["default"](value) === "object" && !(value instanceof Array);
  var isObjectTrue = value !== null && isObject && Object.keys(value).length;
  return isArrayTrue || isObjectTrue || !["null", "", "undefined"].includes(String(value));
}
function filterEmptyData(data) {
  var newFilterValue = {};
  Object.keys(data).forEach(function (key) {
    var item = data[key];
    if (isFilterValueExist(item)) {
      newFilterValue[key] = item;
    }
  });
  return newFilterValue;
}
function useFilter(props, context) {
  var primaryTableRef = Vue.ref(null);
  var _useConfig = configProvider_hooks_useConfig.useConfig("table", props.locale),
    t = _useConfig.t,
    globalConfig = _useConfig.globalConfig;
  var renderTNode = index.useTNodeDefault();
  var _toRefs = Vue.toRefs(props),
    filterValue = _toRefs.filterValue,
    columns = _toRefs.columns;
  var _useClassName = table_hooks_useClassName["default"](),
    tableFilterClasses = _useClassName.tableFilterClasses,
    isFocusClass = _useClassName.isFocusClass;
  var isTableOverflowHidden = Vue.ref();
  var _useDefaultValue = index$1.useDefaultValue(filterValue, props.defaultFilterValue, props.onFilterChange, "filterValue"),
    _useDefaultValue2 = _slicedToArray__default["default"](_useDefaultValue, 2),
    tFilterValue = _useDefaultValue2[0],
    setTFilterValue = _useDefaultValue2[1];
  var innerFilterValue = Vue.ref(tFilterValue.value);
  var hasEmptyCondition = Vue.computed(function () {
    var filterEmpty = filterEmptyData(tFilterValue.value || {});
    return !tFilterValue.value || !Object.keys(filterEmpty).length;
  });
  Vue.watch([tFilterValue], function (_ref) {
    var _ref2 = _slicedToArray__default["default"](_ref, 1),
      val = _ref2[0];
    innerFilterValue.value = val;
  });
  function renderFirstFilterRow() {
    var _props$pagination, _props$data;
    if (hasEmptyCondition.value) return null;
    var defaultNode = Vue.createVNode("div", {
      "class": tableFilterClasses.result
    }, [Vue.createVNode("span", null, [t(globalConfig.value.searchResultText, {
      result: getFilterResultContent(),
      count: ((_props$pagination = props.pagination) === null || _props$pagination === void 0 ? void 0 : _props$pagination.total) || ((_props$data = props.data) === null || _props$data === void 0 ? void 0 : _props$data.length)
    })]), Vue.createVNode(button_index.Button, {
      "theme": "primary",
      "variant": "text",
      "onClick": onResetAll
    }, {
      "default": function _default() {
        return [globalConfig.value.clearFilterResultButtonText];
      }
    })]);
    var filterContent = renderTNode("filterRow");
    if (props.filterRow && !filterContent || props.filterRow === null) return null;
    return Vue.createVNode("div", {
      "class": tableFilterClasses.inner
    }, [filterContent || defaultNode]);
  }
  function getFilterResultContent() {
    var arr = [];
    var columns2 = [];
    getAllColumns(props.columns, columns2);
    columns2.filter(function (col) {
      return col.filter;
    }).forEach(function (col, index) {
      var value = tFilterValue.value[col.colKey];
      if (col.filter.list && !["null"].includes(String(value))) {
        var formattedValue = value instanceof Array ? value : [value];
        var label = [];
        col.filter.list.forEach(function (option) {
          if (formattedValue.includes(option.value)) {
            label.push(option.label);
          }
        });
        value = label.join();
      }
      if (isFilterValueExist(value)) {
        var _col$filter, _col$filter2;
        var _label = isFunction.isFunction((_col$filter = col.filter) === null || _col$filter === void 0 ? void 0 : _col$filter.label) ? col.filter.label(Vue.h) : (_col$filter2 = col.filter) === null || _col$filter2 === void 0 ? void 0 : _col$filter2.label;
        var title = table_hooks_useTableHeader.renderTitle(context.slots, col, index);
        arr.push("".concat(_label || title, "\uFF1A").concat(value));
      }
    });
    return arr.join("\uFF1B");
  }
  function getAllColumns(col, columns2) {
    col.forEach(function (column) {
      if (column.children) {
        getAllColumns(column.children, columns2);
      }
      columns2.push(column);
    });
  }
  function onInnerFilterChange(val, column) {
    var filterValue2 = _objectSpread(_objectSpread({}, innerFilterValue.value), {}, _defineProperty__default["default"]({}, column.colKey, val));
    innerFilterValue.value = filterValue2;
    if (!column.filter.showConfirmAndReset) {
      emitFilterChange(filterValue2, "filter-change", column);
    }
  }
  function emitFilterChange(filterValue2, trigger, column) {
    var _props$onChange;
    setTFilterValue(filterValue2, {
      col: column,
      trigger: trigger
    });
    (_props$onChange = props.onChange) === null || _props$onChange === void 0 || _props$onChange.call(props, {
      filter: filterValue2
    }, {
      trigger: "filter"
    });
  }
  function onReset(column) {
    var _ref3, _column$filter$resetV;
    var filterValue2 = _objectSpread(_objectSpread({}, tFilterValue.value), {}, _defineProperty__default["default"]({}, column.colKey, (_ref3 = (_column$filter$resetV = column.filter.resetValue) !== null && _column$filter$resetV !== void 0 ? _column$filter$resetV : {
      single: "",
      multiple: [],
      input: ""
    }[column.filter.type]) !== null && _ref3 !== void 0 ? _ref3 : ""));
    emitFilterChange(filterValue2, "reset", column);
  }
  function onResetAll() {
    var resetValue = utils.getColumnsResetValue(columns.value);
    emitFilterChange(resetValue, "clear", void 0);
  }
  function onConfirm(column) {
    emitFilterChange(innerFilterValue.value, "confirm", column);
  }
  function renderFilterIcon(_ref4) {
    var _primaryTableRef$valu;
    var col = _ref4.col,
      colIndex = _ref4.colIndex;
    return Vue.createVNode(table_components_filterController["default"], {
      "column": col,
      "colIndex": colIndex,
      "filterIcon": props.filterIcon,
      "tFilterValue": tFilterValue.value,
      "innerFilterValue": innerFilterValue.value,
      "tableFilterClasses": tableFilterClasses,
      "isFocusClass": isFocusClass,
      "popupProps": col.filter.popupProps,
      "attach": props.attach,
      "locale": props.locale,
      "onInnerFilterChange": onInnerFilterChange,
      "onReset": onReset,
      "onConfirm": onConfirm,
      "primaryTableElement": (_primaryTableRef$valu = primaryTableRef.value) === null || _primaryTableRef$valu === void 0 ? void 0 : _primaryTableRef$valu.$el,
      "onVisibleChange": onPopupVisibleChange
    }, {
      filterIcon: context.slots.filterIcon
    });
  }
  function setFilterPrimaryTableRef(primaryTableElement) {
    primaryTableRef.value = primaryTableElement;
  }
  function onPopupVisibleChange(visible) {
    if (visible && !isTableOverflowHidden.value) {
      isTableOverflowHidden.value = !visible;
    }
  }
  return {
    hasEmptyCondition: hasEmptyCondition,
    isTableOverflowHidden: isTableOverflowHidden,
    renderFilterIcon: renderFilterIcon,
    renderFirstFilterRow: renderFirstFilterRow,
    setFilterPrimaryTableRef: setFilterPrimaryTableRef
  };
}

exports["default"] = useFilter;
