// tdesign-vue-next/cjs/date-picker/DateRangePicker.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var dayjs = require('dayjs');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-c7472ccd.js');
var index = require('../_chunks/dep-b56e7ff5.js');
var index$1 = require('../_chunks/dep-a8de5365.js');
require('../_chunks/dep-d1a4892a.js');
var index$2 = require('../_chunks/dep-4725a846.js');
require('@babel/runtime/helpers/defineProperty');
var datePicker_dateRangePickerProps = require('./date-range-picker-props.js');
var rangeInput_index = require('../range-input/index.js');
var datePicker_components_panel_RangePanel = require('./components/panel/RangePanel.js');
var datePicker_hooks_useRange = require('./hooks/useRange.js');
var format = require('../_chunks/dep-d28d8b91.js');
var utils = require('../_chunks/dep-41271801.js');
var datePicker_utils_index = require('./utils/index.js');
var isArray = require('../_chunks/dep-d1d374b4.js');
var isFunction = require('../_chunks/dep-37af9215.js');
require('../_chunks/dep-bf2e7501.js');
require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-e2205a6f.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-2e9fed7b.js');
require('../_chunks/dep-6f4a5c31.js');
require('../_chunks/dep-a324ad6a.js');
require('../_chunks/dep-06e53606.js');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('../_chunks/dep-850526b5.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../range-input/range-input.js');
require('tdesign-icons-vue-next');
require('../input/index.js');
require('../input/input.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../input/props.js');
require('../_chunks/dep-0e56b1da.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../_chunks/dep-6c498eb2.js');
require('../_chunks/dep-125f71ff.js');
require('../input/hooks/useInput.js');
require('../form/constants/index.js');
require('../_chunks/dep-d149369c.js');
require('../input/hooks/useLengthLimit.js');
require('../_chunks/dep-f7deb88a.js');
require('../_chunks/dep-585d18dc.js');
require('../_chunks/dep-8308792f.js');
require('../_chunks/dep-d5e8d527.js');
require('../_chunks/dep-8e4f83bf.js');
require('../_chunks/dep-880baab0.js');
require('../input/hooks/useInputEventHandler.js');
require('../input/hooks/useInputWidth.js');
require('../_chunks/dep-6344c74f.js');
require('../input/input-group.js');
require('../input/input-group-props.js');
require('../_chunks/dep-742210e1.js');
require('../range-input/props.js');
require('../_chunks/dep-6f95225b.js');
require('../range-input/range-input-popup.js');
require('../popup/index.js');
require('../popup/popup.js');
require('@popperjs/core');
require('../_chunks/dep-fd403354.js');
require('../popup/container.js');
require('../popup/props.js');
require('../_chunks/dep-d8ab5fb5.js');
require('../_chunks/dep-cde052f0.js');
require('../_chunks/dep-a3501e24.js');
require('../range-input/range-input-popup-props.js');
require('../select-input/hooks/useOverlayInnerStyle.js');
require('./components/panel/PanelContent.js');
require('./components/base/Header.js');
require('../pagination/index.js');
require('../pagination/pagination.js');
require('../_chunks/dep-86694091.js');
require('../input-number/index.js');
require('../input-number/input-number.js');
require('../button/index.js');
require('../button/button.js');
require('../loading/index.js');
require('../_chunks/dep-548039eb.js');
require('../loading/plugin.js');
require('../_chunks/dep-4a5eca86.js');
require('../loading/icon/gradient.js');
require('../loading/props.js');
require('../_chunks/dep-f8f462e0.js');
require('../_chunks/dep-47af4fff.js');
require('../_chunks/dep-ea9fd5a2.js');
require('../_chunks/dep-a8f427e5.js');
require('../_chunks/dep-9997364c.js');
require('../_chunks/dep-1e2d078f.js');
require('../_chunks/dep-7d888cd2.js');
require('../_chunks/dep-ffd72e3b.js');
require('../_chunks/dep-6ed4388e.js');
require('../button/props.js');
require('../_chunks/dep-a90757f2.js');
require('../input-number/props.js');
require('../input-number/hooks/useInputNumber.js');
require('../_chunks/dep-3688812b.js');
require('../_chunks/dep-77efa7eb.js');
require('../_chunks/dep-cf727d09.js');
require('../select/index.js');
require('../select/select.js');
require('@babel/runtime/helpers/asyncToGenerator');
require('@babel/runtime/regenerator');
require('../common-components/fake-arrow.js');
require('../select-input/index.js');
require('../select-input/select-input.js');
require('../select-input/props.js');
require('../select-input/hooks/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/tag-input.js');
require('../tag-input/props.js');
require('../tag-input/hooks/useDragSorter.js');
require('../tag-input/hooks/useHover.js');
require('../tag-input/hooks/useTagScroll.js');
require('../tag-input/hooks/useTagList.js');
require('../tag/index.js');
require('../tag/tag.js');
require('tinycolor2');
require('../tag/props.js');
require('../tag/check-tag.js');
require('../tag/check-tag-props.js');
require('../_chunks/dep-1177d395.js');
require('../tag/check-tag-group.js');
require('../tag/check-tag-group-props.js');
require('../select-input/hooks/useSingle.js');
require('../_chunks/dep-a66df673.js');
require('../_chunks/dep-f3b67279.js');
require('../_chunks/dep-117093bb.js');
require('../select/components/select-panel.js');
require('../select/option.js');
require('../select/option-props.js');
require('../checkbox/index.js');
require('../checkbox/checkbox.js');
require('../checkbox/props.js');
require('../checkbox/constants/index.js');
require('../checkbox/hooks/useCheckboxLazyLoad.js');
require('../_chunks/dep-b4842612.js');
require('../checkbox/hooks/useKeyboardEvent.js');
require('../checkbox/group.js');
require('../checkbox/checkbox-group-props.js');
require('../_chunks/dep-e53cd0a5.js');
require('../_chunks/dep-2bfa66aa.js');
require('../_chunks/dep-c2cca927.js');
require('../_chunks/dep-5e1fbdf8.js');
require('../select/utils/index.js');
require('../select/constants/index.js');
require('../select/option-group.js');
require('../select/option-group-props.js');
require('../select/props.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../_chunks/dep-d3b20f18.js');
require('../_chunks/dep-3d9f41ae.js');
require('../select/hooks/useKeyboardControl.js');
require('../select/hooks/useSelectOptions.js');
require('../_chunks/dep-5c150838.js');
require('../input-adornment/index.js');
require('../input-adornment/input-adornment.js');
require('../input-adornment/props.js');
require('../pagination/props.js');
require('../pagination/hooks/useMoreAction.js');
require('../pagination/hooks/usePaginationClasses.js');
require('../pagination/pagination-mini.js');
require('../pagination/pagination-mini-props.js');
require('../tooltip/index.js');
require('../tooltip/tooltip.js');
require('../tooltip/props.js');
require('../tooltip/utils/index.js');
require('./hooks/useSelectRange.js');
require('./components/base/Table.js');
require('./components/base/Cell.js');
require('../time-picker/index.js');
require('../time-picker/time-picker.js');
require('../_chunks/dep-a7068a83.js');
require('../time-picker/panel/time-picker-panel.js');
require('../_chunks/dep-83ab8a79.js');
require('../time-picker/props.js');
require('../_chunks/dep-5fee93c0.js');
require('../_chunks/dep-e01a4fce.js');
require('../time-picker/time-range-picker.js');
require('../time-picker/time-range-picker-props.js');
require('./components/panel/ExtraContent.js');
require('./components/base/Footer.js');
require('./hooks/useDisableDate.js');
require('./hooks/useTableData.js');
require('./hooks/useRangeValue.js');
require('@babel/runtime/helpers/objectDestructuringEmpty');
require('@babel/runtime/helpers/extends');
require('../_chunks/dep-0797927a.js');
require('../_chunks/dep-2f14e275.js');
require('../_chunks/dep-c8284192.js');
require('../_chunks/dep-c2f54d6f.js');
require('../_chunks/dep-e70e28c7.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var _DateRangePicker = Vue.defineComponent({
  name: "TDateRangePicker",
  props: datePicker_dateRangePickerProps["default"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var COMPONENT_NAME = index.usePrefixClass("date-range-picker");
    var _useRange = datePicker_hooks_useRange.useRange(props2),
      inputValue = _useRange.inputValue,
      popupVisible = _useRange.popupVisible,
      rangeInputProps = _useRange.rangeInputProps,
      popupProps = _useRange.popupProps,
      value = _useRange.value,
      year = _useRange.year,
      month = _useRange.month,
      time = _useRange.time,
      cacheValue = _useRange.cacheValue,
      activeIndex = _useRange.activeIndex,
      isHoverCell = _useRange.isHoverCell,
      isFirstValueSelected = _useRange.isFirstValueSelected,
      onChange = _useRange.onChange;
    var isDisabled = index$1.useDisabled();
    var isReadOnly = index$2.useReadonly();
    var formatRef = Vue.computed(function () {
      return format.getDefaultFormat({
        mode: props2.mode,
        enableTimePicker: props2.enableTimePicker,
        format: props2.format,
        valueType: props2.valueType
      });
    });
    var isSelected = Vue.ref(false);
    Vue.watch(popupVisible, function (visible) {
      if (visible) {
        if (isArray.isArray(props2.disabled)) activeIndex.value = props2.disabled[0] ? 1 : 0;
        isSelected.value = false;
        cacheValue.value = format.formatDate(value.value || [], {
          format: formatRef.value.valueType,
          targetFormat: formatRef.value.format
        });
        time.value = format.formatTime(value.value || [dayjs__default["default"]().format(formatRef.value.timeFormat), dayjs__default["default"]().format(formatRef.value.timeFormat)], formatRef.value.format, formatRef.value.timeFormat, props2.defaultTime);
        if (!value.value.length) {
          if (props2.range && isArray.isArray(props2.range) || props2.panelActiveDate) {
            var _startRangeBounds$min, _startRangeBounds$min2, _startRangeBounds$max, _startRangeBounds$min3, _startRangeBounds$min4, _startRangeBounds$max2, _endRangeBounds$min$g, _endRangeBounds$min, _endRangeBounds$max, _endRangeBounds$min$g2, _endRangeBounds$min2, _endRangeBounds$max2, _startPanelActiveDate, _startPanelActiveDate2, _startPanelActiveDate3, _startPanelActiveDate4, _endPanelActiveDate$y, _endPanelActiveDate, _endPanelActiveDate2, _endPanelActiveDate3;
            var startRange = props2.range;
            var endRange = props2.range;
            if (isArray.isArray(props2.range)) {
              var _props2$range = _slicedToArray__default["default"](props2.range, 2),
                first = _props2$range[0],
                second = _props2$range[1];
              if (isArray.isArray(first) || isFunction.isFunction(first) || isArray.isArray(second) || isFunction.isFunction(second)) {
                startRange = first;
                endRange = second;
              }
            }
            var startRangeBounds = utils.getRangeBounds(startRange);
            var endRangeBounds = utils.getRangeBounds(endRange);
            var startYearFromRange = (_startRangeBounds$min = (_startRangeBounds$min2 = startRangeBounds.min) === null || _startRangeBounds$min2 === void 0 ? void 0 : _startRangeBounds$min2.getFullYear()) !== null && _startRangeBounds$min !== void 0 ? _startRangeBounds$min : (_startRangeBounds$max = startRangeBounds.max) === null || _startRangeBounds$max === void 0 ? void 0 : _startRangeBounds$max.getFullYear();
            var startMonthFromRange = (_startRangeBounds$min3 = (_startRangeBounds$min4 = startRangeBounds.min) === null || _startRangeBounds$min4 === void 0 ? void 0 : _startRangeBounds$min4.getMonth()) !== null && _startRangeBounds$min3 !== void 0 ? _startRangeBounds$min3 : (_startRangeBounds$max2 = startRangeBounds.max) === null || _startRangeBounds$max2 === void 0 ? void 0 : _startRangeBounds$max2.getMonth();
            var endYearFromRange = (_endRangeBounds$min$g = (_endRangeBounds$min = endRangeBounds.min) === null || _endRangeBounds$min === void 0 ? void 0 : _endRangeBounds$min.getFullYear()) !== null && _endRangeBounds$min$g !== void 0 ? _endRangeBounds$min$g : (_endRangeBounds$max = endRangeBounds.max) === null || _endRangeBounds$max === void 0 ? void 0 : _endRangeBounds$max.getFullYear();
            var endMonthFromRange = (_endRangeBounds$min$g2 = (_endRangeBounds$min2 = endRangeBounds.min) === null || _endRangeBounds$min2 === void 0 ? void 0 : _endRangeBounds$min2.getMonth()) !== null && _endRangeBounds$min$g2 !== void 0 ? _endRangeBounds$min$g2 : (_endRangeBounds$max2 = endRangeBounds.max) === null || _endRangeBounds$max2 === void 0 ? void 0 : _endRangeBounds$max2.getMonth();
            var startPanelActiveDate = props2.panelActiveDate;
            var endPanelActiveDate = props2.panelActiveDate;
            if (isArray.isArray(props2.panelActiveDate)) {
              var _props2$panelActiveDa = _slicedToArray__default["default"](props2.panelActiveDate, 2);
              startPanelActiveDate = _props2$panelActiveDa[0];
              endPanelActiveDate = _props2$panelActiveDa[1];
            }
            var leftYear = (_startPanelActiveDate = (_startPanelActiveDate2 = startPanelActiveDate) === null || _startPanelActiveDate2 === void 0 ? void 0 : _startPanelActiveDate2.year) !== null && _startPanelActiveDate !== void 0 ? _startPanelActiveDate : startYearFromRange;
            var leftMonth = (_startPanelActiveDate3 = startPanelActiveDate) !== null && _startPanelActiveDate3 !== void 0 && _startPanelActiveDate3.month ? Number((_startPanelActiveDate4 = startPanelActiveDate) === null || _startPanelActiveDate4 === void 0 ? void 0 : _startPanelActiveDate4.month) - 1 : startMonthFromRange;
            var rightYear = (_endPanelActiveDate$y = (_endPanelActiveDate = endPanelActiveDate) === null || _endPanelActiveDate === void 0 ? void 0 : _endPanelActiveDate.year) !== null && _endPanelActiveDate$y !== void 0 ? _endPanelActiveDate$y : endYearFromRange;
            var rightMonth = (_endPanelActiveDate2 = endPanelActiveDate) !== null && _endPanelActiveDate2 !== void 0 && _endPanelActiveDate2.month ? Number((_endPanelActiveDate3 = endPanelActiveDate) === null || _endPanelActiveDate3 === void 0 ? void 0 : _endPanelActiveDate3.month) - 1 : endMonthFromRange;
            var _initYearMonthTime = format.initYearMonthTime({
                value: value.value,
                mode: props2.mode,
                format: formatRef.value.format,
                enableTimePicker: props2.enableTimePicker
              }),
              defaultYear = _initYearMonthTime.year,
              defaultMonth = _initYearMonthTime.month;
            year.value = [leftYear !== null && leftYear !== void 0 ? leftYear : defaultYear[0], rightYear !== null && rightYear !== void 0 ? rightYear : defaultYear[1]];
            month.value = [leftMonth !== null && leftMonth !== void 0 ? leftMonth : defaultMonth[0], rightMonth !== null && rightMonth !== void 0 ? rightMonth : defaultMonth[1]];
            if (year.value[0] === year.value[1] && month.value[0] === month.value[1] && !props2.enableTimePicker) {
              if (startRange === endRange) {
                if (props2.mode === "year") year.value[1] += 10;else if (props2.mode === "month" || props2.mode === "quarter") year.value[1] += 1;else month.value[1] += 1;
              }
            }
          } else {
            var _initYearMonthTime2 = format.initYearMonthTime({
                value: value.value,
                mode: props2.mode,
                format: formatRef.value.format,
                enableTimePicker: props2.enableTimePicker
              }),
              _defaultYear = _initYearMonthTime2.year,
              _defaultMonth = _initYearMonthTime2.month;
            year.value = _defaultYear;
            month.value = _defaultMonth;
          }
        } else if (value.value.length === 2 && !props2.enableTimePicker) {
          var nextMonth = value.value.map(function (v) {
            return format.parseToDayjs(v, formatRef.value.format).month();
          });
          year.value = value.value.map(function (v) {
            return format.parseToDayjs(v, formatRef.value.valueType).year();
          });
          if (year.value[0] === year.value[1] && nextMonth[0] === nextMonth[1]) {
            nextMonth[0] === 11 ? nextMonth[0] -= 1 : nextMonth[1] += 1;
          }
          month.value = nextMonth;
          if ((props2.mode === "month" || props2.mode === "quarter") && year.value[0] === year.value[1]) {
            year.value = [year.value[0], year.value[0] + 1];
          }
        } else {
          year.value = value.value.map(function (v) {
            return format.parseToDayjs(v, formatRef.value.format).year();
          });
          if (year.value.length === 1) year.value = [year.value[0], year.value[0]];
          month.value = value.value.map(function (v) {
            return format.parseToDayjs(v, formatRef.value.format).month();
          });
          if (month.value.length === 1) month.value = [month.value[0], Math.min(month.value[0] + 1, 11)];
        }
      } else {
        activeIndex.value = 0;
        isHoverCell.value = false;
        isFirstValueSelected.value = false;
        if (props2.needConfirm) {
          inputValue.value = format.formatDate(value.value, {
            format: formatRef.value.valueType,
            targetFormat: formatRef.value.format
          });
        } else {
          confirmValueChange();
        }
      }
    });
    function onCellMouseEnter(date) {
      isHoverCell.value = true;
      var nextValue = _toConsumableArray__default["default"](inputValue.value);
      nextValue[activeIndex.value] = format.formatDate(date, {
        format: formatRef.value.format
      });
      inputValue.value = nextValue;
    }
    function onCellMouseLeave() {
      isHoverCell.value = false;
      inputValue.value = cacheValue.value;
    }
    function onCellClick(date, _ref2) {
      var _props2$onPick;
      var e = _ref2.e;
      (_props2$onPick = props2.onPick) === null || _props2$onPick === void 0 || _props2$onPick.call(props2, date, {
        e: e,
        partial: activeIndex.value ? "end" : "start"
      });
      isHoverCell.value = false;
      isSelected.value = true;
      var nextValue = _toConsumableArray__default["default"](inputValue.value);
      nextValue[activeIndex.value] = format.formatDate(date, {
        format: formatRef.value.format
      });
      cacheValue.value = nextValue;
      inputValue.value = nextValue;
      if (props2.enableTimePicker) return;
      var notValidIndex = nextValue.findIndex(function (v) {
        return !v || !format.isValidDate(v, formatRef.value.format);
      });
      if (notValidIndex === -1 && nextValue.length === 2) {
        if (!isFirstValueSelected.value && format.parseToDayjs(nextValue[0], formatRef.value.format).isAfter(format.parseToDayjs(nextValue[1], formatRef.value.format))) {
          nextValue[activeIndex.value ? 0 : 1] = "";
          cacheValue.value = nextValue;
          inputValue.value = nextValue;
        } else {
          onChange === null || onChange === void 0 || onChange(format.formatDate(nextValue, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
            autoSwap: true,
            defaultTime: props2.defaultTime
          }), {
            dayjsValue: nextValue.map(function (v, i) {
              var _props2$defaultTime;
              return format.parseToDayjs(v, formatRef.value.format, void 0, void 0, (_props2$defaultTime = props2.defaultTime) === null || _props2$defaultTime === void 0 ? void 0 : _props2$defaultTime[i]);
            }),
            trigger: "pick"
          });
        }
      }
      if (Array.isArray(props2.disabled)) return;
      if (!isFirstValueSelected.value) {
        var nextIndex = notValidIndex;
        if (nextIndex === -1) nextIndex = activeIndex.value ? 0 : 1;
        activeIndex.value = nextIndex;
        isFirstValueSelected.value = nextValue.some(function (v) {
          return !!v;
        });
      } else {
        popupVisible.value = false;
      }
    }
    function onJumperClick(_ref3) {
      var trigger = _ref3.trigger,
        partial = _ref3.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      var monthCountMap = {
        date: 1,
        week: 1,
        month: 12,
        quarter: 12,
        year: 120
      };
      var monthCount = monthCountMap[props2.mode] || 0;
      var current = new Date(year.value[partialIndex], month.value[partialIndex]);
      var next = null;
      if (trigger === "prev") {
        next = utils.subtractMonth(current, monthCount);
      } else if (trigger === "current") {
        next = new Date();
      } else if (trigger === "next") {
        next = utils.addMonth(current, monthCount);
      }
      var nextYear = _toConsumableArray__default["default"](year.value);
      nextYear[partialIndex] = next.getFullYear();
      var nextMonth = _toConsumableArray__default["default"](month.value);
      nextMonth[partialIndex] = next.getMonth();
      var onlyYearSelect = ["year", "quarter", "month"].includes(props2.mode);
      var correctedDate = datePicker_utils_index.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
      nextYear = correctedDate.nextYear;
      nextMonth = correctedDate.nextMonth;
      var yearChanged = year.value[partialIndex] !== nextYear[partialIndex];
      var monthChanged = month.value[partialIndex] !== nextMonth[partialIndex];
      year.value = nextYear;
      month.value = nextMonth;
      if (yearChanged) {
        var _props2$onYearChange;
        (_props2$onYearChange = props2.onYearChange) === null || _props2$onYearChange === void 0 || _props2$onYearChange.call(props2, {
          partial: partial,
          year: nextYear[partialIndex],
          date: value.value.map(function (v) {
            return dayjs__default["default"](v).toDate();
          }),
          trigger: trigger === "current" ? "today" : "year-".concat(datePicker_utils_index.triggerMap[trigger])
        });
      }
      if (monthChanged) {
        var _props2$onMonthChange;
        (_props2$onMonthChange = props2.onMonthChange) === null || _props2$onMonthChange === void 0 || _props2$onMonthChange.call(props2, {
          partial: partial,
          month: nextMonth[partialIndex],
          date: value.value.map(function (v) {
            return dayjs__default["default"](v).toDate();
          }),
          trigger: trigger === "current" ? "today" : "month-".concat(datePicker_utils_index.triggerMap[trigger])
        });
      }
    }
    function onTimePickerChange(val) {
      var _extractTimeObj = utils.extractTimeObj(val),
        hours = _extractTimeObj.hours,
        minutes = _extractTimeObj.minutes,
        seconds = _extractTimeObj.seconds,
        milliseconds = _extractTimeObj.milliseconds,
        meridiem = _extractTimeObj.meridiem;
      var nextInputValue = _toConsumableArray__default["default"](inputValue.value);
      var changedInputValue = nextInputValue[activeIndex.value];
      var currentDate = !dayjs__default["default"](changedInputValue, formatRef.value.format).isValid() ? dayjs__default["default"]().year(year.value[activeIndex.value]).month(month.value[activeIndex.value]) : dayjs__default["default"](changedInputValue, formatRef.value.format);
      var nextHours = hours;
      if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
      if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
      var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
      nextInputValue[activeIndex.value] = nextDate;
      var nextTime = _toConsumableArray__default["default"](time.value);
      nextTime[activeIndex.value] = val;
      time.value = nextTime;
      isSelected.value = true;
      inputValue.value = format.formatDate(nextInputValue, {
        format: formatRef.value.format
      });
      cacheValue.value = format.formatDate(nextInputValue, {
        format: formatRef.value.format
      });
    }
    var confirmValueChange = function confirmValueChange(e) {
      var nextValue = _toConsumableArray__default["default"](inputValue.value);
      var notValidIndex = nextValue.findIndex(function (v) {
        return !v || !format.isValidDate(v, formatRef.value.format);
      });
      if (notValidIndex === -1 && nextValue.length === 2) {
        if (!isFirstValueSelected.value && format.parseToDayjs(nextValue[0], formatRef.value.format).isAfter(format.parseToDayjs(nextValue[1], formatRef.value.format))) {
          nextValue[activeIndex.value ? 0 : 1] = "";
          cacheValue.value = nextValue;
          inputValue.value = nextValue;
        } else {
          var _props2$onConfirm;
          var formattedValue = format.formatDate(nextValue, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
            autoSwap: true,
            defaultTime: props2.defaultTime
          });
          var isSame = Array.isArray(formattedValue) && formattedValue.length === 2 && formattedValue[0] != null && formattedValue[1] != null && value.value[0] != null && value.value[1] != null && dayjs__default["default"](formattedValue[0]).valueOf() === dayjs__default["default"](value.value[0]).valueOf() && dayjs__default["default"](formattedValue[1]).valueOf() === dayjs__default["default"](value.value[1]).valueOf();
          props2 === null || props2 === void 0 || (_props2$onConfirm = props2.onConfirm) === null || _props2$onConfirm === void 0 || _props2$onConfirm.call(props2, {
            date: nextValue.map(function (v) {
              return dayjs__default["default"](v).toDate();
            }),
            e: e || null,
            partial: activeIndex.value ? "end" : "start"
          });
          if (!isSame) {
            onChange === null || onChange === void 0 || onChange(formattedValue, {
              dayjsValue: nextValue.map(function (v, i) {
                var _props2$defaultTime2;
                return format.parseToDayjs(v, formatRef.value.format, void 0, void 0, (_props2$defaultTime2 = props2.defaultTime) === null || _props2$defaultTime2 === void 0 ? void 0 : _props2$defaultTime2[i]);
              }),
              trigger: "confirm"
            });
          }
        }
      }
    };
    function onConfirmClick(_ref4) {
      var e = _ref4.e;
      confirmValueChange(e);
      var nextValue = _toConsumableArray__default["default"](inputValue.value);
      var notValidIndex = nextValue.findIndex(function (v) {
        return !v || !format.isValidDate(v, formatRef.value.format);
      });
      if (Array.isArray(props2.disabled)) {
        popupVisible.value = false;
      } else if (!isFirstValueSelected.value) {
        var nextIndex = notValidIndex;
        if (nextIndex === -1) nextIndex = activeIndex.value ? 0 : 1;
        activeIndex.value = nextIndex;
        isFirstValueSelected.value = nextValue.some(function (v) {
          return !!v;
        });
      } else if (nextValue.length === 2) {
        popupVisible.value = false;
      }
    }
    function onPresetClick(preset, context) {
      var presetValue = preset;
      if (isFunction.isFunction(preset)) {
        presetValue = preset();
      }
      if (!isArray.isArray(presetValue)) {
        console.error("preset: ".concat(preset, " \u9884\u8BBE\u503C\u5FC5\u987B\u662F\u6570\u7EC4!"));
      } else {
        var _props2$onPresetClick;
        onChange === null || onChange === void 0 || onChange(format.formatDate(presetValue, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.valueType,
          autoSwap: true,
          defaultTime: props2.defaultTime
        }), {
          dayjsValue: presetValue.map(function (p, i) {
            var _props2$defaultTime3;
            return format.parseToDayjs(p, formatRef.value.format, void 0, void 0, (_props2$defaultTime3 = props2.defaultTime) === null || _props2$defaultTime3 === void 0 ? void 0 : _props2$defaultTime3[i]);
          }),
          trigger: "preset"
        });
        inputValue.value = format.formatDate(presetValue, {
          format: formatRef.value.format
        });
        popupVisible.value = false;
        (_props2$onPresetClick = props2.onPresetClick) === null || _props2$onPresetClick === void 0 || _props2$onPresetClick.call(props2, context);
      }
    }
    function onYearChange(nextVal, _ref5) {
      var _props2$onYearChange2;
      var partial = _ref5.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      if (props2.enableTimePicker) partialIndex = activeIndex.value;
      var nextYear = _toConsumableArray__default["default"](year.value);
      var nextMonth = _toConsumableArray__default["default"](month.value);
      nextYear[partialIndex] = nextVal;
      var onlyYearSelect = ["year", "quarter", "month"].includes(props2.mode);
      var correctedDate = datePicker_utils_index.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
      nextYear = correctedDate.nextYear;
      nextMonth = correctedDate.nextMonth;
      year.value = nextYear;
      if (!onlyYearSelect) month.value = nextMonth;
      (_props2$onYearChange2 = props2.onYearChange) === null || _props2$onYearChange2 === void 0 || _props2$onYearChange2.call(props2, {
        partial: partial,
        year: nextYear[partialIndex],
        date: value.value.map(function (v) {
          return dayjs__default["default"](v).toDate();
        }),
        trigger: "year-select"
      });
    }
    function onMonthChange(nextVal, _ref6) {
      var _props2$onMonthChange2;
      var partial = _ref6.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      if (props2.enableTimePicker) partialIndex = activeIndex.value;
      var nextMonth = _toConsumableArray__default["default"](month.value);
      nextMonth[partialIndex] = nextVal;
      if (year.value[0] === year.value[1]) {
        if (partialIndex === 0) {
          if (nextMonth[1] <= nextMonth[0]) {
            nextMonth[1] = nextMonth[0] + 1;
            if (nextMonth[1] === 12) {
              var _year$value, _year$value2;
              nextMonth[1] = 0;
              year.value = [(_year$value = year.value) === null || _year$value === void 0 ? void 0 : _year$value[0], ((_year$value2 = year.value) === null || _year$value2 === void 0 ? void 0 : _year$value2[1]) + 1];
            }
          }
        }
        if (partialIndex === 1) {
          nextMonth[0] = Math.min(nextMonth[0], nextMonth[1]);
          if (nextMonth[0] >= nextMonth[1]) {
            nextMonth[0] -= 1;
            if (nextMonth[0] === -1) {
              var _year$value3, _year$value4;
              nextMonth[0] = 11;
              year.value = [((_year$value3 = year.value) === null || _year$value3 === void 0 ? void 0 : _year$value3[0]) - 1, (_year$value4 = year.value) === null || _year$value4 === void 0 ? void 0 : _year$value4[1]];
            }
          }
        }
      }
      month.value = nextMonth;
      (_props2$onMonthChange2 = props2.onMonthChange) === null || _props2$onMonthChange2 === void 0 || _props2$onMonthChange2.call(props2, {
        partial: partial,
        month: nextMonth[partialIndex],
        date: value.value.map(function (v) {
          return dayjs__default["default"](v).toDate();
        }),
        trigger: "month-select"
      });
    }
    var panelProps = Vue.computed(function () {
      return {
        hoverValue: isHoverCell.value ? inputValue.value : [],
        value: isSelected.value ? cacheValue.value : value.value,
        isFirstValueSelected: isFirstValueSelected.value,
        activeIndex: activeIndex.value,
        year: year.value,
        month: month.value,
        format: formatRef.value.format,
        mode: props2.mode,
        presets: props2.presets,
        time: time.value,
        disableDate: props2.disableDate,
        disabled: isDisabled.value,
        disableTime: props2.disableTime,
        firstDayOfWeek: props2.firstDayOfWeek,
        timePickerProps: props2.timePickerProps,
        enableTimePicker: props2.enableTimePicker,
        presetsPlacement: props2.presetsPlacement,
        popupVisible: popupVisible.value,
        range: props2.range,
        panelPreselection: props2.panelPreselection,
        cancelRangeSelectLimit: props2.cancelRangeSelectLimit,
        needConfirm: props2.needConfirm,
        onCellClick: onCellClick,
        onCellMouseEnter: onCellMouseEnter,
        onCellMouseLeave: onCellMouseLeave,
        onJumperClick: onJumperClick,
        onConfirmClick: onConfirmClick,
        onPresetClick: onPresetClick,
        onYearChange: onYearChange,
        onMonthChange: onMonthChange,
        onTimePickerChange: onTimePickerChange,
        cell: props2.cell
      };
    });
    return function () {
      return Vue.createVNode("div", {
        "class": COMPONENT_NAME.value
      }, [Vue.createVNode(rangeInput_index.RangeInputPopup, {
        "readonly": isReadOnly.value,
        "disabled": isDisabled.value,
        "label": props2.label,
        "status": props2.status,
        "tips": props2.tips || slots.tips,
        "inputValue": inputValue.value,
        "popupProps": popupProps.value,
        "rangeInputProps": rangeInputProps.value,
        "popupVisible": popupVisible.value,
        "panel": function panel() {
          return Vue.createVNode(datePicker_components_panel_RangePanel["default"], panelProps.value, slots);
        }
      }, null)]);
    };
  }
});

exports["default"] = _DateRangePicker;
