// tdesign-vue-next/cjs/date-picker/DateRangePickerPanel.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var dayjs = require('dayjs');
var datePicker_dateRangePickerPanelProps = require('./date-range-picker-panel-props.js');
var datePicker_dateRangePickerProps = require('./date-range-picker-props.js');
var datePicker_components_panel_RangePanel = require('./components/panel/RangePanel.js');
var datePicker_hooks_useRangeValue = require('./hooks/useRangeValue.js');
var format = require('../_chunks/dep-d28d8b91.js');
var utils = require('../_chunks/dep-41271801.js');
var datePicker_utils_index = require('./utils/index.js');
var isFunction = require('../_chunks/dep-37af9215.js');
var isArray = require('../_chunks/dep-d1d374b4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-c7472ccd.js');
require('../_chunks/dep-bf2e7501.js');
require('../_chunks/dep-b56e7ff5.js');
require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-e2205a6f.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-2e9fed7b.js');
require('../_chunks/dep-6f4a5c31.js');
require('../_chunks/dep-a324ad6a.js');
require('../_chunks/dep-06e53606.js');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('../_chunks/dep-f7deb88a.js');
require('../_chunks/dep-d1a4892a.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('./components/panel/PanelContent.js');
require('./components/base/Header.js');
require('../pagination/index.js');
require('../pagination/pagination.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('tdesign-icons-vue-next');
require('../_chunks/dep-0e56b1da.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../_chunks/dep-86694091.js');
require('../_chunks/dep-125f71ff.js');
require('../_chunks/dep-d149369c.js');
require('../input-number/index.js');
require('../input-number/input-number.js');
require('../button/index.js');
require('../button/button.js');
require('../loading/index.js');
require('../_chunks/dep-548039eb.js');
require('../loading/plugin.js');
require('../_chunks/dep-4a5eca86.js');
require('../loading/icon/gradient.js');
require('../_chunks/dep-fd403354.js');
require('../_chunks/dep-585d18dc.js');
require('../_chunks/dep-8308792f.js');
require('../_chunks/dep-d5e8d527.js');
require('../_chunks/dep-8e4f83bf.js');
require('../_chunks/dep-6c498eb2.js');
require('../loading/props.js');
require('../_chunks/dep-f8f462e0.js');
require('../_chunks/dep-47af4fff.js');
require('../_chunks/dep-ea9fd5a2.js');
require('../_chunks/dep-a8f427e5.js');
require('../_chunks/dep-9997364c.js');
require('../_chunks/dep-1e2d078f.js');
require('../_chunks/dep-7d888cd2.js');
require('../_chunks/dep-ffd72e3b.js');
require('../_chunks/dep-6ed4388e.js');
require('../_chunks/dep-742210e1.js');
require('../button/props.js');
require('../_chunks/dep-a8de5365.js');
require('../_chunks/dep-850526b5.js');
require('../_chunks/dep-a90757f2.js');
require('../input/index.js');
require('../input/input.js');
require('../input/props.js');
require('../_chunks/dep-4725a846.js');
require('../input/hooks/useInput.js');
require('../form/constants/index.js');
require('../input/hooks/useLengthLimit.js');
require('../_chunks/dep-880baab0.js');
require('../input/hooks/useInputEventHandler.js');
require('../input/hooks/useInputWidth.js');
require('../_chunks/dep-6344c74f.js');
require('../input/input-group.js');
require('../input/input-group-props.js');
require('../input-number/props.js');
require('../_chunks/dep-6f95225b.js');
require('../input-number/hooks/useInputNumber.js');
require('../_chunks/dep-3688812b.js');
require('../_chunks/dep-77efa7eb.js');
require('../_chunks/dep-cf727d09.js');
require('../select/index.js');
require('../select/select.js');
require('@babel/runtime/helpers/asyncToGenerator');
require('@babel/runtime/regenerator');
require('../common-components/fake-arrow.js');
require('../select-input/index.js');
require('../select-input/select-input.js');
require('../popup/index.js');
require('../popup/popup.js');
require('@popperjs/core');
require('../popup/container.js');
require('../popup/props.js');
require('../_chunks/dep-d8ab5fb5.js');
require('../_chunks/dep-cde052f0.js');
require('../_chunks/dep-a3501e24.js');
require('../select-input/props.js');
require('../select-input/hooks/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/tag-input.js');
require('../tag-input/props.js');
require('../tag-input/hooks/useDragSorter.js');
require('../tag-input/hooks/useHover.js');
require('../tag-input/hooks/useTagScroll.js');
require('../tag-input/hooks/useTagList.js');
require('../tag/index.js');
require('../tag/tag.js');
require('tinycolor2');
require('../tag/props.js');
require('../tag/check-tag.js');
require('../tag/check-tag-props.js');
require('../_chunks/dep-1177d395.js');
require('../tag/check-tag-group.js');
require('../tag/check-tag-group-props.js');
require('../select-input/hooks/useOverlayInnerStyle.js');
require('../select-input/hooks/useSingle.js');
require('../_chunks/dep-a66df673.js');
require('../_chunks/dep-f3b67279.js');
require('../_chunks/dep-117093bb.js');
require('../select/components/select-panel.js');
require('../select/option.js');
require('../select/option-props.js');
require('../checkbox/index.js');
require('../checkbox/checkbox.js');
require('../checkbox/props.js');
require('../checkbox/constants/index.js');
require('../checkbox/hooks/useCheckboxLazyLoad.js');
require('../_chunks/dep-b4842612.js');
require('../checkbox/hooks/useKeyboardEvent.js');
require('../checkbox/group.js');
require('../checkbox/checkbox-group-props.js');
require('../_chunks/dep-e53cd0a5.js');
require('../_chunks/dep-2bfa66aa.js');
require('../_chunks/dep-c2cca927.js');
require('../_chunks/dep-5e1fbdf8.js');
require('../select/utils/index.js');
require('../select/constants/index.js');
require('../select/option-group.js');
require('../select/option-group-props.js');
require('../select/props.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../_chunks/dep-d3b20f18.js');
require('../_chunks/dep-3d9f41ae.js');
require('../select/hooks/useKeyboardControl.js');
require('../select/hooks/useSelectOptions.js');
require('../_chunks/dep-5c150838.js');
require('../input-adornment/index.js');
require('../input-adornment/input-adornment.js');
require('../input-adornment/props.js');
require('../pagination/props.js');
require('../pagination/hooks/useMoreAction.js');
require('../pagination/hooks/usePaginationClasses.js');
require('../pagination/pagination-mini.js');
require('../pagination/pagination-mini-props.js');
require('../tooltip/index.js');
require('../tooltip/tooltip.js');
require('../tooltip/props.js');
require('../tooltip/utils/index.js');
require('./hooks/useSelectRange.js');
require('./components/base/Table.js');
require('./components/base/Cell.js');
require('../time-picker/index.js');
require('../time-picker/time-picker.js');
require('../_chunks/dep-a7068a83.js');
require('../time-picker/panel/time-picker-panel.js');
require('../_chunks/dep-83ab8a79.js');
require('../time-picker/props.js');
require('../_chunks/dep-5fee93c0.js');
require('../_chunks/dep-e01a4fce.js');
require('../time-picker/time-range-picker.js');
require('../range-input/index.js');
require('../range-input/range-input.js');
require('../range-input/props.js');
require('../range-input/range-input-popup.js');
require('../range-input/range-input-popup-props.js');
require('../time-picker/time-range-picker-props.js');
require('./components/panel/ExtraContent.js');
require('./components/base/Footer.js');
require('./hooks/useDisableDate.js');
require('./hooks/useTableData.js');
require('@babel/runtime/helpers/objectDestructuringEmpty');
require('@babel/runtime/helpers/extends');
require('../_chunks/dep-0797927a.js');
require('../_chunks/dep-2f14e275.js');
require('../_chunks/dep-c8284192.js');
require('../_chunks/dep-c2f54d6f.js');
require('../_chunks/dep-e70e28c7.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _DateRangePickerPanel = Vue.defineComponent({
  name: "TDateRangePickerPanel",
  props: _objectSpread({
    value: datePicker_dateRangePickerProps["default"].value,
    defaultValue: datePicker_dateRangePickerProps["default"].defaultValue,
    modelValue: datePicker_dateRangePickerProps["default"].modelValue,
    disabled: datePicker_dateRangePickerProps["default"].disabled,
    disableDate: datePicker_dateRangePickerProps["default"].disableDate,
    enableTimePicker: datePicker_dateRangePickerProps["default"].enableTimePicker,
    firstDayOfWeek: datePicker_dateRangePickerProps["default"].firstDayOfWeek,
    format: datePicker_dateRangePickerProps["default"].format,
    mode: datePicker_dateRangePickerProps["default"].mode,
    presets: datePicker_dateRangePickerProps["default"].presets,
    presetsPlacement: datePicker_dateRangePickerProps["default"].presetsPlacement,
    timePickerProps: datePicker_dateRangePickerProps["default"].timePickerProps,
    panelPreselection: datePicker_dateRangePickerProps["default"].panelPreselection
  }, datePicker_dateRangePickerPanelProps["default"]),
  setup: function setup(props) {
    var _useRangeValue = datePicker_hooks_useRangeValue.useRangeValue(props),
      value = _useRangeValue.value,
      year = _useRangeValue.year,
      month = _useRangeValue.month,
      time = _useRangeValue.time,
      cacheValue = _useRangeValue.cacheValue,
      isFirstValueSelected = _useRangeValue.isFirstValueSelected,
      onChange = _useRangeValue.onChange;
    var formatRef = Vue.computed(function () {
      return format.getDefaultFormat({
        mode: props.mode,
        enableTimePicker: props.enableTimePicker,
        format: props.format
      });
    });
    var isSelected = Vue.ref(false);
    var isHoverCell = Vue.ref(false);
    var hoverValue = Vue.ref([]);
    var activeIndex = Vue.computed(function () {
      return isFirstValueSelected.value ? 1 : 0;
    });
    Vue.onMounted(function () {
      if (value.value.length === 2 && !props.enableTimePicker) {
        var nextMonth = value.value.map(function (v) {
          return format.parseToDayjs(v, formatRef.value.format).month();
        });
        year.value = value.value.map(function (v) {
          return format.parseToDayjs(v, formatRef.value.valueType).year();
        });
        if (year.value[0] === year.value[1] && nextMonth[0] === nextMonth[1]) {
          nextMonth[0] === 11 ? nextMonth[0] -= 1 : nextMonth[1] += 1;
        }
        month.value = nextMonth;
        if (["month", "quarter"].includes(props.mode) && year.value[0] === year.value[1]) {
          year.value = [year.value[0], year.value[0] + 1];
        }
      }
    });
    function onCellMouseEnter(date) {
      isHoverCell.value = true;
      var nextValue = _toConsumableArray__default["default"](hoverValue.value);
      nextValue[activeIndex.value] = format.formatDate(date, {
        format: formatRef.value.format
      });
      hoverValue.value = nextValue;
    }
    function onCellMouseLeave() {
      isHoverCell.value = false;
      hoverValue.value = cacheValue.value;
    }
    function onCellClick(date, _ref) {
      var _props$onCellClick;
      var e = _ref.e;
      isHoverCell.value = false;
      isSelected.value = true;
      var nextValue = _toConsumableArray__default["default"](cacheValue.value);
      nextValue[activeIndex.value] = format.formatDate(date, {
        format: formatRef.value.format
      });
      cacheValue.value = nextValue;
      (_props$onCellClick = props.onCellClick) === null || _props$onCellClick === void 0 || _props$onCellClick.call(props, {
        e: e,
        partial: activeIndex.value ? "end" : "start",
        date: nextValue.map(function (v) {
          return dayjs__default["default"](v).toDate();
        })
      });
      if (props.enableTimePicker) return;
      if (nextValue.length === 2 && isFirstValueSelected.value) {
        onChange === null || onChange === void 0 || onChange(format.formatDate(nextValue, {
          format: formatRef.value.format,
          autoSwap: true,
          defaultTime: props.defaultTime
        }), {
          dayjsValue: nextValue.map(function (v, i) {
            var _props$defaultTime;
            return format.parseToDayjs(v, formatRef.value.format, void 0, void 0, (_props$defaultTime = props.defaultTime) === null || _props$defaultTime === void 0 ? void 0 : _props$defaultTime[i]);
          }),
          trigger: "pick"
        });
        isFirstValueSelected.value = false;
      } else {
        isFirstValueSelected.value = true;
      }
    }
    function onJumperClick(_ref2) {
      var trigger = _ref2.trigger,
        partial = _ref2.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      var monthCountMap = {
        date: 1,
        week: 1,
        month: 12,
        quarter: 12,
        year: 120
      };
      var monthCount = monthCountMap[props.mode] || 0;
      var current = new Date(year.value[partialIndex], month.value[partialIndex]);
      var next = null;
      if (trigger === "prev") {
        next = utils.subtractMonth(current, monthCount);
      } else if (trigger === "current") {
        next = new Date();
      } else if (trigger === "next") {
        next = utils.addMonth(current, monthCount);
      }
      var nextYear = _toConsumableArray__default["default"](year.value);
      nextYear[partialIndex] = next.getFullYear();
      var nextMonth = _toConsumableArray__default["default"](month.value);
      nextMonth[partialIndex] = next.getMonth();
      var onlyYearSelect = ["year", "quarter", "month"].includes(props.mode);
      var correctedDate = datePicker_utils_index.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
      nextYear = correctedDate.nextYear;
      nextMonth = correctedDate.nextMonth;
      if (year.value.some(function (y) {
        return !nextYear.includes(y);
      })) {
        var _props$onYearChange;
        (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 || _props$onYearChange.call(props, {
          partial: partial,
          year: nextYear[partialIndex],
          date: value.value.map(function (v) {
            return dayjs__default["default"](v).toDate();
          }),
          trigger: trigger === "current" ? "today" : "year-".concat(datePicker_utils_index.triggerMap[trigger])
        });
      }
      if (month.value.some(function (m) {
        return !nextMonth.includes(m);
      })) {
        var _props$onMonthChange;
        (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 || _props$onMonthChange.call(props, {
          partial: partial,
          month: nextMonth[partialIndex],
          date: value.value.map(function (v) {
            return dayjs__default["default"](v).toDate();
          }),
          trigger: trigger === "current" ? "today" : "month-".concat(datePicker_utils_index.triggerMap[trigger])
        });
      }
      year.value = nextYear;
      month.value = nextMonth;
    }
    function onTimePickerChange(val) {
      var _props$onTimeChange;
      var _extractTimeObj = utils.extractTimeObj(val),
        hours = _extractTimeObj.hours,
        minutes = _extractTimeObj.minutes,
        seconds = _extractTimeObj.seconds,
        milliseconds = _extractTimeObj.milliseconds,
        meridiem = _extractTimeObj.meridiem;
      var nextInputValue = _toConsumableArray__default["default"](cacheValue.value);
      var changedInputValue = cacheValue.value[activeIndex.value];
      var currentDate = !dayjs__default["default"](changedInputValue, formatRef.value.format).isValid() ? dayjs__default["default"]().year(year.value[activeIndex.value]).month(month.value[activeIndex.value]) : dayjs__default["default"](changedInputValue, formatRef.value.format);
      var nextHours = hours;
      if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
      if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
      var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
      nextInputValue[activeIndex.value] = nextDate;
      var nextTime = _toConsumableArray__default["default"](time.value);
      nextTime[activeIndex.value] = val;
      time.value = nextTime;
      isSelected.value = true;
      cacheValue.value = format.formatDate(nextInputValue, {
        format: formatRef.value.format
      });
      (_props$onTimeChange = props.onTimeChange) === null || _props$onTimeChange === void 0 || _props$onTimeChange.call(props, {
        time: val,
        date: value.value.map(function (v) {
          return format.parseToDayjs(v, formatRef.value.format).toDate();
        }),
        partial: activeIndex.value ? "end" : "start",
        trigger: "time-hour"
      });
    }
    function onConfirmClick(_ref3) {
      var _props$onConfirm;
      var e = _ref3.e;
      var nextValue = _toConsumableArray__default["default"](cacheValue.value);
      if (nextValue.length === 2 && isFirstValueSelected.value) {
        onChange === null || onChange === void 0 || onChange(format.formatDate(nextValue, {
          format: formatRef.value.format,
          autoSwap: true,
          defaultTime: props.defaultTime
        }), {
          dayjsValue: nextValue.map(function (v, i) {
            var _props$defaultTime2;
            return format.parseToDayjs(v, formatRef.value.format, void 0, void 0, (_props$defaultTime2 = props.defaultTime) === null || _props$defaultTime2 === void 0 ? void 0 : _props$defaultTime2[i]);
          }),
          trigger: "confirm"
        });
        year.value = nextValue.map(function (v) {
          return dayjs__default["default"](v, formatRef.value.format).year();
        });
        month.value = nextValue.map(function (v) {
          return dayjs__default["default"](v, formatRef.value.format).month();
        });
        isFirstValueSelected.value = false;
      } else {
        isFirstValueSelected.value = true;
      }
      (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, {
        date: value.value.map(function (v) {
          return dayjs__default["default"](v).toDate();
        }),
        e: e
      });
    }
    function onPresetClick(preset, context) {
      var presetValue = preset;
      if (isFunction.isFunction(preset)) {
        presetValue = preset();
      }
      if (!isArray.isArray(presetValue)) {
        console.error("preset: ".concat(preset, " \u9884\u8BBE\u503C\u5FC5\u987B\u662F\u6570\u7EC4!"));
      } else {
        var _props$onPresetClick;
        onChange === null || onChange === void 0 || onChange(format.formatDate(presetValue, {
          format: formatRef.value.format,
          autoSwap: true,
          defaultTime: props.defaultTime
        }), {
          dayjsValue: presetValue.map(function (p, i) {
            var _props$defaultTime3;
            return format.parseToDayjs(p, formatRef.value.format, void 0, void 0, (_props$defaultTime3 = props.defaultTime) === null || _props$defaultTime3 === void 0 ? void 0 : _props$defaultTime3[i]);
          }),
          trigger: "preset"
        });
        (_props$onPresetClick = props.onPresetClick) === null || _props$onPresetClick === void 0 || _props$onPresetClick.call(props, context);
      }
    }
    function onYearChange(nextVal, _ref4) {
      var _props$onYearChange2;
      var partial = _ref4.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      if (props.enableTimePicker) partialIndex = activeIndex.value;
      var nextYear = _toConsumableArray__default["default"](year.value);
      nextYear[partialIndex] = nextVal;
      var nextMonth = _toConsumableArray__default["default"](month.value);
      var onlyYearSelect = ["year", "quarter", "month"].includes(props.mode);
      var correctedDate = datePicker_utils_index.dateCorrection(partialIndex, nextYear, nextMonth, onlyYearSelect);
      nextYear = correctedDate.nextYear;
      nextMonth = correctedDate.nextMonth;
      year.value = nextYear;
      if (!onlyYearSelect) month.value = nextMonth;
      (_props$onYearChange2 = props.onYearChange) === null || _props$onYearChange2 === void 0 || _props$onYearChange2.call(props, {
        partial: partial,
        year: nextYear[partialIndex],
        date: value.value.map(function (v) {
          return dayjs__default["default"](v).toDate();
        }),
        trigger: "year-select"
      });
    }
    function onMonthChange(nextVal, _ref5) {
      var _props$onMonthChange2;
      var partial = _ref5.partial;
      var partialIndex = partial === "start" ? 0 : 1;
      if (props.enableTimePicker) partialIndex = activeIndex.value;
      var nextMonth = _toConsumableArray__default["default"](month.value);
      nextMonth[partialIndex] = nextVal;
      if (year.value[0] === year.value[1]) {
        if (partialIndex === 0) {
          if (nextMonth[1] <= nextMonth[0]) {
            nextMonth[1] = nextMonth[0] + 1;
            if (nextMonth[1] === 12) {
              var _year$value, _year$value2;
              nextMonth[1] = 0;
              year.value = [(_year$value = year.value) === null || _year$value === void 0 ? void 0 : _year$value[0], ((_year$value2 = year.value) === null || _year$value2 === void 0 ? void 0 : _year$value2[1]) + 1];
            }
          }
        }
        if (partialIndex === 1) {
          nextMonth[0] = Math.min(nextMonth[0], nextMonth[1]);
          if (nextMonth[0] >= nextMonth[1]) {
            nextMonth[0] -= 1;
            if (nextMonth[0] === -1) {
              var _year$value3, _year$value4;
              nextMonth[0] = 11;
              year.value = [((_year$value3 = year.value) === null || _year$value3 === void 0 ? void 0 : _year$value3[0]) - 1, (_year$value4 = year.value) === null || _year$value4 === void 0 ? void 0 : _year$value4[1]];
            }
          }
        }
      }
      month.value = nextMonth;
      (_props$onMonthChange2 = props.onMonthChange) === null || _props$onMonthChange2 === void 0 || _props$onMonthChange2.call(props, {
        partial: partial,
        month: nextMonth[partialIndex],
        date: value.value.map(function (v) {
          return dayjs__default["default"](v).toDate();
        }),
        trigger: "month-select"
      });
    }
    var panelProps = Vue.computed(function () {
      return {
        hoverValue: isHoverCell.value ? hoverValue.value : [],
        value: isSelected.value ? cacheValue.value : value.value,
        activeIndex: activeIndex.value,
        year: year.value,
        month: month.value,
        mode: props.mode,
        format: formatRef.value.format,
        presets: props.presets,
        time: time.value,
        disableDate: props.disableDate,
        range: props.range,
        firstDayOfWeek: props.firstDayOfWeek,
        timePickerProps: props.timePickerProps,
        enableTimePicker: props.enableTimePicker,
        presetsPlacement: props.presetsPlacement,
        panelPreselection: props.panelPreselection,
        popupVisible: true,
        panelClick: props.onPanelClick,
        onCellClick: onCellClick,
        onCellMouseEnter: onCellMouseEnter,
        onCellMouseLeave: onCellMouseLeave,
        onJumperClick: onJumperClick,
        onConfirmClick: onConfirmClick,
        onPresetClick: onPresetClick,
        onYearChange: onYearChange,
        onMonthChange: onMonthChange,
        onTimePickerChange: onTimePickerChange
      };
    });
    return function () {
      return Vue.createVNode(datePicker_components_panel_RangePanel["default"], panelProps.value, null);
    };
  }
});

exports["default"] = _DateRangePickerPanel;
