// @vueup/vue-quill/dist/vue-quill.cjs.js
/*!
 * VueQuill @vueup/vue-quill v1.5.3
 * https://vueup.github.io/vue-quill/
 * 
 * Includes quill v2.0.2 || >=2.0.4 <3
 * https://quilljs.com/
 * 
 * Copyright (c) 2026 Ahmad Luthfi Masruri
 * Released under the MIT license
 * Date: 2026-06-03T00:25:15.762Z
 */
'use strict';

var Delta = require('quill-delta');
var vue = require('vue');

const toolbarOptions = {
    essential: [
        [{ header: [1, 2, 3, 4, 5, 6, false] }],
        ['bold', 'italic', 'underline'],
        [{ list: 'ordered' }, { list: 'bullet' }, { align: [] }],
        ['blockquote', 'code-block', 'link'],
        [{ color: [] }, 'clean'],
    ],
    minimal: [
        [{ header: 1 }, { header: 2 }],
        ['bold', 'italic', 'underline'],
        [{ list: 'ordered' }, { list: 'bullet' }, { align: [] }],
    ],
    full: [
        ['bold', 'italic', 'underline', 'strike'], // toggled buttons
        ['blockquote', 'code-block'],
        [{ header: 1 }, { header: 2 }], // custom button values
        [{ list: 'ordered' }, { list: 'bullet' }],
        [{ script: 'sub' }, { script: 'super' }], // superscript/subscript
        [{ indent: '-1' }, { indent: '+1' }], // outdent/indent
        [{ direction: 'rtl' }], // text direction
        [{ size: ['small', false, 'large', 'huge'] }], // custom dropdown
        [{ header: [1, 2, 3, 4, 5, 6, false] }],
        [{ color: [] }, { background: [] }], // dropdown with defaults from theme
        [{ font: [] }],
        [{ align: [] }],
        ['link', 'video', 'image'],
        ['clean'], // remove formatting button
    ],
};

let loadedQuill;
let quillLoadPromise;
let pendingRegistrations = [];
const isBrowser = () => typeof document !== 'undefined';
const replayPendingRegistrations = (Quill) => {
    const registrations = pendingRegistrations;
    pendingRegistrations = [];
    for (const args of registrations) {
        Quill.register(...args);
    }
};
const loadQuill = async () => {
    if (loadedQuill)
        return loadedQuill;
    if (!isBrowser()) {
        throw new Error('@vueup/vue-quill: Quill can only be loaded in a browser environment.');
    }
    quillLoadPromise !== null && quillLoadPromise !== void 0 ? quillLoadPromise : (quillLoadPromise = import('quill').then(({ default: Quill }) => {
        loadedQuill = Quill;
        replayPendingRegistrations(Quill);
        return Quill;
    }));
    return quillLoadPromise;
};
const getLoadedQuill = () => loadedQuill;
const getQuillOrThrow = () => {
    if (loadedQuill)
        return loadedQuill;
    throw new Error('@vueup/vue-quill: Quill is not loaded yet. Use loadQuill() in a browser-only lifecycle hook, or access the Quill instance from the ready event.');
};
const Quill = new Proxy(function QuillProxy() { }, {
    get(_target, property) {
        if (property === 'register') {
            return (...args) => {
                if (loadedQuill) {
                    return loadedQuill.register(...args);
                }
                pendingRegistrations.push(args);
                if (isBrowser())
                    void loadQuill();
            };
        }
        const Quill = getQuillOrThrow();
        const value = Quill[property];
        return typeof value === 'function' ? value.bind(Quill) : value;
    },
    construct(_target, args) {
        const Quill = getQuillOrThrow();
        return new Quill(...args);
    },
});

const quillRegistrationPathPrefixes = new Set([
    'attributors',
    'blots',
    'formats',
    'modules',
    'themes',
    'ui',
]);
const isQuillRegistrationPath = (name) => {
    const [prefix] = name.split('/');
    return quillRegistrationPathPrefixes.has(prefix);
};
const getModuleRegistrationName = (name) => {
    return isQuillRegistrationPath(name) ? name : `modules/${name}`;
};
const getModuleOptionName = (name) => {
    const registrationName = getModuleRegistrationName(name);
    const modulePrefix = 'modules/';
    return registrationName.startsWith(modulePrefix)
        ? registrationName.slice(modulePrefix.length)
        : undefined;
};
const globalRegistryPathPrefixes = new Set(['attributors', 'blots', 'formats']);
const usesParchmentRegistry = (name) => {
    const [prefix] = name.split('/');
    return globalRegistryPathPrefixes.has(prefix);
};
const getModuleOptions = (modules) => {
    var _a;
    const modulesOption = {};
    const registrations = Array.isArray(modules) ? modules : [modules];
    for (const module of registrations) {
        const optionName = getModuleOptionName(module.name);
        if (optionName)
            modulesOption[optionName] = (_a = module.options) !== null && _a !== void 0 ? _a : {};
    }
    return Object.keys(modulesOption).length > 0 ? modulesOption : undefined;
};
const registerModule = (Quill, module, registry) => {
    const moduleName = getModuleRegistrationName(module.name);
    const quillImports = Quill.imports;
    if (!quillImports || !(moduleName in quillImports)) {
        Quill.register(moduleName, module.module);
    }
    if (registry && usesParchmentRegistry(moduleName)) {
        registry.register(module.module);
    }
};

const QuillEditor = vue.defineComponent({
    name: 'QuillEditor',
    inheritAttrs: false,
    props: {
        content: {
            type: [String, Object],
        },
        contentType: {
            type: String,
            default: 'delta',
            validator: (value) => {
                return ['delta', 'html', 'text'].includes(value);
            },
        },
        enable: {
            type: Boolean,
            default: true,
        },
        readOnly: {
            type: Boolean,
            default: false,
        },
        placeholder: {
            type: String,
            required: false,
        },
        theme: {
            type: String,
            default: 'snow',
            validator: (value) => {
                return ['snow', 'bubble', ''].includes(value);
            },
        },
        toolbar: {
            type: [String, Array, Object, Boolean],
            required: false,
            default: undefined,
            validator: (value) => {
                if (typeof value === 'boolean')
                    return value === false;
                if (typeof value === 'string' && value !== '') {
                    return value.charAt(0) === '#'
                        ? true
                        : Object.keys(toolbarOptions).indexOf(value) !== -1;
                }
                return true;
            },
        },
        modules: {
            type: Object,
            required: false,
        },
        options: {
            type: Object,
            required: false,
        },
        globalOptions: {
            type: Object,
            required: false,
        },
    },
    emits: [
        'textChange',
        'selectionChange',
        'editorChange',
        'update:content',
        'focus',
        'blur',
        'ready',
    ],
    setup: (props, ctx) => {
        vue.onMounted(() => {
            isUnmounted = false;
            void initialize();
        });
        vue.onBeforeUnmount(() => {
            quill = null;
            isUnmounted = true;
        });
        let isUnmounted = false;
        let quill;
        let options;
        const editor = vue.ref();
        // Initialize Quill
        const initialize = async () => {
            var _a, _b;
            if (!editor.value)
                return;
            const editorElement = editor.value;
            const Quill = await loadQuill();
            if (isUnmounted || !editor.value || editor.value !== editorElement)
                return;
            options = composeOptions();
            // Register modules
            if (props.modules) {
                if (Array.isArray(props.modules)) {
                    for (const module of props.modules) {
                        registerModule(Quill, module, options.registry);
                    }
                }
                else {
                    registerModule(Quill, props.modules, options.registry);
                }
            }
            // Create new Quill instance
            quill = new Quill(editor.value, options);
            // Set editor content
            setContents(props.content);
            // Set event handlers
            quill.on('text-change', handleTextChange);
            quill.on('selection-change', handleSelectionChange);
            quill.on('editor-change', handleEditorChange);
            // Remove editor class when theme changes
            if (props.theme !== 'bubble')
                editor.value.classList.remove('ql-bubble');
            if (props.theme !== 'snow')
                editor.value.classList.remove('ql-snow');
            // Fix clicking the quill toolbar is detected as blur event
            (_b = (_a = getToolbarModule()) === null || _a === void 0 ? void 0 : _a.container) === null || _b === void 0 ? void 0 : _b.addEventListener('mousedown', (e) => {
                e.preventDefault();
            });
            // Emit ready event
            ctx.emit('ready', quill);
        };
        // Compose Options
        const composeOptions = () => {
            const clientOptions = {};
            if (props.theme !== '')
                clientOptions.theme = props.theme;
            if (props.readOnly)
                clientOptions.readOnly = props.readOnly;
            if (props.placeholder)
                clientOptions.placeholder = props.placeholder;
            if (props.toolbar === false) {
                clientOptions.modules = {
                    toolbar: false,
                };
            }
            else if (props.toolbar && props.toolbar !== '') {
                clientOptions.modules = {
                    toolbar: (() => {
                        if (typeof props.toolbar === 'object') {
                            return props.toolbar;
                        }
                        else if (typeof props.toolbar === 'string') {
                            const str = props.toolbar;
                            return str.charAt(0) === '#'
                                ? props.toolbar
                                : toolbarOptions[props.toolbar];
                        }
                        return;
                    })(),
                };
            }
            if (props.modules) {
                const modules = getModuleOptions(props.modules);
                if (modules) {
                    clientOptions.modules = Object.assign({}, clientOptions.modules, modules);
                }
            }
            return Object.assign({}, props.globalOptions, props.options, clientOptions);
        };
        const maybeClone = (delta) => {
            return typeof delta === 'object' && delta ? delta.slice() : delta;
        };
        const getToolbarModule = () => {
            return quill === null || quill === void 0 ? void 0 : quill.getModule('toolbar');
        };
        const deltaHasValuesOtherThanRetain = (delta) => {
            return Object.values(delta.ops).some((v) => !v.retain || Object.keys(v).length !== 1);
        };
        // Doesn't need reactivity, but does need to be cloned to avoid deep mutations always registering as equal
        let internalModel;
        const internalModelEquals = (against) => {
            if (typeof internalModel === typeof against) {
                if (against === internalModel) {
                    return true;
                }
                // Ref/Proxy does not support instanceof, so do a loose check
                if (typeof against === 'object' &&
                    against &&
                    typeof internalModel === 'object' &&
                    internalModel) {
                    return !deltaHasValuesOtherThanRetain(internalModel.diff(against));
                }
            }
            return false;
        };
        const handleTextChange = (delta, oldContents, source) => {
            internalModel = maybeClone(getContents());
            // Update v-model:content when text changes
            if (!internalModelEquals(props.content)) {
                ctx.emit('update:content', internalModel);
            }
            ctx.emit('textChange', { delta, oldContents, source });
        };
        const isEditorFocus = vue.ref();
        const handleSelectionChange = (range, oldRange, source) => {
            // Set isEditorFocus if quill.hasFocus()
            isEditorFocus.value = !!(quill === null || quill === void 0 ? void 0 : quill.hasFocus());
            ctx.emit('selectionChange', { range, oldRange, source });
        };
        vue.watch(isEditorFocus, (focus) => {
            // Focus and blur events document the editor ref as their payload.
            // eslint-disable-next-line vue/no-ref-as-operand
            if (focus)
                ctx.emit('focus', editor);
            // eslint-disable-next-line vue/no-ref-as-operand
            else
                ctx.emit('blur', editor);
        });
        const handleEditorChange = (...args) => {
            if (args[0] === 'text-change')
                ctx.emit('editorChange', {
                    name: args[0],
                    delta: args[1],
                    oldContents: args[2],
                    source: args[3],
                });
            if (args[0] === 'selection-change')
                ctx.emit('editorChange', {
                    name: args[0],
                    range: args[1],
                    oldRange: args[2],
                    source: args[3],
                });
        };
        const getEditor = () => {
            return editor.value;
        };
        const getToolbar = () => {
            var _a;
            return (_a = getToolbarModule()) === null || _a === void 0 ? void 0 : _a.container;
        };
        const getQuill = () => {
            if (quill)
                return quill;
            else
                throw `The quill editor hasn't been instantiated yet,
                  make sure to call this method when the editor ready
                  or use v-on:ready="onReady(quill)" event instead.`;
        };
        const getContents = (index, length) => {
            if (props.contentType === 'html') {
                return getHTML();
            }
            else if (props.contentType === 'text') {
                return getText(index, length);
            }
            return quill === null || quill === void 0 ? void 0 : quill.getContents(index, length);
        };
        const setContents = (content, source = 'api') => {
            const normalizedContent = !content
                ? props.contentType === 'delta'
                    ? new Delta()
                    : ''
                : content;
            if (props.contentType === 'html') {
                setHTML(normalizedContent);
            }
            else if (props.contentType === 'text') {
                setText(normalizedContent, source);
            }
            else {
                quill === null || quill === void 0 ? void 0 : quill.setContents(normalizedContent, source);
            }
            internalModel = maybeClone(normalizedContent);
        };
        const getText = (index, length) => {
            var _a;
            return (_a = quill === null || quill === void 0 ? void 0 : quill.getText(index, length)) !== null && _a !== void 0 ? _a : '';
        };
        const setText = (text, source = 'api') => {
            quill === null || quill === void 0 ? void 0 : quill.setText(text, source);
        };
        const getHTML = () => {
            var _a;
            return (_a = quill === null || quill === void 0 ? void 0 : quill.root.innerHTML) !== null && _a !== void 0 ? _a : '';
        };
        const setHTML = (html) => {
            if (quill)
                quill.root.innerHTML = html;
        };
        const pasteHTML = (html, source = 'api') => {
            const delta = quill === null || quill === void 0 ? void 0 : quill.clipboard.convert({ html });
            if (delta)
                quill === null || quill === void 0 ? void 0 : quill.setContents(delta, source);
        };
        const focus = () => {
            quill === null || quill === void 0 ? void 0 : quill.focus();
        };
        const reinit = () => {
            vue.nextTick(() => {
                var _a, _b;
                if (!ctx.slots.toolbar && quill)
                    (_b = (_a = getToolbarModule()) === null || _a === void 0 ? void 0 : _a.container) === null || _b === void 0 ? void 0 : _b.remove();
                void initialize();
            });
        };
        vue.watch(() => props.content, (newContent) => {
            if (!quill || !newContent || internalModelEquals(newContent))
                return;
            // Restore the selection and cursor position after updating the content
            const selection = quill.getSelection();
            if (selection) {
                vue.nextTick(() => quill === null || quill === void 0 ? void 0 : quill.setSelection(selection));
            }
            setContents(newContent);
        }, { deep: true });
        vue.watch(() => props.enable, (newValue) => {
            if (quill)
                quill.enable(newValue);
        });
        return {
            editor,
            getEditor,
            getToolbar,
            getQuill,
            getContents,
            setContents,
            getHTML,
            setHTML,
            pasteHTML,
            focus,
            getText,
            setText,
            reinit,
        };
    },
    render() {
        var _a, _b;
        return [
            (_b = (_a = this.$slots).toolbar) === null || _b === void 0 ? void 0 : _b.call(_a),
            vue.h('div', { ref: 'editor', ...this.$attrs }),
        ];
    },
});

exports.Delta = Delta;
exports.Quill = Quill;
exports.QuillEditor = QuillEditor;
exports.getLoadedQuill = getLoadedQuill;
exports.loadQuill = loadQuill;
