// tdesign-vue-next/cjs/tree-select/tree-select.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _asyncToGenerator = require('@babel/runtime/helpers/asyncToGenerator');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _regeneratorRuntime = require('@babel/runtime/regenerator');
var _typeof = require('@babel/runtime/helpers/typeof');
var tree_index = require('../tree/index.js');
var selectInput_index = require('../select-input/index.js');
var commonComponents_fakeArrow = require('../common-components/fake-arrow.js');
var treeSelect_props = require('./props.js');
require('../_chunks/dep-c7472ccd.js');
var index = require('../_chunks/dep-0e56b1da.js');
var index$1 = require('../_chunks/dep-b56e7ff5.js');
var index$5 = require('../_chunks/dep-86694091.js');
var index$2 = require('../_chunks/dep-a8de5365.js');
require('../_chunks/dep-d1a4892a.js');
var index$3 = require('../_chunks/dep-4725a846.js');
var index$4 = require('../_chunks/dep-d149369c.js');
var index$6 = require('../_chunks/dep-6f95225b.js');
var isFunction = require('../_chunks/dep-37af9215.js');
var isBoolean = require('../_chunks/dep-850526b5.js');
var isArray = require('../_chunks/dep-d1d374b4.js');
var isEmpty = require('../_chunks/dep-954dbe59.js');
var isNil = require('../_chunks/dep-e70e28c7.js');
var configProvider_hooks_useConfig = require('../config-provider/hooks/useConfig.js');
require('../tree/tree.js');
require('../tree/utils/adapt.js');
require('../_chunks/dep-742210e1.js');
require('../_chunks/dep-125f71ff.js');
require('../_chunks/dep-b292e00e.js');
require('../_chunks/dep-b4842612.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-e2205a6f.js');
require('../_chunks/dep-6f4a5c31.js');
require('../_chunks/dep-c3906cd5.js');
require('../_chunks/dep-6344c74f.js');
require('../_chunks/dep-ffe05cb1.js');
require('../_chunks/dep-cde052f0.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-06e53606.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-2e9fed7b.js');
require('../_chunks/dep-a3501e24.js');
require('tdesign-icons-vue-next');
require('../checkbox/index.js');
require('../checkbox/checkbox.js');
require('../checkbox/props.js');
require('../_chunks/dep-6c498eb2.js');
require('../_chunks/dep-a90757f2.js');
require('../_chunks/dep-fd403354.js');
require('../checkbox/constants/index.js');
require('../checkbox/hooks/useCheckboxLazyLoad.js');
require('../checkbox/hooks/useKeyboardEvent.js');
require('../_chunks/dep-1177d395.js');
require('../_chunks/dep-bf2e7501.js');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('dayjs');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-a324ad6a.js');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../checkbox/group.js');
require('../checkbox/checkbox-group-props.js');
require('../_chunks/dep-e53cd0a5.js');
require('../_chunks/dep-2bfa66aa.js');
require('../_chunks/dep-ea9fd5a2.js');
require('../_chunks/dep-c2cca927.js');
require('../_chunks/dep-5e1fbdf8.js');
require('../loading/index.js');
require('../_chunks/dep-548039eb.js');
require('../loading/plugin.js');
require('../_chunks/dep-4a5eca86.js');
require('../loading/icon/gradient.js');
require('../_chunks/dep-585d18dc.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-8308792f.js');
require('../_chunks/dep-d5e8d527.js');
require('../_chunks/dep-8e4f83bf.js');
require('../loading/props.js');
require('../_chunks/dep-f8f462e0.js');
require('../_chunks/dep-47af4fff.js');
require('../_chunks/dep-a8f427e5.js');
require('../_chunks/dep-9997364c.js');
require('../_chunks/dep-1e2d078f.js');
require('../_chunks/dep-7d888cd2.js');
require('../_chunks/dep-ffd72e3b.js');
require('../_chunks/dep-6ed4388e.js');
require('../_chunks/dep-0d190b0b.js');
require('../_chunks/dep-f7deb88a.js');
require('../_chunks/dep-e492b603.js');
require('../_chunks/dep-c8284192.js');
require('../_chunks/dep-0797927a.js');
require('../_chunks/dep-cf727d09.js');
require('../_chunks/dep-114eefa3.js');
require('../_chunks/dep-c2f54d6f.js');
require('../_chunks/dep-2f14e275.js');
require('../tree/hooks/useDragHandle.js');
require('../tree/hooks/useDraggable.js');
require('../tree/utils/index.js');
require('../tree/hooks/useTreeAction.js');
require('../tree/hooks/useTreeNodes.js');
require('../tree/tree-item.js');
require('../tree/hooks/useItemState.js');
require('../tree/hooks/useTreeItem.js');
require('../tree/hooks/useItemEvents.js');
require('../tree/hooks/useRenderIcon.js');
require('../tree/hooks/useRenderLabel.js');
require('../tree/hooks/useRenderLine.js');
require('../tree/hooks/useRenderOperations.js');
require('../tree/hooks/useTreeEvents.js');
require('../tree/hooks/useTreeScroll.js');
require('../tree/hooks/useTreeState.js');
require('../tree/hooks/useTreeStore.js');
require('../_chunks/dep-bd85d2ca.js');
require('mitt');
require('../_chunks/dep-a66df673.js');
require('../_chunks/dep-f3b67279.js');
require('../_chunks/dep-117093bb.js');
require('../tree/hooks/useTreeStyles.js');
require('../tree/props.js');
require('../select-input/select-input.js');
require('../popup/index.js');
require('../popup/popup.js');
require('@popperjs/core');
require('../popup/container.js');
require('../popup/props.js');
require('../_chunks/dep-d8ab5fb5.js');
require('../select-input/props.js');
require('../select-input/hooks/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/tag-input.js');
require('../input/index.js');
require('../input/input.js');
require('../input/props.js');
require('../input/hooks/useInput.js');
require('../form/constants/index.js');
require('../input/hooks/useLengthLimit.js');
require('../_chunks/dep-880baab0.js');
require('../input/hooks/useInputEventHandler.js');
require('../input/hooks/useInputWidth.js');
require('../input/input-group.js');
require('../input/input-group-props.js');
require('../tag-input/props.js');
require('../tag-input/hooks/useDragSorter.js');
require('../tag-input/hooks/useHover.js');
require('../tag-input/hooks/useTagScroll.js');
require('../tag-input/hooks/useTagList.js');
require('../tag/index.js');
require('../tag/tag.js');
require('tinycolor2');
require('../tag/props.js');
require('../tag/check-tag.js');
require('../tag/check-tag-props.js');
require('../tag/check-tag-group.js');
require('../tag/check-tag-group-props.js');
require('../select-input/hooks/useOverlayInnerStyle.js');
require('../select-input/hooks/useSingle.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _asyncToGenerator__default = /*#__PURE__*/_interopDefaultLegacy(_asyncToGenerator);
var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _regeneratorRuntime__default = /*#__PURE__*/_interopDefaultLegacy(_regeneratorRuntime);
var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);

function findParentValues(options, targetValue, realValue, realChildren) {
  var currentTargetValue = targetValue;
  if (currentTargetValue != null && _typeof__default["default"](currentTargetValue) === "object") {
    var _currentTargetValue;
    currentTargetValue = (_currentTargetValue = currentTargetValue) === null || _currentTargetValue === void 0 ? void 0 : _currentTargetValue[realValue];
  }
  if (currentTargetValue == null) return [];
  function findPath(nodes, parentPath) {
    var result = null;
    nodes.some(function (node) {
      var newPath = [].concat(_toConsumableArray__default["default"](parentPath), [node[realValue]]);
      if (node[realValue] === currentTargetValue) {
        result = parentPath;
        return true;
      }
      if (Array.isArray(node[realChildren]) && node[realChildren].length) {
        result = findPath(node[realChildren], newPath);
        if (result !== null) {
          return true;
        }
      }
      return false;
    });
    return result;
  }
  return findPath(options, []) || [];
}

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _TreeSelect = Vue.defineComponent({
  name: "TTreeSelect",
  props: treeSelect_props["default"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots,
      expose = _ref.expose;
    var renderTNodeJSX = index.useTNodeJSX();
    var renderDefaultTNode = index.useTNodeDefault();
    var classPrefix = index$1.usePrefixClass();
    var _useConfig = configProvider_hooks_useConfig.useConfig("treeSelect"),
      globalConfig = _useConfig.globalConfig;
    var formDisabled = index$2.useDisabled();
    var isReadonly = index$3.useReadonly();
    var treeRef = Vue.ref(null);
    var actived = Vue.ref([]);
    var expanded = Vue.ref([]);
    var nodeInfo = Vue.ref(null);
    var treeKey = Vue.ref(0);
    var _toRefs = Vue.toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue,
      popupVisible = _toRefs.popupVisible,
      inputValue = _toRefs.inputValue;
    var _useVModel = index$4.useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      treeSelectValue = _useVModel2[0],
      setTreeSelectValue = _useVModel2[1];
    var _useDefaultValue = index$5.useDefaultValue(popupVisible, false, props2.onPopupVisibleChange, "popupVisible"),
      _useDefaultValue2 = _slicedToArray__default["default"](_useDefaultValue, 2),
      innerVisible = _useDefaultValue2[0],
      setInnerVisible = _useDefaultValue2[1];
    var _useDefaultValue3 = index$5.useDefaultValue(inputValue, props2.defaultInputValue, props2.onInputChange, "inputValue"),
      _useDefaultValue4 = _slicedToArray__default["default"](_useDefaultValue3, 2),
      innerInputValue = _useDefaultValue4[0],
      setInnerInputValue = _useDefaultValue4[1];
    var setTreeParentsExpanded = function setTreeParentsExpanded() {
      var getParents = function getParents(value2) {
        return findParentValues(props2.data, value2, realValue.value, realChildren.value);
      };
      var treeParents = [];
      if (treeSelectValue.value) {
        if (Array.isArray(treeSelectValue.value) && props2.multiple) {
          treeParents = treeSelectValue.value.flatMap(getParents);
        } else if (!Array.isArray(treeSelectValue.value) && !props2.multiple) {
          treeParents = getParents(treeSelectValue.value);
        }
      }
      expanded.value = Array.from(/* @__PURE__ */new Set([].concat(_toConsumableArray__default["default"](expanded.value), _toConsumableArray__default["default"](treeParents))));
    };
    Vue.watch(treeSelectValue, /*#__PURE__*/_asyncToGenerator__default["default"](/*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee() {
      return _regeneratorRuntime__default["default"].wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.next = 1;
            return changeNodeInfo();
          case 1:
            if (!props2.multiple) {
              actived.value = nodeInfo.value ? [nodeInfo.value.value] : [];
            }
          case 2:
          case "end":
            return _context.stop();
        }
      }, _callee);
    })));
    Vue.watch(function () {
      return props2.data;
    }, /*#__PURE__*/_asyncToGenerator__default["default"](/*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee2() {
      return _regeneratorRuntime__default["default"].wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 1;
            return changeNodeInfo();
          case 1:
            treeRerender();
          case 2:
          case "end":
            return _context2.stop();
        }
      }, _callee2);
    })), {
      deep: true
    });
    var filterByText = Vue.computed(function () {
      var value2 = innerInputValue.value || "";
      if (value2 === "") {
        return null;
      }
      return function (node) {
        if (isFunction.isFunction(props2.filter)) {
          var filter = props2.filter(String(value2), node);
          if (isBoolean.isBoolean(filter)) {
            return filter;
          }
        }
        return node.data[realLabel.value].indexOf(value2) >= 0;
      };
    });
    var tDisabled = Vue.computed(function () {
      return formDisabled.value || props2.disabled;
    });
    var inputPlaceholder = Vue.computed(function () {
      var _nodeInfo$value;
      return innerVisible.value && ((_nodeInfo$value = nodeInfo.value) === null || _nodeInfo$value === void 0 ? void 0 : _nodeInfo$value.label) || props2.placeholder || globalConfig.value.placeholder;
    });
    var popupClass = Vue.computed(function () {
      return ["".concat(classPrefix.value, "-select__dropdown"), "narrow-scrollbar"];
    });
    var dropdownInnerSize = Vue.computed(function () {
      return {
        small: "s",
        medium: "m",
        large: "l"
      }[props2.size];
    });
    var isObjectValue = Vue.computed(function () {
      return props2.valueType === "object";
    });
    var checked = Vue.computed(function () {
      if (props2.multiple) {
        if (isObjectValue.value) {
          return isArray.isArray(treeSelectValue.value) ? treeSelectValue.value.map(function (item) {
            return item.value;
          }) : [];
        }
        return isArray.isArray(treeSelectValue.value) ? treeSelectValue.value.map(function (item) {
          return item;
        }) : [];
      }
      return [];
    });
    var multiLimitDisabled = Vue.computed(function () {
      return props2.multiple && !!props2.max && isArray.isArray(treeSelectValue.value) && props2.max <= treeSelectValue.value.length;
    });
    var realLabel = Vue.computed(function () {
      var _props2$keys;
      if (!isEmpty.isEmpty(props2.treeProps) && !isEmpty.isEmpty(props2.treeProps.keys)) {
        return props2.treeProps.keys.label || "label";
      }
      return ((_props2$keys = props2.keys) === null || _props2$keys === void 0 ? void 0 : _props2$keys.label) || "label";
    });
    var realValue = Vue.computed(function () {
      var _props2$keys2;
      if (!isEmpty.isEmpty(props2.treeProps) && !isEmpty.isEmpty(props2.treeProps.keys)) {
        return props2.treeProps.keys.value || "value";
      }
      return ((_props2$keys2 = props2.keys) === null || _props2$keys2 === void 0 ? void 0 : _props2$keys2.value) || "value";
    });
    var realChildren = Vue.computed(function () {
      var _props2$keys3;
      if (!isEmpty.isEmpty(props2.treeProps) && !isEmpty.isEmpty(props2.treeProps.keys)) {
        return props2.treeProps.keys.children || "children";
      }
      return ((_props2$keys3 = props2.keys) === null || _props2$keys3 === void 0 ? void 0 : _props2$keys3.children) || "children";
    });
    Vue.onMounted(/*#__PURE__*/_asyncToGenerator__default["default"](/*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee3() {
      var _treeSelectValue$valu;
      return _regeneratorRuntime__default["default"].wrap(function (_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            setTreeParentsExpanded();
            if (!treeSelectValue.value && props2.defaultValue) {
              change(props2.defaultValue, null, "uncheck");
            }
            if (isObjectValue.value) {
              actived.value = isArray.isArray(treeSelectValue.value) ? treeSelectValue.value.map(function (item) {
                return item === null || item === void 0 ? void 0 : item.value;
              }) : [(_treeSelectValue$valu = treeSelectValue.value) === null || _treeSelectValue$valu === void 0 ? void 0 : _treeSelectValue$valu.value];
            } else {
              actived.value = isArray.isArray(treeSelectValue.value) ? treeSelectValue.value : [treeSelectValue.value];
            }
            changeNodeInfo();
          case 1:
          case "end":
            return _context3.stop();
        }
      }, _callee3);
    })));
    var change = function change(valueParam, node, trigger) {
      setTreeSelectValue(valueParam, {
        node: node,
        trigger: trigger
      });
      changeNodeInfo();
    };
    var clear = function clear(content) {
      var _props2$onClear;
      var defaultValue = props2.multiple ? [] : "";
      actived.value = [];
      change(defaultValue, null, "clear");
      (_props2$onClear = props2.onClear) === null || _props2$onClear === void 0 || _props2$onClear.call(props2, {
        e: content.e
      });
    };
    var treeNodeChange = function treeNodeChange(valueParam, context) {
      var current = valueParam;
      if (isObjectValue.value) {
        current = valueParam.map(getNodeItem);
      }
      change(current, context.node, "check");
    };
    var treeNodeActive = function treeNodeActive(valueParam, context) {
      if (!props2.multiple) {
        setInnerVisible(false, context);
      }
      if (props2.multiple || !context.node.actived) {
        return;
      }
      if (treeSelectValue.value === context.node.data[realValue.value]) {
        return;
      }
      var current = valueParam;
      if (isObjectValue.value) {
        var nodeValue = isEmpty.isEmpty(valueParam) ? "" : valueParam[0];
        current = getNodeItem(nodeValue);
      } else {
        current = isEmpty.isEmpty(valueParam) ? "" : valueParam[0];
      }
      change(current, context.node, "check");
      actived.value = valueParam;
    };
    var treeNodeExpand = function treeNodeExpand(valueParam) {
      expanded.value = valueParam;
    };
    var treeNodeLoad = function treeNodeLoad() {
      changeNodeInfo();
    };
    var inputChange = function inputChange(value2) {
      var _props2$onSearch2;
      if (!innerVisible.value) {
        var _props2$onSearch;
        (_props2$onSearch = props2.onSearch) === null || _props2$onSearch === void 0 || _props2$onSearch.call(props2, String(value2));
        return;
      }
      setInnerInputValue(value2);
      (_props2$onSearch2 = props2.onSearch) === null || _props2$onSearch2 === void 0 || _props2$onSearch2.call(props2, String(value2));
    };
    var tagChange = function tagChange(value2, context) {
      var _props2$onRemove;
      var trigger = context.trigger,
        index = context.index;
      if (["tag-remove", "backspace"].includes(trigger)) {
        isArray.isArray(treeSelectValue.value) && treeSelectValue.value.splice(index, 1);
      }
      (_props2$onRemove = props2.onRemove) === null || _props2$onRemove === void 0 || _props2$onRemove.call(props2, {
        value: value2,
        data: null,
        e: context && context.e
      });
      change(treeSelectValue.value, null, trigger);
    };
    var handlePopupVisibleChange = function handlePopupVisibleChange(visible, context) {
      setInnerVisible(visible, context);
      if (visible && context.trigger === "trigger-element-click") setInnerInputValue("");
    };
    var changeNodeInfo = /*#__PURE__*/function () {
      var _ref5 = _asyncToGenerator__default["default"](/*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee4() {
        return _regeneratorRuntime__default["default"].wrap(function (_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              if (!props2.multiple) {
                if (treeSelectValue.value || treeSelectValue.value === 0) {
                  nodeInfo.value = getSingleNodeInfo();
                } else {
                  nodeInfo.value = "";
                }
              } else if (props2.multiple) {
                if (isArray.isArray(treeSelectValue.value)) {
                  nodeInfo.value = getMultipleNodeInfo();
                } else {
                  nodeInfo.value = [];
                }
              } else {
                nodeInfo.value = null;
              }
            case 1:
            case "end":
              return _context4.stop();
          }
        }, _callee4);
      }));
      return function changeNodeInfo() {
        return _ref5.apply(this, arguments);
      };
    }();
    var getSingleNodeInfo = function getSingleNodeInfo() {
      var nodeValue = isObjectValue.value ? treeSelectValue.value.value : treeSelectValue.value;
      return getNodeItem(nodeValue);
    };
    var getMultipleNodeInfo = function getMultipleNodeInfo() {
      return treeSelectValue.value.map(function (value2) {
        var nodeValue = isObjectValue.value ? value2.value : value2;
        return getNodeItem(nodeValue);
      });
    };
    var getNodeItem = function getNodeItem(targetValue) {
      if (treeRef.value) {
        var node2 = treeRef.value.getItem(targetValue);
        if (node2) {
          return _objectSpread(_objectSpread({}, node2.data), {}, {
            label: node2.data[realLabel.value],
            value: node2.data[realValue.value]
          });
        }
      }
      var node = _getTreeNode(props2.data, targetValue);
      if (node) {
        return node;
      }
      return {
        label: targetValue,
        value: targetValue
      };
    };
    var _getTreeNode = function getTreeNode(data, targetValue) {
      for (var i = 0, len = data.length; i < len; i++) {
        var item = data[i];
        if (item[realValue.value] === targetValue) {
          return _objectSpread(_objectSpread({}, item), {}, {
            label: item[realLabel.value],
            value: item[realValue.value]
          });
        }
        if (item !== null && item !== void 0 && item[realChildren.value]) {
          var result = _getTreeNode(item === null || item === void 0 ? void 0 : item[realChildren.value], targetValue);
          if (!isNil.isNil(result)) {
            return result;
          }
        }
      }
      return null;
    };
    var treeRerender = function treeRerender() {
      treeKey.value += 1;
    };
    var renderTree = function renderTree() {
      var treeEvents = index$6.useEventForward(props2.treeProps, {
        onChange: treeNodeChange,
        onActive: treeNodeActive,
        onExpand: treeNodeExpand,
        onLoad: treeNodeLoad
      });
      return Vue.withDirectives(Vue.createVNode(tree_index.Tree, Vue.mergeProps({
        "ref": treeRef,
        "key": treeKey.value,
        "value": _toConsumableArray__default["default"](checked.value),
        "hover": true,
        "keys": props2.keys,
        "data": props2.data,
        "activable": !props2.multiple,
        "checkable": props2.multiple,
        "disabled": tDisabled.value || multiLimitDisabled.value,
        "filter": filterByText.value,
        "actived": actived.value,
        "expanded": expanded.value,
        "activeMultiple": props2.multiple,
        "expandOnClickNode": false
      }, treeEvents.value), {
        empty: function empty() {
          return renderDefaultTNode("empty", {
            defaultNode: Vue.createVNode("div", {
              "class": "".concat(classPrefix.value, "-select__empty")
            }, [globalConfig.value.empty])
          });
        }
      }), [[Vue.vShow, !props2.loading]]);
    };
    var renderSuffixIcon = function renderSuffixIcon() {
      return Vue.createVNode(commonComponents_fakeArrow["default"], {
        "isActive": innerVisible.value,
        "overlayClassName": _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(classPrefix.value, "-fake-arrow--highlight"), innerVisible.value), "".concat(classPrefix.value, "-fake-arrow--disable"), props2.disabled)
      }, null);
    };
    expose({
      treeRef: treeRef
    });
    var selectInputEvent = index$6.useEventForward(props2.selectInputProps, {
      onInputChange: inputChange,
      onTagChange: tagChange,
      onPopupVisibleChange: handlePopupVisibleChange,
      onClear: clear,
      onBlur: function onBlur(_, context) {
        var _props2$onBlur;
        (_props2$onBlur = props2.onBlur) === null || _props2$onBlur === void 0 || _props2$onBlur.call(props2, {
          value: treeSelectValue.value,
          e: context.e
        });
      },
      onFocus: function onFocus(_, context) {
        var _props2$onFocus;
        (_props2$onFocus = props2.onFocus) === null || _props2$onFocus === void 0 || _props2$onFocus.call(props2, {
          value: treeSelectValue.value,
          e: context.e
        });
      }
    });
    return function () {
      return Vue.createVNode(selectInput_index.SelectInput, Vue.mergeProps({
        "class": "".concat(classPrefix.value, "-tree-select"),
        "value": nodeInfo.value,
        "inputValue": innerVisible.value ? innerInputValue.value : "",
        "popupVisible": innerVisible.value,
        "disabled": tDisabled.value,
        "multiple": props2.multiple,
        "loading": props2.loading,
        "clearable": props2.clearable,
        "autoWidth": props2.autoWidth,
        "borderless": props2.borderless,
        "readonly": isReadonly.value,
        "placeholder": inputPlaceholder.value,
        "allowInput": props2.filterable || isFunction.isFunction(props2.filter),
        "minCollapsedNum": props2.minCollapsedNum,
        "collapsed-items": props2.collapsedItems,
        "popupProps": _objectSpread({
          overlayClassName: popupClass.value
        }, props2.popupProps),
        "inputProps": _objectSpread({
          size: props2.size
        }, props2.inputProps),
        "tagInputProps": {
          size: props2.size
        },
        "tagProps": _objectSpread({
          maxWidth: 300
        }, props2.tagProps),
        "label": function label() {
          return renderTNodeJSX("label") || renderTNodeJSX("prefixIcon");
        },
        "suffix": props2.suffix,
        "suffixIcon": function suffixIcon() {
          if (props2.suffixIcon || slots.suffixIcon) {
            return renderTNodeJSX("suffixIcon");
          }
          return renderSuffixIcon();
        },
        "valueDisplay": function valueDisplay() {
          return renderTNodeJSX("valueDisplay", {
            params: props2.multiple ? {
              value: nodeInfo.value,
              onClose: function onClose(index) {
                var value2 = nodeInfo.value.map(function (node) {
                  return node.value;
                });
                tagChange(value2, {
                  trigger: "tag-remove",
                  index: index,
                  item: value2[index]
                });
              }
            } : {
              value: nodeInfo.value || _defineProperty__default["default"](_defineProperty__default["default"]({}, realLabel.value, ""), realValue.value, void 0)
            }
          });
        }
      }, selectInputEvent.value), {
        suffix: slots.suffix,
        panel: function panel() {
          return Vue.createVNode("div", {
            "class": ["".concat(classPrefix.value, "-select__dropdown-inner"), "".concat(classPrefix.value, "-select__dropdown-inner--size-").concat(dropdownInnerSize.value)]
          }, [renderTNodeJSX("panelTopContent"), Vue.withDirectives(Vue.createVNode("p", {
            "class": ["".concat(classPrefix.value, "-select-loading-tips"), "".concat(classPrefix.value, "-select__right-icon-polyfill")]
          }, [renderDefaultTNode("loadingText", {
            defaultNode: Vue.createVNode("div", {
              "class": "".concat(classPrefix.value, "-select__empty")
            }, [globalConfig.value.loadingText])
          })]), [[Vue.vShow, props2.loading && !tDisabled.value]]), renderTree(), renderTNodeJSX("panelBottomContent")]);
        },
        collapsedItems: slots.collapsedItems
      });
    };
  }
});

exports["default"] = _TreeSelect;
