// tdesign-vue-next/esm/menu/head-menu.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, getCurrentInstance, watchEffect, toRefs, ref, computed, reactive, provide, watch, onMounted, onUpdated, nextTick, onBeforeUnmount, h, createVNode, cloneVNode, isVNode } from 'vue';
import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { EllipsisIcon } from 'tdesign-icons-vue-next';
import { isArray, isFunction } from 'lodash-es';
import { l as log } from '../_chunks/dep-f2301d1c.js';
import props from './head-menu-props.js';
import { Tabs, TabPanel } from '../tabs/index.js';
import _Submenu from './submenu.js';
import PopupOverflowContent from './components/popup-overflow-content.js';
import { VMenu } from './utils/v-menu.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-651bdc55.js';
import { u as usePrefixClass } from '../_chunks/dep-59f77e78.js';
import { u as useDefaultValue } from '../_chunks/dep-c6879e67.js';
import '../_chunks/dep-204ae706.js';
import { u as useResizeObserver } from '../_chunks/dep-f5c5752f.js';
import { u as useVModel } from '../_chunks/dep-f32cc39c.js';
import '../tabs/tabs.js';
import '../tabs/tab-panel.js';
import '../tabs/tab-panel-props.js';
import '../_chunks/dep-eff3e90f.js';
import '../_chunks/dep-a7922f61.js';
import '../_chunks/dep-69502ff8.js';
import '../_chunks/dep-7daa5fe4.js';
import '../config-provider/hooks/useConfig.js';
import '../config-provider/utils/context.js';
import '../_chunks/dep-5c1cdd0b.js';
import '../_chunks/dep-76473034.js';
import 'dayjs';
import '../_chunks/dep-85e515d6.js';
import '../tabs/tab-nav.js';
import '../tabs/props.js';
import '../tabs/tab-nav-item.js';
import '../_chunks/dep-3d495891.js';
import '../_chunks/dep-16e13fa9.js';
import '../_chunks/dep-d15b7e74.js';
import '../tabs/tab-nav-bar.js';
import '../_chunks/dep-bc28556f.js';
import '../_chunks/dep-5dfdb698.js';
import 'tdesign-vue-next/esm/common/style/web/components/tabs/_index.less';
import '@babel/runtime/helpers/createClass';
import '@babel/runtime/helpers/classCallCheck';
import '../_chunks/dep-b8c8ffff.js';
import './submenu-props.js';
import '../common-components/fake-arrow.js';
import '../_chunks/dep-a9edb4d8.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../popup/index.js';
import '../popup/popup.js';
import '@popperjs/core';
import '../popup/container.js';
import '../popup/props.js';
import 'tdesign-vue-next/esm/common/style/web/components/popup/_index.less';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var MORE_SUBMENU_VALUE = "__t_head_menu_more__";
var _HeadMenu = defineComponent({
  name: "THeadMenu",
  props: props,
  setup: function setup(props2, ctx) {
    var classPrefix = usePrefixClass();
    var _getCurrentInstance = getCurrentInstance(),
      proxy = _getCurrentInstance.proxy;
    watchEffect(function () {
      if (ctx.slots.options) {
        log.warnOnce("TMenu", "`options` slot is going to be deprecated, please use `operations` for slot instead.");
      }
    });
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue,
      expanded = _toRefs.expanded;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      activeValue = _useVModel2[0],
      setActiveValue = _useVModel2[1];
    var _useDefaultValue = useDefaultValue(expanded, props2.defaultExpanded, props2.onExpand, "expanded"),
      _useDefaultValue2 = _slicedToArray(_useDefaultValue, 2),
      expandValues = _useDefaultValue2[0],
      setExpanded = _useDefaultValue2[1];
    var activeValues = ref([]);
    var theme = computed(function () {
      return props2.theme;
    });
    var menuClass = computed(function () {
      return ["".concat(classPrefix.value, "-menu"), "".concat(classPrefix.value, "-head-menu"), "".concat(classPrefix.value, "-menu--").concat(props2.theme)];
    });
    var mode = ref(props2.expandType);
    var submenu = reactive([]);
    var vMenu = new VMenu({
      isMutex: true,
      expandValues: expandValues.value
    });
    provide("TdMenu", {
      mode: mode,
      theme: theme,
      vMenu: vMenu,
      isHead: true,
      expandValues: expandValues,
      activeValue: activeValue,
      activeValues: activeValues,
      select: function select(value2) {
        setActiveValue(value2);
      },
      open: function open(value2, type) {
        var expanded2 = _toConsumableArray(expandValues.value);
        var index = expanded2.indexOf(value2);
        if (mode.value === "popup") {
          if (type === "add") {
            if (index === -1) {
              expanded2.push(value2);
            }
          } else if (type === "remove") {
            expanded2.splice(index, 1);
          }
        } else if (mode.value === "normal" && value2 !== void 0) {
          expanded2.splice(0, 1);
          if (index === -1) {
            expanded2.push(value2);
          }
        }
        setExpanded(expanded2);
      }
    });
    var handleTabChange = function handleTabChange(value2) {
      setActiveValue(value2);
      handleClickSubMenuItem(value2);
    };
    var handleSubmenuExpand = function handleSubmenuExpand(value2) {
      var ans = vMenu.getChild(value2);
      submenu.length = 0;
      submenu.push.apply(submenu, _toConsumableArray(ans));
    };
    watch(expandValues, function (value2) {
      if (mode.value === "normal") {
        handleSubmenuExpand(value2[0]);
      }
    });
    var cachedFlatItemValues = [];
    var updateActiveValues = function updateActiveValues(value2) {
      var pureVals = vMenu.select(value2).filter(function (v) {
        return v !== MORE_SUBMENU_VALUE;
      });
      activeValues.value = pureVals;
      syncMoreActiveState();
    };
    var syncMoreActiveState = function syncMoreActiveState() {
      var vals = activeValues.value.filter(function (v) {
        return v !== MORE_SUBMENU_VALUE;
      });
      var isFolded = foldStartIndex.value >= 0;
      var needActive = false;
      if (isFolded && vals.length > 0 && cachedFlatItemValues.length > 0) {
        var foldedTopValues = new Set(cachedFlatItemValues.slice(foldStartIndex.value));
        var topLevelAncestor = vals.find(function (v) {
          return v != null;
        });
        needActive = topLevelAncestor != null && foldedTopValues.has(topLevelAncestor);
      }
      if (needActive) {
        activeValues.value = [].concat(_toConsumableArray(vals), [MORE_SUBMENU_VALUE]);
      } else {
        activeValues.value = vals;
      }
    };
    watch(activeValue, updateActiveValues);
    watch(function () {
      return props2.expandType;
    }, function (value2) {
      mode.value = value2;
    });
    onMounted(function () {
      var _expandValues$value;
      activeValues.value = vMenu.select(activeValue.value);
      if (((_expandValues$value = expandValues.value) === null || _expandValues$value === void 0 ? void 0 : _expandValues$value.length) > 0) {
        handleSubmenuExpand(expandValues.value[0]);
      }
    });
    var handleClickSubMenuItem = function handleClickSubMenuItem(value2) {
      var _activeMenuItem$onCli;
      var activeMenuItem = submenu.find(function (v) {
        return v.value === value2;
      });
      (_activeMenuItem$onCli = activeMenuItem.onClick) === null || _activeMenuItem$onCli === void 0 || _activeMenuItem$onCli.call(activeMenuItem, {
        value: value2
      });
      var to = activeMenuItem.to,
        href = activeMenuItem.href,
        replace = activeMenuItem.replace;
      if (href) {
        window.location.href = activeMenuItem.href;
      }
      var router = activeMenuItem.router || proxy.$router;
      if (to && router) {
        replace ? router.replace(to) : router.push(to);
      }
    };
    var renderNormalSubmenu = function renderNormalSubmenu() {
      var _slot;
      if (submenu.length === 0) return null;
      return createVNode("ul", {
        "class": ["".concat(classPrefix.value, "-head-menu__submenu"), "".concat(classPrefix.value, "-submenu")]
      }, [createVNode(Tabs, {
        "value": activeValue.value,
        "onChange": handleTabChange
      }, _isSlot(_slot = submenu.map(function (item) {
        var _item$vnode$;
        return createVNode(TabPanel, {
          "value": item.value,
          "label": (_item$vnode$ = item.vnode()[0]) === null || _item$vnode$ === void 0 ? void 0 : _item$vnode$.children
        }, null);
      })) ? _slot : {
        "default": function _default() {
          return [_slot];
        }
      })]);
    };
    var menuRef = ref();
    var innerRef = ref();
    var logoRef = ref();
    var operationRef = ref();
    var foldStartIndex = ref(-1);
    var cachedItemWidths = [];
    var getMenuItemElements = function getMenuItemElements() {
      if (!menuRef.value) return [];
      var moreClass = "".concat(classPrefix.value, "-head-menu__submenu--more");
      var menuItemClass = "".concat(classPrefix.value, "-menu__item");
      var submenuClass = "".concat(classPrefix.value, "-submenu");
      var result = [];
      var _collectItems = function collectItems(parent, depth) {
        if (depth > 3) return;
        Array.from(parent.children).forEach(function (el) {
          if (!(el instanceof HTMLElement)) return;
          if (el.classList.contains(moreClass)) return;
          if (el.classList.contains(menuItemClass) || el.classList.contains(submenuClass)) {
            result.push(el);
          } else {
            _collectItems(el, depth + 1);
          }
        });
      };
      _collectItems(menuRef.value, 0);
      return result;
    };
    var getMoreButtonWidth = function getMoreButtonWidth() {
      if (!menuRef.value) return 0;
      var moreEl = menuRef.value.querySelector(".".concat(classPrefix.value, "-head-menu__submenu--more"));
      if (!moreEl) return 0;
      var wasHidden = moreEl.style.display === "none";
      if (wasHidden) {
        moreEl.style.visibility = "hidden";
        moreEl.style.display = "";
      }
      var style = window.getComputedStyle(moreEl);
      var rect = moreEl.getBoundingClientRect();
      var w = rect.width + Number.parseFloat(style.marginLeft) + Number.parseFloat(style.marginRight);
      if (wasHidden) {
        moreEl.style.display = "none";
        moreEl.style.visibility = "";
      }
      return w;
    };
    var getElementWidth = function getElementWidth(el) {
      var style = window.getComputedStyle(el);
      var rect = el.getBoundingClientRect();
      return rect.width + Number.parseFloat(style.marginLeft) + Number.parseFloat(style.marginRight);
    };
    var getWrapperElements = function getWrapperElements() {
      if (!menuRef.value) return [];
      var moreClass = "".concat(classPrefix.value, "-head-menu__submenu--more");
      var menuItemClass = "".concat(classPrefix.value, "-menu__item");
      var submenuClass = "".concat(classPrefix.value, "-submenu");
      var wrappers = [];
      Array.from(menuRef.value.children).forEach(function (el) {
        if (!(el instanceof HTMLElement)) return;
        if (el.classList.contains(moreClass)) return;
        if (el.classList.contains(menuItemClass) || el.classList.contains(submenuClass)) return;
        wrappers.push(el);
      });
      return wrappers;
    };
    var handleResize = function handleResize() {
      if (props2.expandType !== "popup") return;
      if (!menuRef.value) return;
      var wrappers = getWrapperElements();
      var savedWrapperDisplays = [];
      wrappers.forEach(function (el) {
        savedWrapperDisplays.push(el.style.display);
        el.style.display = "contents";
      });
      var itemNodes = getMenuItemElements();
      if (itemNodes.length === 0) {
        wrappers.forEach(function (el, i) {
          el.style.display = savedWrapperDisplays[i];
        });
        foldStartIndex.value = -1;
        applyFoldState();
        return;
      }
      var moreEl = menuRef.value.querySelector(".".concat(classPrefix.value, "-head-menu__submenu--more"));
      if (moreEl) moreEl.style.display = "none";
      var savedFlexShrinks = [];
      var savedDisplays = [];
      itemNodes.forEach(function (el) {
        savedDisplays.push(el.style.display);
        el.style.display = "";
        savedFlexShrinks.push(el.style.flexShrink);
        el.style.flexShrink = "0";
      });
      cachedItemWidths.length = 0;
      itemNodes.forEach(function (el) {
        cachedItemWidths.push(getElementWidth(el));
      });
      itemNodes.forEach(function (el, index) {
        el.style.flexShrink = savedFlexShrinks[index];
        el.style.display = savedDisplays[index];
      });
      if (moreEl) moreEl.style.display = savedDisplays.length > 0 ? "" : "none";
      wrappers.forEach(function (el) {
        el.style.display = "contents";
      });
      var menuWidth = calcMenuWidth();
      var totalItemsWidth = cachedItemWidths.reduce(function (sum, w) {
        return sum + w;
      }, 0);
      var newFoldIndex = -1;
      if (totalItemsWidth > menuWidth) {
        var subMoreWidth = getMoreButtonWidth();
        var currentWidth = 0;
        for (var i = 0; i < itemNodes.length; i++) {
          if (currentWidth + cachedItemWidths[i] + subMoreWidth > menuWidth) {
            newFoldIndex = i;
            break;
          }
          currentWidth += cachedItemWidths[i];
        }
        if (newFoldIndex === -1 && currentWidth + subMoreWidth > menuWidth) {
          newFoldIndex = itemNodes.length - 1;
        }
      }
      foldStartIndex.value = newFoldIndex;
      applyFoldState();
    };
    var applyFoldState = function applyFoldState() {
      if (!menuRef.value) return;
      var wrappers = getWrapperElements();
      wrappers.forEach(function (el) {
        el.style.display = "contents";
      });
      var itemNodes = getMenuItemElements();
      var moreEl = menuRef.value.querySelector(".".concat(classPrefix.value, "-head-menu__submenu--more"));
      var isFolded = foldStartIndex.value >= 0;
      itemNodes.forEach(function (el, index) {
        el.style.display = isFolded && index >= foldStartIndex.value ? "none" : "";
      });
      if (moreEl) {
        moreEl.style.display = isFolded ? "" : "none";
      }
      syncMoreActiveState();
    };
    useResizeObserver(innerRef, handleResize);
    watch(logoRef, function (el) {
      if (el) {
        var imgs = el.querySelectorAll("img");
        imgs.forEach(function (img) {
          if (!img.complete) {
            img.onload = function () {
              return handleResize();
            };
          }
        });
      }
    });
    var getComputedCss = function getComputedCss(el, cssProperty) {
      var _getComputedStyle$css;
      return (_getComputedStyle$css = getComputedStyle(el)[cssProperty]) !== null && _getComputedStyle$css !== void 0 ? _getComputedStyle$css : "";
    };
    var getComputedCssValue = function getComputedCssValue(el, cssProperty) {
      return Number.parseInt(String(getComputedCss(el, cssProperty)), 10);
    };
    var calcMenuWidth = function calcMenuWidth() {
      if (!innerRef.value || !menuRef.value) return 0;
      var menuPaddingLeft = getComputedCssValue(menuRef.value, "paddingLeft");
      var menuPaddingRight = getComputedCssValue(menuRef.value, "paddingRight");
      var menuMarginLeft = getComputedCssValue(menuRef.value, "marginLeft");
      var menuMarginRight = getComputedCssValue(menuRef.value, "marginRight");
      var totalWidth = innerRef.value.clientWidth;
      if (logoRef.value) {
        var logoMarginLeft = getComputedCssValue(logoRef.value, "marginLeft");
        var logoMarginRight = getComputedCssValue(logoRef.value, "marginRight");
        totalWidth = totalWidth - logoRef.value.offsetWidth - logoMarginLeft - logoMarginRight;
      }
      if (operationRef.value) {
        var operationMarginLeft = getComputedCssValue(operationRef.value, "marginLeft");
        var operationMarginRight = getComputedCssValue(operationRef.value, "marginRight");
        totalWidth = totalWidth - operationRef.value.offsetWidth - operationMarginLeft - operationMarginRight;
      }
      return totalWidth - menuPaddingLeft - menuPaddingRight - menuMarginLeft - menuMarginRight;
    };
    var _flattenSlotNodes = function flattenSlotNodes(nodes) {
      var result = [];
      var _iterator = _createForOfIteratorHelper(nodes),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var _node$type, _node$type2, _node$children;
          var node = _step.value;
          if (node.type === Symbol["for"]("v-fgm") || node.type === Symbol["for"]("")) {
            if (isArray(node.children)) {
              result.push.apply(result, _toConsumableArray(_flattenSlotNodes(node.children)));
            }
          } else if (((_node$type = node.type) === null || _node$type === void 0 ? void 0 : _node$type.name) === "TSubmenu" || ((_node$type2 = node.type) === null || _node$type2 === void 0 ? void 0 : _node$type2.name) === "TMenuItem") {
            result.push(node);
          } else if (isArray(node.children)) {
            result.push.apply(result, _toConsumableArray(_flattenSlotNodes(node.children)));
          } else if (isFunction((_node$children = node.children) === null || _node$children === void 0 ? void 0 : _node$children["default"])) {
            result.push.apply(result, _toConsumableArray(_flattenSlotNodes(node.children["default"]())));
          } else {
            result.push(node);
          }
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
      return result;
    };
    var _initVMenu = function initVMenu(slots, parentValue) {
      slots.forEach(function (node) {
        var _node$props, _node$type3, _node$type4, _node$children2;
        var nodeValue = (_node$props = node.props) === null || _node$props === void 0 ? void 0 : _node$props.value;
        if (((_node$type3 = node.type) === null || _node$type3 === void 0 ? void 0 : _node$type3.name) === "TSubmenu" || ((_node$type4 = node.type) === null || _node$type4 === void 0 ? void 0 : _node$type4.name) === "TMenuItem") {
          vMenu.add(_objectSpread({
            value: nodeValue,
            parent: parentValue,
            vnode: node.children["default"]
          }, node.props));
        }
        if (isFunction((_node$children2 = node.children) === null || _node$children2 === void 0 ? void 0 : _node$children2["default"])) {
          _initVMenu(node.children["default"](), nodeValue);
          return;
        }
        if (isArray(node.children)) {
          _initVMenu(node.children, nodeValue);
        }
      });
    };
    var pendingResizeRAF = null;
    var prevSlotKey = "";
    onUpdated(function () {
      applyFoldState();
    });
    var scheduleResize = function scheduleResize() {
      if (pendingResizeRAF != null) return;
      pendingResizeRAF = requestAnimationFrame(function () {
        pendingResizeRAF = null;
        handleResize();
      });
    };
    onMounted(function () {
      nextTick(function () {
        return handleResize();
      });
    });
    onBeforeUnmount(function () {
      if (pendingResizeRAF != null) {
        cancelAnimationFrame(pendingResizeRAF);
        pendingResizeRAF = null;
      }
    });
    return function () {
      var _props2$logo, _ctx$slots$logo, _ctx$slots, _props2$operations, _ctx$slots$operations, _ctx$slots2, _ctx$slots$options, _ctx$slots3, _ctx$slots$default, _ctx$slots4, _ctx$slots$content, _ctx$slots5;
      var logo = ((_props2$logo = props2.logo) === null || _props2$logo === void 0 ? void 0 : _props2$logo.call(props2, h)) || ((_ctx$slots$logo = (_ctx$slots = ctx.slots).logo) === null || _ctx$slots$logo === void 0 ? void 0 : _ctx$slots$logo.call(_ctx$slots));
      var operations = ((_props2$operations = props2.operations) === null || _props2$operations === void 0 ? void 0 : _props2$operations.call(props2, h)) || ((_ctx$slots$operations = (_ctx$slots2 = ctx.slots).operations) === null || _ctx$slots$operations === void 0 ? void 0 : _ctx$slots$operations.call(_ctx$slots2)) || ((_ctx$slots$options = (_ctx$slots3 = ctx.slots).options) === null || _ctx$slots$options === void 0 ? void 0 : _ctx$slots$options.call(_ctx$slots3));
      var originalContent = ((_ctx$slots$default = (_ctx$slots4 = ctx.slots)["default"]) === null || _ctx$slots$default === void 0 ? void 0 : _ctx$slots$default.call(_ctx$slots4)) || ((_ctx$slots$content = (_ctx$slots5 = ctx.slots).content) === null || _ctx$slots$content === void 0 ? void 0 : _ctx$slots$content.call(_ctx$slots5)) || [];
      var flatItems = _flattenSlotNodes(originalContent);
      var isFolded = foldStartIndex.value >= 0;
      var canSliceVNodes = isFolded && foldStartIndex.value < flatItems.length;
      var content = originalContent;
      var slotKey = flatItems.map(function (n) {
        var _n$props;
        return (_n$props = n.props) === null || _n$props === void 0 ? void 0 : _n$props.value;
      }).join(",");
      if (slotKey !== prevSlotKey) {
        prevSlotKey = slotKey;
        cachedFlatItemValues = flatItems.map(function (n) {
          var _n$props2;
          return (_n$props2 = n.props) === null || _n$props2 === void 0 ? void 0 : _n$props2.value;
        });
        vMenu.data.children = [];
        vMenu.cache.clear();
        _initVMenu(flatItems);
        scheduleResize();
      }
      var buildPopupSlot = function buildPopupSlot() {
        var _slot2;
        if (!isFolded) return [];
        if (canSliceVNodes) {
          return flatItems.slice(foldStartIndex.value).map(function (vnode) {
            return cloneVNode(vnode);
          });
        }
        return createVNode(PopupOverflowContent, {
          "foldIndex": foldStartIndex.value
        }, _isSlot(_slot2 = originalContent.map(function (vnode) {
          return cloneVNode(vnode);
        })) ? _slot2 : {
          "default": function _default() {
            return [_slot2];
          }
        });
      };
      return createVNode("div", {
        "class": menuClass.value
      }, [createVNode("div", {
        "class": "".concat(classPrefix.value, "-head-menu__inner"),
        "ref": innerRef
      }, [logo && createVNode("div", {
        "class": "".concat(classPrefix.value, "-menu__logo"),
        "ref": logoRef
      }, [logo]), createVNode("ul", {
        "class": "".concat(classPrefix.value, "-menu"),
        "ref": menuRef
      }, [content, props2.expandType === "popup" && createVNode(_Submenu, {
        "class": "".concat(classPrefix.value, "-head-menu__submenu--more"),
        "value": MORE_SUBMENU_VALUE,
        "expandType": "popup",
        "title": function title() {
          return createVNode(EllipsisIcon, null, null);
        },
        "style": {
          display: isFolded ? "" : "none"
        }
      }, {
        "default": buildPopupSlot
      })]), operations && createVNode("div", {
        "class": "".concat(classPrefix.value, "-menu__operations"),
        "ref": operationRef
      }, [operations])]), mode.value === "normal" && renderNormalSubmenu()]);
    };
  }
});

export { _HeadMenu as default };
