// tdesign-vue-next/cjs/date-picker/DatePicker.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var dayjs = require('dayjs');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-c7472ccd.js');
var index$2 = require('../_chunks/dep-0e56b1da.js');
var index = require('../_chunks/dep-b56e7ff5.js');
var index$1 = require('../_chunks/dep-a8de5365.js');
var index$4 = require('../_chunks/dep-125f71ff.js');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-d1a4892a.js');
var index$3 = require('../_chunks/dep-4725a846.js');
require('@babel/runtime/helpers/defineProperty');
var index$5 = require('../_chunks/dep-6f95225b.js');
var datePicker_hooks_useSingle = require('./hooks/useSingle.js');
var format = require('../_chunks/dep-d28d8b91.js');
var utils = require('../_chunks/dep-41271801.js');
var datePicker_props = require('./props.js');
var selectInput_index = require('../select-input/index.js');
var datePicker_components_panel_SinglePanel = require('./components/panel/SinglePanel.js');
var datePicker_utils_index = require('./utils/index.js');
var _baseGetTag = require('../_chunks/dep-e2205a6f.js');
var isObjectLike = require('../_chunks/dep-6f4a5c31.js');
var _baseUnary = require('../_chunks/dep-2e9fed7b.js');
var _overArg = require('../_chunks/dep-a324ad6a.js');
var isArray = require('../_chunks/dep-d1d374b4.js');
var isFunction = require('../_chunks/dep-37af9215.js');
var configProvider_hooks_useConfig = require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-bf2e7501.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../_chunks/dep-850526b5.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-06e53606.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('./hooks/useSingleValue.js');
require('../_chunks/dep-d149369c.js');
require('../_chunks/dep-d3b20f18.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('../_chunks/dep-7d888cd2.js');
require('../_chunks/dep-3d9f41ae.js');
require('../_chunks/dep-117093bb.js');
require('../_chunks/dep-a7068a83.js');
require('../_chunks/dep-f7deb88a.js');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-cf727d09.js');
require('@babel/runtime/helpers/objectDestructuringEmpty');
require('@babel/runtime/helpers/extends');
require('../_chunks/dep-e01a4fce.js');
require('../_chunks/dep-0797927a.js');
require('../_chunks/dep-2f14e275.js');
require('../_chunks/dep-c8284192.js');
require('../_chunks/dep-c2f54d6f.js');
require('../select-input/select-input.js');
require('../popup/index.js');
require('../popup/popup.js');
require('@popperjs/core');
require('../_chunks/dep-6c498eb2.js');
require('../_chunks/dep-fd403354.js');
require('../popup/container.js');
require('../popup/props.js');
require('../_chunks/dep-6344c74f.js');
require('../_chunks/dep-d8ab5fb5.js');
require('../_chunks/dep-cde052f0.js');
require('../_chunks/dep-a3501e24.js');
require('../_chunks/dep-742210e1.js');
require('../select-input/props.js');
require('../select-input/hooks/useMultiple.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../tag-input/index.js');
require('../tag-input/tag-input.js');
require('../input/index.js');
require('../input/input.js');
require('../input/props.js');
require('../input/hooks/useInput.js');
require('../form/constants/index.js');
require('../input/hooks/useLengthLimit.js');
require('../_chunks/dep-585d18dc.js');
require('../_chunks/dep-8308792f.js');
require('../_chunks/dep-d5e8d527.js');
require('../_chunks/dep-8e4f83bf.js');
require('../_chunks/dep-880baab0.js');
require('../input/hooks/useInputEventHandler.js');
require('../input/hooks/useInputWidth.js');
require('../input/input-group.js');
require('../input/input-group-props.js');
require('../tag-input/props.js');
require('../_chunks/dep-86694091.js');
require('../tag-input/hooks/useDragSorter.js');
require('../tag-input/hooks/useHover.js');
require('../tag-input/hooks/useTagScroll.js');
require('../tag-input/hooks/useTagList.js');
require('../tag/index.js');
require('../tag/tag.js');
require('tinycolor2');
require('../tag/props.js');
require('../tag/check-tag.js');
require('../tag/check-tag-props.js');
require('../_chunks/dep-1177d395.js');
require('../tag/check-tag-group.js');
require('../tag/check-tag-group-props.js');
require('../loading/index.js');
require('../_chunks/dep-548039eb.js');
require('../loading/plugin.js');
require('../_chunks/dep-4a5eca86.js');
require('../loading/icon/gradient.js');
require('../loading/props.js');
require('../_chunks/dep-f8f462e0.js');
require('../_chunks/dep-47af4fff.js');
require('../_chunks/dep-ea9fd5a2.js');
require('../_chunks/dep-a8f427e5.js');
require('../_chunks/dep-9997364c.js');
require('../_chunks/dep-1e2d078f.js');
require('../_chunks/dep-ffd72e3b.js');
require('../_chunks/dep-6ed4388e.js');
require('../select-input/hooks/useOverlayInnerStyle.js');
require('../select-input/hooks/useSingle.js');
require('../_chunks/dep-a66df673.js');
require('../_chunks/dep-f3b67279.js');
require('./components/panel/PanelContent.js');
require('./components/base/Header.js');
require('../pagination/index.js');
require('../pagination/pagination.js');
require('../input-number/index.js');
require('../input-number/input-number.js');
require('../button/index.js');
require('../button/button.js');
require('../button/props.js');
require('../_chunks/dep-a90757f2.js');
require('../input-number/props.js');
require('../input-number/hooks/useInputNumber.js');
require('../_chunks/dep-3688812b.js');
require('../_chunks/dep-77efa7eb.js');
require('../select/index.js');
require('../select/select.js');
require('@babel/runtime/helpers/asyncToGenerator');
require('@babel/runtime/regenerator');
require('../common-components/fake-arrow.js');
require('../select/components/select-panel.js');
require('../select/option.js');
require('../select/option-props.js');
require('../checkbox/index.js');
require('../checkbox/checkbox.js');
require('../checkbox/props.js');
require('../checkbox/constants/index.js');
require('../checkbox/hooks/useCheckboxLazyLoad.js');
require('../_chunks/dep-b4842612.js');
require('../checkbox/hooks/useKeyboardEvent.js');
require('../checkbox/group.js');
require('../checkbox/checkbox-group-props.js');
require('../_chunks/dep-e53cd0a5.js');
require('../_chunks/dep-2bfa66aa.js');
require('../_chunks/dep-c2cca927.js');
require('../_chunks/dep-5e1fbdf8.js');
require('../select/utils/index.js');
require('../select/constants/index.js');
require('../select/option-group.js');
require('../select/option-group-props.js');
require('../select/props.js');
require('../select/hooks/usePanelVirtualScroll.js');
require('../select/hooks/useKeyboardControl.js');
require('../select/hooks/useSelectOptions.js');
require('../_chunks/dep-5c150838.js');
require('../input-adornment/index.js');
require('../input-adornment/input-adornment.js');
require('../input-adornment/props.js');
require('../pagination/props.js');
require('../pagination/hooks/useMoreAction.js');
require('../pagination/hooks/usePaginationClasses.js');
require('../pagination/pagination-mini.js');
require('../pagination/pagination-mini-props.js');
require('../tooltip/index.js');
require('../tooltip/tooltip.js');
require('../tooltip/props.js');
require('../tooltip/utils/index.js');
require('./hooks/useSelectRange.js');
require('./components/base/Table.js');
require('./components/base/Cell.js');
require('../time-picker/index.js');
require('../time-picker/time-picker.js');
require('../time-picker/panel/time-picker-panel.js');
require('../_chunks/dep-83ab8a79.js');
require('../time-picker/props.js');
require('../_chunks/dep-5fee93c0.js');
require('../time-picker/time-range-picker.js');
require('../range-input/index.js');
require('../range-input/range-input.js');
require('../range-input/props.js');
require('../range-input/range-input-popup.js');
require('../range-input/range-input-popup-props.js');
require('../time-picker/time-range-picker-props.js');
require('./components/panel/ExtraContent.js');
require('./components/base/Footer.js');
require('./hooks/useDisableDate.js');
require('./hooks/useTableData.js');
require('../_chunks/dep-e70e28c7.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

/** `Object#toString` result references. */
var dateTag = '[object Date]';

/**
 * The base implementation of `_.isDate` without Node.js optimizations.
 *
 * @private
 * @param {*} value The value to check.
 * @returns {boolean} Returns `true` if `value` is a date object, else `false`.
 */
function baseIsDate(value) {
  return isObjectLike.isObjectLike(value) && _baseGetTag.baseGetTag(value) == dateTag;
}

/* Node.js helper references. */
var nodeIsDate = _overArg.nodeUtil && _overArg.nodeUtil.isDate;

/**
 * Checks if `value` is classified as a `Date` object.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param {*} value The value to check.
 * @returns {boolean} Returns `true` if `value` is a date object, else `false`.
 * @example
 *
 * _.isDate(new Date);
 * // => true
 *
 * _.isDate('Mon April 23 2012');
 * // => false
 */
var isDate = nodeIsDate ? _baseUnary.baseUnary(nodeIsDate) : baseIsDate;
var isDate$1 = isDate;

var _DatePicker = Vue.defineComponent({
  name: "TDatePicker",
  props: datePicker_props["default"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var COMPONENT_NAME = index.usePrefixClass("date-picker");
    var _useSingle = datePicker_hooks_useSingle.useSingle(props2),
      inputValue = _useSingle.inputValue,
      popupVisible = _useSingle.popupVisible,
      inputProps = _useSingle.inputProps,
      popupProps = _useSingle.popupProps,
      isHoverCell = _useSingle.isHoverCell,
      cacheValue = _useSingle.cacheValue,
      value = _useSingle.value,
      year = _useSingle.year,
      month = _useSingle.month,
      time = _useSingle.time,
      inputRef = _useSingle.inputRef,
      onChange = _useSingle.onChange;
    var isDisabled = index$1.useDisabled();
    var renderTNodeJSX = index$2.useTNodeJSX();
    var _useConfig = configProvider_hooks_useConfig.useConfig("datePicker"),
      globalConfig = _useConfig.globalConfig;
    var isReadOnly = index$3.useReadonly();
    var _useGlobalIcon = index$4.useGlobalIcon({
        CalendarIcon: tdesignIconsVueNext.CalendarIcon
      }),
      CalendarIcon = _useGlobalIcon.CalendarIcon;
    var formatRef = Vue.computed(function () {
      return format.getDefaultFormat({
        mode: props2.mode,
        format: props2.format,
        valueType: props2.valueType,
        enableTimePicker: props2.multiple ? false : props2.enableTimePicker
      });
    });
    var valueDisplayParams = Vue.computed(function () {
      return {
        value: value.value,
        displayValue: inputValue.value
      };
    });
    Vue.watch(popupVisible, function (visible) {
      var _formatRef$value;
      if (props2.multiple) return;
      if (!props2.needConfirm && props2.enableTimePicker && !visible) {
        var nextValue = format.formatDate(inputValue.value, {
          format: formatRef.value.format
        });
        if (nextValue) {
          onChange === null || onChange === void 0 || onChange(format.formatDate(inputValue.value, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
            defaultTime: props2.defaultTime
          }), {
            dayjsValue: format.parseToDayjs(inputValue.value, formatRef.value.format, void 0, void 0, props2.defaultTime),
            trigger: "confirm"
          });
        } else {
          inputValue.value = format.formatDate(value.value, {
            format: formatRef.value.format
          });
        }
      }
      var dateValue = value.value && !isDate$1(value.value) && !["week", "quarter"].includes(props2.mode) ? utils.covertToDate(value.value, (_formatRef$value = formatRef.value) === null || _formatRef$value === void 0 ? void 0 : _formatRef$value.valueType) : value.value;
      cacheValue.value = format.formatDate(dateValue, {
        format: formatRef.value.valueType,
        targetFormat: formatRef.value.format
      });
      inputValue.value = format.formatDate(dateValue, {
        format: formatRef.value.valueType,
        targetFormat: formatRef.value.format
      });
      if (visible) {
        if ((props2.range && isArray.isArray(props2.range) || props2.panelActiveDate) && !value.value) {
          var _rangeBounds$min$getF, _rangeBounds$min, _rangeBounds$max, _rangeBounds$min$getM, _rangeBounds$min2, _rangeBounds$max2, _props2$panelActiveDa, _props2$panelActiveDa2, _props2$panelActiveDa3, _props2$panelActiveDa4;
          var rangeBounds = utils.getRangeBounds(props2.range);
          var yearFromRange = (_rangeBounds$min$getF = (_rangeBounds$min = rangeBounds.min) === null || _rangeBounds$min === void 0 ? void 0 : _rangeBounds$min.getFullYear()) !== null && _rangeBounds$min$getF !== void 0 ? _rangeBounds$min$getF : (_rangeBounds$max = rangeBounds.max) === null || _rangeBounds$max === void 0 ? void 0 : _rangeBounds$max.getFullYear();
          var monthFromRange = (_rangeBounds$min$getM = (_rangeBounds$min2 = rangeBounds.min) === null || _rangeBounds$min2 === void 0 ? void 0 : _rangeBounds$min2.getMonth()) !== null && _rangeBounds$min$getM !== void 0 ? _rangeBounds$min$getM : (_rangeBounds$max2 = rangeBounds.max) === null || _rangeBounds$max2 === void 0 ? void 0 : _rangeBounds$max2.getMonth();
          year.value = (_props2$panelActiveDa = (_props2$panelActiveDa2 = props2.panelActiveDate) === null || _props2$panelActiveDa2 === void 0 ? void 0 : _props2$panelActiveDa2.year) !== null && _props2$panelActiveDa !== void 0 ? _props2$panelActiveDa : yearFromRange;
          month.value = (_props2$panelActiveDa3 = props2.panelActiveDate) !== null && _props2$panelActiveDa3 !== void 0 && _props2$panelActiveDa3.month ? Number((_props2$panelActiveDa4 = props2.panelActiveDate) === null || _props2$panelActiveDa4 === void 0 ? void 0 : _props2$panelActiveDa4.month) - 1 : monthFromRange;
        } else {
          year.value = format.parseToDayjs(value.value, formatRef.value.valueType).year();
          month.value = format.parseToDayjs(value.value, formatRef.value.format).month();
        }
        time.value = format.formatTime(value.value, formatRef.value.format, formatRef.value.timeFormat, props2.defaultTime);
      } else {
        isHoverCell.value = false;
      }
    });
    function onCellMouseEnter(date) {
      if (props2.multiple) return;
      isHoverCell.value = true;
      inputValue.value = format.formatDate(date, {
        format: formatRef.value.format
      });
    }
    function onCellMouseLeave() {
      if (props2.multiple) return;
      isHoverCell.value = false;
      inputValue.value = format.formatDate(cacheValue.value, {
        format: formatRef.value.format
      });
    }
    function onCellClick(date) {
      var _props2$onPick;
      isHoverCell.value = false;
      if (props2.mode === "date") {
        year.value = date.getFullYear();
        month.value = date.getMonth();
      }
      if (props2.enableTimePicker) {
        cacheValue.value = format.formatDate(date, {
          format: formatRef.value.format
        });
      } else {
        if (props2.multiple) {
          var newDate = processDate(date, props2.defaultTime);
          onChange(newDate, {
            dayjsValue: format.parseToDayjs(date, formatRef.value.format, void 0, void 0, props2.defaultTime),
            trigger: "pick"
          });
          return;
        }
        onChange === null || onChange === void 0 || onChange(format.formatDate(date, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.valueType,
          defaultTime: props2.defaultTime
        }), {
          dayjsValue: format.parseToDayjs(date, formatRef.value.format, void 0, void 0, props2.defaultTime),
          trigger: "pick"
        });
        popupVisible.value = false;
      }
      (_props2$onPick = props2.onPick) === null || _props2$onPick === void 0 || _props2$onPick.call(props2, date);
    }
    function processDate(date, defaultTime) {
      var isSameDate;
      var currentValue = value.value || [];
      var dayjsLocale = globalConfig.value.dayjsLocale;
      var currentDate;
      if (props2.mode !== "week") isSameDate = currentValue.some(function (val) {
        return utils.isSame(format.parseToDayjs(val, formatRef.value.format).toDate(), date, props2.mode, dayjsLocale);
      });else {
        isSameDate = currentValue.some(function (val) {
          return val === dayjs__default["default"](date).locale(dayjsLocale).format(formatRef.value.format);
        });
      }
      if (!isSameDate) {
        currentDate = currentValue.concat(format.formatDate(date, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.valueType,
          defaultTime: defaultTime
        }));
      } else {
        currentDate = currentValue.filter(function (val) {
          return format.formatDate(val, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
            defaultTime: defaultTime
          }) !== format.formatDate(date, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
            defaultTime: defaultTime
          });
        });
      }
      return currentDate;
    }
    function onTagRemoveClick(ctx) {
      if (["week", "quarter"].includes(props2.mode)) {
        onChange === null || onChange === void 0 || onChange(ctx.value, {
          trigger: "tag-remove"
        });
        return;
      }
      var removeDate = dayjs__default["default"](ctx.item).toDate();
      var newDate = processDate(removeDate, props2.defaultTime);
      onChange === null || onChange === void 0 || onChange(newDate, {
        dayjsValue: format.parseToDayjs(removeDate, formatRef.value.format, void 0, void 0, props2.defaultTime),
        trigger: "tag-remove"
      });
    }
    function onTagClearClick(_ref2) {
      var _props2$onClear;
      var e = _ref2.e;
      e.stopPropagation();
      popupVisible.value = false;
      onChange === null || onChange === void 0 || onChange(props2.multiple ? [] : "", {
        dayjsValue: dayjs__default["default"](),
        trigger: "clear"
      });
      (_props2$onClear = props2.onClear) === null || _props2$onClear === void 0 || _props2$onClear.call(props2, {
        e: e
      });
    }
    function onJumperClick(_ref3) {
      var trigger = _ref3.trigger;
      var monthCountMap = {
        date: 1,
        week: 1,
        month: 12,
        quarter: 12,
        year: 120
      };
      var monthCount = monthCountMap[props2.mode] || 0;
      var current = new Date(year.value, month.value);
      var next = null;
      if (trigger === "prev") {
        next = utils.subtractMonth(current, monthCount);
      } else if (trigger === "current") {
        next = new Date();
      } else if (trigger === "next") {
        next = utils.addMonth(current, monthCount);
      }
      var nextYear = next.getFullYear();
      var nextMonth = next.getMonth();
      var yearChanged = year.value !== nextYear;
      var monthChanged = month.value !== nextMonth;
      year.value = nextYear;
      month.value = nextMonth;
      if (yearChanged) {
        var _props2$onYearChange;
        (_props2$onYearChange = props2.onYearChange) === null || _props2$onYearChange === void 0 || _props2$onYearChange.call(props2, {
          year: nextYear,
          date: new Date(nextYear, nextMonth),
          trigger: trigger === "current" ? "today" : "year-".concat(datePicker_utils_index.triggerMap[trigger])
        });
      }
      if (monthChanged) {
        var _props2$onMonthChange;
        (_props2$onMonthChange = props2.onMonthChange) === null || _props2$onMonthChange === void 0 || _props2$onMonthChange.call(props2, {
          month: nextMonth,
          date: new Date(nextYear, nextMonth),
          trigger: trigger === "current" ? "today" : "month-".concat(datePicker_utils_index.triggerMap[trigger])
        });
      }
    }
    function onTimePickerChange(val) {
      var _props2$onPick2;
      time.value = val;
      var _extractTimeObj = utils.extractTimeObj(val),
        hours = _extractTimeObj.hours,
        minutes = _extractTimeObj.minutes,
        seconds = _extractTimeObj.seconds,
        milliseconds = _extractTimeObj.milliseconds,
        meridiem = _extractTimeObj.meridiem;
      var nextHours = hours;
      if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
      if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
      var currentDate = !dayjs__default["default"](inputValue.value, formatRef.value.format).isValid() ? dayjs__default["default"]() : dayjs__default["default"](inputValue.value, formatRef.value.format);
      var nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds).toDate();
      inputValue.value = format.formatDate(nextDate, {
        format: formatRef.value.format
      });
      cacheValue.value = format.formatDate(nextDate, {
        format: formatRef.value.format
      });
      (_props2$onPick2 = props2.onPick) === null || _props2$onPick2 === void 0 || _props2$onPick2.call(props2, nextDate);
    }
    function onConfirmClick(_ref4) {
      var e = _ref4.e;
      var nextValue = format.formatDate(inputValue.value, {
        format: formatRef.value.format
      });
      if (nextValue) {
        var _props2$onConfirm;
        props2 === null || props2 === void 0 || (_props2$onConfirm = props2.onConfirm) === null || _props2$onConfirm === void 0 || _props2$onConfirm.call(props2, {
          date: dayjs__default["default"](nextValue).toDate(),
          e: e
        });
        onChange === null || onChange === void 0 || onChange(format.formatDate(inputValue.value, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.valueType,
          defaultTime: props2.defaultTime
        }), {
          dayjsValue: format.parseToDayjs(inputValue.value, formatRef.value.format, void 0, void 0, props2.defaultTime),
          trigger: "confirm"
        });
      } else {
        inputValue.value = format.formatDate(value.value, {
          format: formatRef.value.format
        });
      }
      popupVisible.value = false;
    }
    function onPresetClick(presetValue, context) {
      var _props2$onPresetClick;
      var presetVal = isFunction.isFunction(presetValue) ? presetValue() : presetValue;
      onChange === null || onChange === void 0 || onChange(format.formatDate(presetVal, {
        format: formatRef.value.format,
        targetFormat: formatRef.value.valueType
      }), {
        dayjsValue: format.parseToDayjs(presetVal, formatRef.value.format, void 0, void 0, props2.defaultTime),
        trigger: "preset"
      });
      inputValue.value = format.formatDate(presetVal, {
        format: formatRef.value.format
      });
      popupVisible.value = false;
      (_props2$onPresetClick = props2.onPresetClick) === null || _props2$onPresetClick === void 0 || _props2$onPresetClick.call(props2, context);
    }
    function onYearChange(nextYear) {
      var _props2$onYearChange2;
      year.value = nextYear;
      (_props2$onYearChange2 = props2.onYearChange) === null || _props2$onYearChange2 === void 0 || _props2$onYearChange2.call(props2, {
        year: nextYear,
        date: dayjs__default["default"](value.value).toDate(),
        trigger: "year-select"
      });
    }
    function onMonthChange(nextMonth) {
      var _props2$onMonthChange2;
      month.value = nextMonth;
      (_props2$onMonthChange2 = props2.onMonthChange) === null || _props2$onMonthChange2 === void 0 || _props2$onMonthChange2.call(props2, {
        month: nextMonth,
        date: dayjs__default["default"](value.value).toDate(),
        trigger: "month-select"
      });
    }
    var panelProps = Vue.computed(function () {
      return {
        value: cacheValue.value,
        year: year.value,
        month: month.value,
        format: formatRef.value.format,
        mode: props2.mode,
        presets: props2.presets,
        multiple: props2.multiple,
        time: props2.multiple ? "" : time.value,
        disableDate: props2.disableDate,
        firstDayOfWeek: props2.firstDayOfWeek,
        timePickerProps: props2.timePickerProps,
        enableTimePicker: props2.multiple ? false : props2.enableTimePicker,
        presetsPlacement: props2.presetsPlacement,
        popupVisible: popupVisible.value,
        needConfirm: props2.needConfirm,
        disableTime: props2.disableTime,
        range: props2.range,
        cell: props2.cell,
        onCellClick: onCellClick,
        onCellMouseEnter: onCellMouseEnter,
        onCellMouseLeave: onCellMouseLeave,
        onJumperClick: onJumperClick,
        onConfirmClick: onConfirmClick,
        onPresetClick: onPresetClick,
        onYearChange: onYearChange,
        onMonthChange: onMonthChange,
        onTimePickerChange: onTimePickerChange,
        onPanelClick: function onPanelClick() {
          var _inputRef$value, _inputRef$value$focus;
          return (_inputRef$value = inputRef.value) === null || _inputRef$value === void 0 || (_inputRef$value$focus = _inputRef$value.focus) === null || _inputRef$value$focus === void 0 ? void 0 : _inputRef$value$focus.call(_inputRef$value);
        }
      };
    });
    var selectInputEvents = index$5.useEventForward(props2.selectInputProps, {
      onClear: onTagClearClick
    });
    return function () {
      return Vue.createVNode("div", {
        "class": COMPONENT_NAME.value
      }, [Vue.createVNode(selectInput_index.SelectInput, Vue.mergeProps({
        "borderless": props2.borderless,
        "disabled": isDisabled.value,
        "value": inputValue.value,
        "label": function label() {
          return renderTNodeJSX("label");
        },
        "status": props2.status,
        "tips": props2.tips,
        "clearable": props2.clearable,
        "readonly": isReadOnly.value,
        "multiple": props2.multiple,
        "popupProps": popupProps.value,
        "inputProps": inputProps.value,
        "placeholder": props2.placeholder || globalConfig.value.placeholder[props2.mode],
        "popupVisible": !isReadOnly.value && popupVisible.value,
        "valueDisplay": function valueDisplay() {
          return renderTNodeJSX("valueDisplay", {
            params: valueDisplayParams.value
          });
        }
      }, props2.selectInputProps, {
        "panel": function panel() {
          return Vue.createVNode(datePicker_components_panel_SinglePanel["default"], panelProps.value, slots);
        },
        "tagInputProps": {
          onRemove: onTagRemoveClick
        },
        "prefixIcon": function prefixIcon() {
          return renderTNodeJSX("prefixIcon");
        },
        "suffixIcon": function suffixIcon() {
          return renderTNodeJSX("suffixIcon") || Vue.createVNode(CalendarIcon, null, null);
        }
      }, selectInputEvents.value), null)]);
    };
  }
});

exports["default"] = _DatePicker;
