// tdesign-vue-next/es/date-picker/components/base/Header.mjs
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import { defineComponent, toRefs, ref, computed, watch, createVNode, createTextVNode } from 'vue';
import { _ as _toConsumableArray } from '../../../_chunks/dep-fe2215b7.mjs';
import { PaginationMini } from '../../../pagination/index.mjs';
import { Select } from '../../../select/index.mjs';
import { u as usePrefixClass } from '../../../_chunks/dep-d6195387.mjs';
import '../../hooks/index.mjs';
import { useConfig } from '../../../config-provider/hooks/useConfig.mjs';
import { useSelectRange } from '../../hooks/useSelectRange.mjs';
import '../../../_chunks/dep-c13a05a2.mjs';
import '../../../pagination/pagination.mjs';
import '../../../_chunks/dep-15e8aadf.mjs';
import '../../../_chunks/dep-39047c40.mjs';
import 'lodash-es';
import 'tdesign-icons-vue-next';
import '../../../input-number/index.mjs';
import '../../../input-number/input-number.mjs';
import '../../../_chunks/dep-edde6136.mjs';
import '../../../_chunks/dep-cb344d2d.mjs';
import '../../../button/index.mjs';
import '../../../button/button.mjs';
import '../../../loading/index.mjs';
import '../../../loading/directive.mjs';
import '../../../loading/plugin.mjs';
import '../../../loading/loading.mjs';
import '../../../loading/icon/gradient.mjs';
import '../../../_chunks/dep-58da668e.mjs';
import '../../../_chunks/dep-478d8d87.mjs';
import '../../../_chunks/dep-80ab80bc.mjs';
import '../../../config-provider/utils/context.mjs';
import '../../../_chunks/dep-fa2e9d13.mjs';
import '../../../_chunks/dep-c22f10cd.mjs';
import '../../../_chunks/dep-2512dba6.mjs';
import '../../../_chunks/dep-1d48345c.mjs';
import '../../../_chunks/dep-4f3de704.mjs';
import '../../../_chunks/dep-15b76bbf.mjs';
import '../../../_chunks/dep-61161446.mjs';
import '../../../_chunks/dep-671c8553.mjs';
import '../../../loading/props.mjs';
import './style/css.mjs';
import '../../../button/props.mjs';
import '../../../input/index.mjs';
import '../../../input/input.mjs';
import '../../../input/props.mjs';
import '../../../input/hooks/index.mjs';
import '../../../input/hooks/useInput.mjs';
import '../../../form/constants/index.mjs';
import '../../../input/hooks/useLengthLimit.mjs';
import '../../../_chunks/dep-c7fa2bca.mjs';
import '../../../input/hooks/useInputEventHandler.mjs';
import '../../../input/hooks/useInputWidth.mjs';
import '../../../input/input-group.mjs';
import '../../../input/input-group-props.mjs';
import '../../../input-number/props.mjs';
import '../../../input-number/hooks/useInputNumber.mjs';
import '../../../_chunks/dep-71bbcf0c.mjs';
import '../../../_chunks/dep-f3ae51ae.mjs';
import '../../../input-adornment/index.mjs';
import '../../../input-adornment/input-adornment.mjs';
import '../../../input-adornment/props.mjs';
import '../../../pagination/props.mjs';
import '../../../pagination/hooks/index.mjs';
import '../../../pagination/hooks/useMoreAction.mjs';
import '../../../pagination/hooks/usePaginationClasses.mjs';
import '../../../select/select.mjs';
import '../../../_chunks/dep-bcb347a6.mjs';
import '../../../common-components/fake-arrow.mjs';
import '../../../select-input/index.mjs';
import '../../../select-input/select-input.mjs';
import '../../../popup/index.mjs';
import '../../../popup/popup.mjs';
import '@popperjs/core';
import '../../../popup/container.mjs';
import '../../../popup/props.mjs';
import '../../../select-input/props.mjs';
import '../../../select-input/hooks/index.mjs';
import '../../../select-input/hooks/useMultiple.mjs';
import '../../../tag-input/index.mjs';
import '../../../tag-input/tag-input.mjs';
import '../../../tag-input/props.mjs';
import '../../../tag-input/hooks/index.mjs';
import '../../../tag-input/hooks/useDragSorter.mjs';
import '../../../tag-input/hooks/useHover.mjs';
import '../../../tag-input/hooks/useTagScroll.mjs';
import '../../../tag-input/hooks/useTagList.mjs';
import '../../../tag/index.mjs';
import '../../../tag/tag.mjs';
import '../../../_chunks/dep-eaa90005.mjs';
import '../../../tag/props.mjs';
import '../../../tag/check-tag.mjs';
import '../../../tag/check-tag-props.mjs';
import '../../../tag/check-tag-group.mjs';
import '../../../tag/check-tag-group-props.mjs';
import '../../../select-input/hooks/useOverlayInnerStyle.mjs';
import '../../../select-input/hooks/useSingle.mjs';
import '../../../select/components/select-panel.mjs';
import '../../../select/option.mjs';
import '../../../select/option-props.mjs';
import '../../../checkbox/index.mjs';
import '../../../checkbox/checkbox.mjs';
import '../../../checkbox/props.mjs';
import '../../../checkbox/constants/index.mjs';
import '../../../checkbox/hooks/useCheckboxLazyLoad.mjs';
import '../../../checkbox/hooks/useKeyboardEvent.mjs';
import '../../../checkbox/group.mjs';
import '../../../checkbox/checkbox-group-props.mjs';
import '../../../select/utils/index.mjs';
import '../../../select/constants/index.mjs';
import '../../../select/option-group.mjs';
import '../../../select/option-group-props.mjs';
import '../../../select/props.mjs';
import '../../../select/hooks/index.mjs';
import '../../../select/hooks/useKeyboardControl.mjs';
import '../../../select/hooks/usePanelVirtualScroll.mjs';
import '../../../select/hooks/useSelectOptions.mjs';
import '../../../pagination/pagination-mini.mjs';
import '../../../pagination/pagination-mini-props.mjs';
import '../../../tooltip/index.mjs';
import '../../../tooltip/tooltip.mjs';
import '../../../tooltip/props.mjs';
import '../../../tooltip/utils/index.mjs';
import '../../hooks/useDisableDate.mjs';
import '../../../_chunks/dep-66bd7218.mjs';
import '../../../_chunks/dep-25210ce3.mjs';
import '../../../_chunks/dep-65ec4207.mjs';
import '../../../_chunks/dep-53e8af8e.mjs';
import '../../hooks/useRange.mjs';
import '../../hooks/useRangeValue.mjs';
import '../../hooks/useSingle.mjs';
import '../../hooks/useSingleValue.mjs';
import '../../hooks/useTableData.mjs';

var TDateHeader = defineComponent({
  name: "TDatePickerHeader",
  props: {
    mode: {
      type: String,
      "default": "date"
    },
    range: [Array, Function],
    year: Number,
    month: Number,
    internalYear: Array,
    partial: String,
    onMonthChange: Function,
    onYearChange: Function,
    onJumperClick: Function
  },
  setup: function setup(props) {
    var _toRefs = toRefs(props),
      year = _toRefs.year,
      month = _toRefs.month;
    var _useConfig = useConfig("classPrefix"),
      classPrefix = _useConfig.classPrefix;
    var COMPONENT_NAME = usePrefixClass("date-picker__header");
    var _useConfig2 = useConfig("datePicker"),
      globalConfig = _useConfig2.globalConfig;
    var _useSelectRange = useSelectRange({
        range: props.range,
        mode: props.mode,
        year: year,
        month: month
      }),
      monthHasAnyAllowed = _useSelectRange.monthHasAnyAllowed,
      yearHasAnyAllowed = _useSelectRange.yearHasAnyAllowed,
      decadeHasAnyAllowed = _useSelectRange.decadeHasAnyAllowed,
      paginationDisabled = _useSelectRange.paginationDisabled,
      canLoadMoreTop = _useSelectRange.canLoadMoreTop,
      canLoadMoreBottom = _useSelectRange.canLoadMoreBottom;
    var yearOptions = ref(initOptions(props.year));
    var showMonthPicker = computed(function () {
      return props.mode === "date" || props.mode === "week";
    });
    var nearestYear = computed(function () {
      var _yearOptions$value$fi;
      var extraYear = props.partial === "end" && props.mode === "year" && Number(props.internalYear[1]) - Number(props.internalYear[0]) <= 9 ? 9 : 0;
      return ((_yearOptions$value$fi = yearOptions.value.find(function (option) {
        return option.value - (props.year + extraYear) <= 9 && option.value - (props.year + extraYear) >= 0;
      })) === null || _yearOptions$value$fi === void 0 ? void 0 : _yearOptions$value$fi.value) || props.year;
    });
    var monthOptions = computed(function () {
      return globalConfig.value.months.map(function (item, index) {
        return {
          label: item,
          value: index,
          disabled: !monthHasAnyAllowed(props.year, index)
        };
      });
    });
    var showPanelTop = computed(function () {
      var options = yearOptions.value;
      if (!options.length) return false;
      var first = options[0].value;
      return canLoadMoreTop(first);
    });
    var showPanelBottom = computed(function () {
      var options = yearOptions.value;
      if (!options.length) return false;
      var last = options[options.length - 1].value;
      return canLoadMoreBottom(last);
    });
    function initOptions(year2) {
      var options = [];
      if (props.mode === "year") {
        var extraYear = year2 % 10;
        var minYear = year2 - extraYear - 100;
        var maxYear = year2 - extraYear + 100;
        for (var i = minYear; i <= maxYear; i += 10) {
          var end = i + 9;
          if (decadeHasAnyAllowed(end)) {
            options.push({
              label: "".concat(i, " - ").concat(end),
              value: end,
              disabled: false
            });
          }
        }
      } else {
        if (yearHasAnyAllowed(year2)) options.push({
          label: "".concat(year2),
          value: year2,
          disabled: false
        });
        for (var _i = 1; _i <= 10; _i++) {
          if (yearHasAnyAllowed(year2 + _i)) options.push({
            label: "".concat(year2 + _i),
            value: year2 + _i,
            disabled: false
          });
          if (yearHasAnyAllowed(year2 - _i)) options.unshift({
            label: "".concat(year2 - _i),
            value: year2 - _i,
            disabled: false
          });
        }
      }
      return options;
    }
    function loadMoreYear(year2, type) {
      var options = [];
      if (props.mode === "year") {
        var extraYear = year2 % 10;
        if (type === "add") {
          for (var i = year2 - extraYear + 10; i <= year2 - extraYear + 50; i += 10) {
            var end = i + 9;
            if (decadeHasAnyAllowed(end)) options.push({
              label: "".concat(i, " - ").concat(end),
              value: end,
              disabled: false
            });
          }
        } else {
          for (var _i2 = year2 - extraYear - 1; _i2 > year2 - extraYear - 50; _i2 -= 10) {
            if (decadeHasAnyAllowed(_i2)) options.unshift({
              label: "".concat(_i2 - 9, " - ").concat(_i2),
              value: _i2,
              disabled: false
            });
          }
        }
      } else if (type === "add") {
        for (var _i3 = year2 + 1; _i3 <= year2 + 10; _i3++) {
          if (yearHasAnyAllowed(_i3)) options.push({
            label: "".concat(_i3),
            value: _i3,
            disabled: false
          });
        }
      } else {
        for (var _i4 = year2 - 1; _i4 > year2 - 10; _i4--) {
          if (yearHasAnyAllowed(_i4)) options.unshift({
            label: "".concat(_i4),
            value: _i4,
            disabled: false
          });
        }
      }
      return options;
    }
    var labelMap = computed(function () {
      return {
        year: {
          prev: globalConfig.value.preDecade,
          current: globalConfig.value.now,
          next: globalConfig.value.nextDecade
        },
        month: {
          prev: globalConfig.value.preYear,
          current: globalConfig.value.now,
          next: globalConfig.value.nextYear
        },
        date: {
          prev: globalConfig.value.preMonth,
          current: globalConfig.value.now,
          next: globalConfig.value.nextMonth
        },
        quarter: {
          prev: globalConfig.value.preYear,
          current: globalConfig.value.now,
          next: globalConfig.value.nextYear
        },
        week: {
          prev: globalConfig.value.preMonth,
          current: globalConfig.value.now,
          next: globalConfig.value.nextMonth
        }
      };
    });
    function handleScroll(_ref) {
      var e = _ref.e;
      if (e.target.scrollTop === 0) {
        if (showPanelTop.value) handlePanelTopClick(e);
      } else if (e.target.scrollTop === e.target.scrollHeight - e.target.clientHeight) {
        if (showPanelBottom.value) handlePanelBottomClick(e);
      }
    }
    function handlePanelTopClick(e) {
      var _e$stopPropagation;
      e === null || e === void 0 || (_e$stopPropagation = e.stopPropagation) === null || _e$stopPropagation === void 0 || _e$stopPropagation.call(e);
      var firstYear = yearOptions.value[0].value;
      var options = loadMoreYear(firstYear, "reduce");
      yearOptions.value = [].concat(_toConsumableArray(options), _toConsumableArray(yearOptions.value));
    }
    function handlePanelBottomClick(e) {
      var _e$stopPropagation2;
      e === null || e === void 0 || (_e$stopPropagation2 = e.stopPropagation) === null || _e$stopPropagation2 === void 0 || _e$stopPropagation2.call(e);
      var lastYear = yearOptions.value.slice(-1)[0].value;
      var options = loadMoreYear(lastYear, "add");
      yearOptions.value = [].concat(_toConsumableArray(yearOptions.value), _toConsumableArray(options));
    }
    watch(function () {
      return props.mode;
    }, function () {
      yearOptions.value = initOptions(props.year);
    });
    return function () {
      return createVNode("div", {
        "class": COMPONENT_NAME.value
      }, [createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-controller")
      }, [showMonthPicker.value && createVNode(Select, {
        "class": "".concat(COMPONENT_NAME.value, "-controller-month"),
        "value": props.month,
        "options": monthOptions.value,
        "onChange": function onChange(val) {
          var _props$onMonthChange;
          (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 || _props$onMonthChange.call(props, val);
        },
        "popupProps": {
          attach: function attach(triggerElement) {
            return triggerElement.parentNode;
          },
          overlayClassName: "".concat(COMPONENT_NAME.value, "-controller-month-popup")
        }
      }, null), createVNode(Select, {
        "class": "".concat(COMPONENT_NAME.value, "-controller-year"),
        "value": props.mode === "year" ? nearestYear.value : props.year,
        "options": yearOptions.value,
        "onChange": function onChange(val) {
          var _props$onYearChange;
          (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 || _props$onYearChange.call(props, val);
        },
        "popupProps": {
          onScroll: handleScroll,
          attach: function attach(triggerElement) {
            return triggerElement.parentNode;
          },
          overlayClassName: "".concat(COMPONENT_NAME.value, "-controller-year-popup")
        },
        "panelTopContent": function panelTopContent() {
          return showPanelTop.value && createVNode("div", {
            "class": "".concat(classPrefix.value, "-select-option"),
            "onClick": handlePanelTopClick
          }, [createTextVNode("...")]);
        },
        "panelBottomContent": function panelBottomContent() {
          return showPanelBottom.value && createVNode("div", {
            "class": "".concat(classPrefix.value, "-select-option"),
            "onClick": handlePanelBottomClick
          }, [createTextVNode("...")]);
        }
      }, null)]), createVNode(PaginationMini, {
        "tips": labelMap.value[props.mode],
        "size": "small",
        "onChange": props.onJumperClick,
        "disabled": paginationDisabled.value
      }, null)]);
    };
  }
});

export { TDateHeader as default };
