// tdesign-vue-next/cjs/card/props.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  actions: {
    type: [String, Function]
  },
  avatar: {
    type: [String, Function]
  },
  bordered: {
    type: Boolean,
    "default": true
  },
  bodyClassName: {
    type: String
  },
  bodyStyle: {
    type: Object
  },
  content: {
    type: [String, Function]
  },
  cover: {
    type: [String, Function]
  },
  "default": {
    type: [String, Function]
  },
  description: {
    type: [String, Function]
  },
  footer: {
    type: [String, Function]
  },
  footerClassName: {
    type: String
  },
  footerStyle: {
    type: Object
  },
  header: {
    type: [String, Function]
  },
  headerClassName: {
    type: String
  },
  headerStyle: {
    type: Object
  },
  headerBordered: Boolean,
  hoverShadow: Boolean,
  loading: {
    type: [Boolean, Function],
    "default": false
  },
  loadingProps: {
    type: Object
  },
  shadow: Boolean,
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["medium", "small"].includes(val);
    }
  },
  status: {
    type: String,
    "default": ""
  },
  subtitle: {
    type: [String, Function]
  },
  theme: {
    type: String,
    "default": "normal",
    validator: function validator(val) {
      if (!val) return true;
      return ["normal", "poster1", "poster2"].includes(val);
    }
  },
  title: {
    type: [String, Function]
  }
};

exports["default"] = props;
