// tdesign-vue-next/cjs/descriptions/descriptions.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var index$2 = require('../_chunks/dep-e53cd0a5.js');
var index$3 = require('../_chunks/dep-0e56b1da.js');
var index$1 = require('../_chunks/dep-6c498eb2.js');
var index = require('../_chunks/dep-b56e7ff5.js');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-d1a4892a.js');
require('../_chunks/dep-c7472ccd.js');
require('@babel/runtime/helpers/toConsumableArray');
var descriptions_props = require('./props.js');
var descriptions_constants_index = require('./constants/index.js');
var descriptions_descriptionsRow = require('./descriptions-row.js');
var descriptions_utils_index = require('../_chunks/dep-517de294.js');
var isArray = require('../_chunks/dep-d1d374b4.js');
var isNil = require('../_chunks/dep-e70e28c7.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-e2205a6f.js');
require('../_chunks/dep-6f4a5c31.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-37af9215.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('../_chunks/dep-bf2e7501.js');
require('dayjs');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-2e9fed7b.js');
require('../_chunks/dep-a324ad6a.js');
require('../_chunks/dep-06e53606.js');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var _Descriptions = Vue.defineComponent({
  name: "TDescriptions",
  props: descriptions_props["default"],
  setup: function setup(props2) {
    var COMPONENT_NAME = index.usePrefixClass("descriptions");
    var _useCommonClassName = index$1.useCommonClassName(),
      SIZE = _useCommonClassName.SIZE;
    var getChildByName = index$2.useChildComponentSlots();
    var renderTNodeJSX = index$3.useTNodeJSX();
    var itemsType = Vue.ref(descriptions_utils_index.ItemsType.props);
    var getRows = function getRows() {
      var column = props2.column,
        layout = props2.layout;
      var items = [];
      if (isArray.isArray(props2.items)) {
        items = props2.items.map(function (item) {
          return {
            label: descriptions_utils_index.renderCustomNode(item.label),
            content: descriptions_utils_index.renderCustomNode(item.content),
            span: item.span || 1
          };
        });
        itemsType.value = descriptions_utils_index.ItemsType.props;
      } else {
        var slots = getChildByName("TDescriptionsItem");
        if (slots.length !== 0) {
          items = slots;
          itemsType.value = descriptions_utils_index.ItemsType.slots;
        }
      }
      if (layout === "vertical") {
        return [items];
      }
      var temp = [];
      var reset = column;
      var res = [];
      items.forEach(function (item, index) {
        var span = 1;
        if (descriptions_utils_index.itemTypeIsProps(itemsType.value)) {
          span = isNil.isNil(item.span) ? span : item.span;
          span = span > column ? column : span;
        } else {
          var _item$props;
          item.props = item.props || {};
          span = isNil.isNil((_item$props = item.props) === null || _item$props === void 0 ? void 0 : _item$props.span) ? span : item.props.span;
          span = span > column ? column : span;
          item.props.span = span;
        }
        if (reset >= span) {
          temp.push(item);
          reset -= span;
        } else {
          res.push(temp);
          temp = [item];
          reset = column - span;
        }
        if (index === items.length - 1) {
          if (descriptions_utils_index.itemTypeIsProps(itemsType.value)) {
            item.span += reset;
          } else {
            item.props.span += reset;
          }
          res.push(temp);
        }
      });
      return res;
    };
    Vue.provide(descriptions_constants_index.descriptionsKey, props2);
    var renderBody = function renderBody() {
      var tableClass = ["".concat(COMPONENT_NAME.value, "__body"), SIZE.value[props2.size], _defineProperty__default["default"]({}, "".concat(COMPONENT_NAME.value, "__body--fixed"), props2.tableLayout === "fixed"), _defineProperty__default["default"]({}, "".concat(COMPONENT_NAME.value, "__body--border"), props2.bordered)];
      return Vue.createVNode("table", {
        "class": tableClass
      }, [Vue.createVNode("tbody", null, [getRows().map(function (row) {
        return Vue.createVNode(descriptions_descriptionsRow["default"], {
          "item-type": itemsType.value,
          "row": row
        }, null);
      })])]);
    };
    var renderHeader = function renderHeader() {
      var title = renderTNodeJSX("title");
      return title ? Vue.createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "__header")
      }, [title]) : "";
    };
    return function () {
      return Vue.createVNode("div", {
        "class": COMPONENT_NAME.value
      }, [renderHeader(), renderBody()]);
    };
  }
});

exports["default"] = _Descriptions;
