// tdesign-vue-next/esm/qrcode/components/props.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

var QRCodeSubComponentProps = {
  value: {
    type: String,
    "default": ""
  },
  size: {
    type: Number,
    "default": 128
  },
  level: {
    type: String,
    "default": "L"
  },
  bgColor: {
    type: String,
    "default": "#FFFFFF"
  },
  fgColor: {
    type: String,
    "default": "#000000"
  },
  style: {
    type: Object,
    "default": function _default() {
      return {};
    }
  },
  includeMargin: {
    type: Boolean,
    "default": false
  },
  marginSize: {
    type: Number,
    "default": 0
  },
  imageSettings: {
    type: Object,
    "default": function _default() {
      return {};
    }
  },
  title: {
    type: String,
    "default": ""
  },
  minVersion: {
    type: Number,
    "default": 1
  }
};
var QRCodeStatusProps = {
  locale: {
    type: Object,
    "default": function _default() {
      return {};
    }
  },
  onRefresh: {
    type: Function
  },
  statusRender: {
    type: Function
  },
  status: {
    type: String,
    "default": "active"
  }
};

export { QRCodeStatusProps, QRCodeSubComponentProps };
