// tdesign-vue-next/cjs/_chunks/dep-c8d40779.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

var Vue = require('vue');
var isObject = require('./dep-c644d54f.js');
var isString = require('./dep-11e6da14.js');
var isArray = require('./dep-d1d374b4.js');

function getDefaultNode(options) {
  var defaultNode;
  if (isObject.isObject(options) && "defaultNode" in options) {
    defaultNode = options.defaultNode;
  } else if (Vue.isVNode(options) || isString.isString(options)) {
    defaultNode = options;
  }
  return defaultNode;
}
function getChildren(content) {
  var childList = [];
  var _innerGetChildren = function innerGetChildren(content2) {
    if (!isArray.isArray(content2)) return;
    content2.forEach(function (item) {
      if (item.children && isArray.isArray(item.children)) {
        if (item.type !== Vue.Fragment) return;
        _innerGetChildren(item.children);
      } else {
        childList.push(item);
      }
    });
    return childList;
  };
  return _innerGetChildren(content);
}
function getParams(options) {
  return isObject.isObject(options) && "params" in options ? options.params : {};
}
function getSlotFirst(options) {
  return isObject.isObject(options) && "slotFirst" in options ? options.slotFirst : false;
}
function isPropsUsed(propsName) {
  var _getCurrentInstance = Vue.getCurrentInstance(),
    vnode = _getCurrentInstance.vnode,
    slots = _getCurrentInstance.slots;
  var vProps = vnode.props || {};
  if (Object.prototype.hasOwnProperty.call(vProps, propsName)) return true;
  if (slots && Object.prototype.hasOwnProperty.call(slots, propsName)) return true;
  return false;
}

exports.getChildren = getChildren;
exports.getDefaultNode = getDefaultNode;
exports.getParams = getParams;
exports.getSlotFirst = getSlotFirst;
exports.isPropsUsed = isPropsUsed;
