// tdesign-vue-next/cjs/time-picker/time-picker.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var dayjs = require('dayjs');
var customParseFormat = require('../_chunks/dep-a7068a83.js');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var timePicker_panel_timePickerPanel = require('./panel/time-picker-panel.js');
var selectInput_index = require('../select-input/index.js');
var timePicker_panel_singlePanel = require('../_chunks/dep-5fee93c0.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-c7472ccd.js');
var index = require('../_chunks/dep-0e56b1da.js');
var index$2 = require('../_chunks/dep-6c498eb2.js');
var index$1 = require('../_chunks/dep-b56e7ff5.js');
var index$6 = require('../_chunks/dep-a8de5365.js');
var index$3 = require('../_chunks/dep-125f71ff.js');
require('../_chunks/dep-d1a4892a.js');
var index$4 = require('../_chunks/dep-4725a846.js');
var index$5 = require('../_chunks/dep-d149369c.js');
var index$7 = require('../_chunks/dep-6f95225b.js');
var timePicker_props = require('./props.js');
var configProvider_hooks_useConfig = require('../config-provider/hooks/useConfig.js');
require('../_chunks/dep-bf2e7501.js');
require('../_chunks/dep-83ab8a79.js');
require('../_chunks/dep-850526b5.js');
require('../_chunks/dep-e2205a6f.js');
require('../_chunks/dep-6f4a5c31.js');
require('../button/button.js');
require('../loading/index.js');
require('../_chunks/dep-548039eb.js');
require('../loading/plugin.js');
require('../_chunks/dep-4a5eca86.js');
require('../loading/icon/gradient.js');
require('../_chunks/dep-fd403354.js');
require('../_chunks/dep-585d18dc.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-8308792f.js');
require('../_chunks/dep-d5e8d527.js');
require('../_chunks/dep-bde8abb8.js');
require('../_chunks/dep-8e4f83bf.js');
require('../_chunks/dep-eb214be8.js');
require('../_chunks/dep-37af9215.js');
require('../_chunks/dep-c644d54f.js');
require('../_chunks/dep-11e6da14.js');
require('../_chunks/dep-d1d374b4.js');
require('../loading/props.js');
require('../_chunks/dep-07f46433.js');
require('../_chunks/dep-0b08f67c.js');
require('../_chunks/dep-eb1f700b.js');
require('../_chunks/dep-9399a4e2.js');
require('../_chunks/dep-58ff00b9.js');
require('../_chunks/dep-0c9948d8.js');
require('../_chunks/dep-2e9fed7b.js');
require('../_chunks/dep-a324ad6a.js');
require('../_chunks/dep-06e53606.js');
require('../_chunks/dep-c8d40779.js');
require('../_chunks/dep-388cb46f.js');
require('../_chunks/dep-2333b13f.js');
require('../_chunks/dep-5729cb14.js');
require('../_chunks/dep-89d65b78.js');
require('../_chunks/dep-dc9d7b38.js');
require('../_chunks/dep-445d3f65.js');
require('../_chunks/dep-09b19404.js');
require('../_chunks/dep-41cc84a3.js');
require('../_chunks/dep-c51f4345.js');
require('../_chunks/dep-e2c29461.js');
require('../_chunks/dep-52d3718c.js');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/classCallCheck');
require('../_chunks/dep-f8f462e0.js');
require('../_chunks/dep-47af4fff.js');
require('../_chunks/dep-ea9fd5a2.js');
require('../_chunks/dep-a8f427e5.js');
require('../_chunks/dep-9997364c.js');
require('../_chunks/dep-1e2d078f.js');
require('../_chunks/dep-7d888cd2.js');
require('../_chunks/dep-ffd72e3b.js');
require('../_chunks/dep-6ed4388e.js');
require('../_chunks/dep-742210e1.js');
require('../button/props.js');
require('../_chunks/dep-a90757f2.js');
require('../_chunks/dep-f7deb88a.js');
require('../_chunks/dep-5e1fbdf8.js');
require('../select-input/select-input.js');
require('../popup/index.js');
require('../popup/popup.js');
require('@popperjs/core');
require('../popup/container.js');
require('../popup/props.js');
require('../_chunks/dep-6344c74f.js');
require('../_chunks/dep-d8ab5fb5.js');
require('../_chunks/dep-cde052f0.js');
require('../_chunks/dep-a3501e24.js');
require('../select-input/props.js');
require('../select-input/hooks/useMultiple.js');
require('../tag-input/index.js');
require('../tag-input/tag-input.js');
require('../input/index.js');
require('../input/input.js');
require('../input/props.js');
require('../input/hooks/useInput.js');
require('../form/constants/index.js');
require('../input/hooks/useLengthLimit.js');
require('../_chunks/dep-880baab0.js');
require('../input/hooks/useInputEventHandler.js');
require('../input/hooks/useInputWidth.js');
require('../input/input-group.js');
require('../input/input-group-props.js');
require('../tag-input/props.js');
require('../_chunks/dep-86694091.js');
require('../tag-input/hooks/useDragSorter.js');
require('../tag-input/hooks/useHover.js');
require('../tag-input/hooks/useTagScroll.js');
require('../tag-input/hooks/useTagList.js');
require('../tag/index.js');
require('../tag/tag.js');
require('tinycolor2');
require('../tag/props.js');
require('../tag/check-tag.js');
require('../tag/check-tag-props.js');
require('../_chunks/dep-1177d395.js');
require('../tag/check-tag-group.js');
require('../tag/check-tag-group-props.js');
require('../select-input/hooks/useOverlayInnerStyle.js');
require('../select-input/hooks/useSingle.js');
require('../_chunks/dep-a66df673.js');
require('../_chunks/dep-f3b67279.js');
require('../_chunks/dep-117093bb.js');
require('../_chunks/dep-e01a4fce.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
dayjs__default["default"].extend(customParseFormat.customParseFormat);
var _TimePicker = Vue.defineComponent({
  name: "TTimePicker",
  props: timePicker_props["default"],
  setup: function setup(props2) {
    var renderTNodeJSX = index.useTNodeJSX();
    var _useConfig = configProvider_hooks_useConfig.useConfig("timePicker"),
      globalConfig = _useConfig.globalConfig;
    var COMPONENT_NAME = index$1.usePrefixClass("time-picker");
    var _useCommonClassName = index$2.useCommonClassName(),
      STATUS = _useCommonClassName.STATUS;
    var _useGlobalIcon = index$3.useGlobalIcon({
        TimeIcon: tdesignIconsVueNext.TimeIcon
      }),
      TimeIcon = _useGlobalIcon.TimeIcon;
    var currentValue = Vue.ref("");
    var isShowPanel = Vue.ref(false);
    var isReadonly = index$4.useReadonly();
    var _toRefs = Vue.toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = index$5.useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      innerValue = _useVModel2[0],
      setInnerValue = _useVModel2[1];
    var isDisabled = index$6.useDisabled();
    var _toRefs2 = Vue.toRefs(props2),
      allowInput = _toRefs2.allowInput,
      format = _toRefs2.format;
    var inputClasses = Vue.computed(function () {
      return ["".concat(COMPONENT_NAME.value, "__group"), _defineProperty__default["default"]({}, STATUS.value.focused, isShowPanel.value)];
    });
    var handleShowPopup = function handleShowPopup(visible, context) {
      var _props2$onOpen, _props2$onClose;
      isShowPanel.value = visible;
      visible ? (_props2$onOpen = props2.onOpen) === null || _props2$onOpen === void 0 ? void 0 : _props2$onOpen.call(props2, context) : (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 ? void 0 : _props2$onClose.call(props2, context);
    };
    var handleClear = function handleClear(context) {
      var _props2$onClear;
      var e = context.e;
      e.stopPropagation();
      currentValue.value = null;
      setInnerValue(null);
      props2 === null || props2 === void 0 || (_props2$onClear = props2.onClear) === null || _props2$onClear === void 0 || _props2$onClear.call(props2, context);
    };
    var handleInputChange = function handleInputChange(value2) {
      currentValue.value = value2;
    };
    var handleInputBlur = function handleInputBlur(value2, context) {
      var _props2$onBlur;
      if (allowInput.value) {
        var isValidTime = timePicker_panel_singlePanel.validateInputValue(currentValue.value, format.value);
        if (isValidTime) {
          setInnerValue(timePicker_panel_singlePanel.formatInputValue(currentValue.value, format.value));
        }
      }
      (_props2$onBlur = props2.onBlur) === null || _props2$onBlur === void 0 || _props2$onBlur.call(props2, {
        value: value2,
        inputValue: context.inputValue,
        e: context.e
      });
    };
    var handleClickConfirm = function handleClickConfirm(e) {
      var _props2$onConfirm;
      props2 === null || props2 === void 0 || (_props2$onConfirm = props2.onConfirm) === null || _props2$onConfirm === void 0 || _props2$onConfirm.call(props2, {
        e: e
      });
      var isValidTime = timePicker_panel_singlePanel.validateInputValue(currentValue.value, format.value);
      if (isValidTime) setInnerValue(currentValue.value);
      isShowPanel.value = false;
    };
    var handlePanelChange = function handlePanelChange(v, e) {
      var _props2$onPick;
      currentValue.value = v;
      (_props2$onPick = props2.onPick) === null || _props2$onPick === void 0 || _props2$onPick.call(props2, v, {
        e: e
      });
    };
    var valueDisplayParams = Vue.computed(function () {
      var _innerValue$value;
      return {
        value: isShowPanel.value ? currentValue.value : (_innerValue$value = innerValue.value) !== null && _innerValue$value !== void 0 ? _innerValue$value : void 0
      };
    });
    Vue.watch(function () {
      return isShowPanel.value;
    }, function () {
      var _innerValue$value2;
      currentValue.value = isShowPanel.value ? (_innerValue$value2 = innerValue.value) !== null && _innerValue$value2 !== void 0 ? _innerValue$value2 : "" : "";
    });
    var selectEvents = index$7.useEventForward(props2.selectInputProps, {
      onInputChange: handleInputChange,
      onBlur: handleInputBlur,
      onPopupVisibleChange: handleShowPopup
    });
    return function () {
      var _innerValue$value3, _innerValue$value4;
      return Vue.createVNode("div", {
        "class": COMPONENT_NAME.value
      }, [Vue.createVNode(selectInput_index.SelectInput, Vue.mergeProps({
        "onFocus": props2.onFocus,
        "onClear": handleClear,
        "borderless": props2.borderless,
        "disabled": isDisabled.value,
        "clearable": props2.clearable,
        "allowInput": allowInput.value,
        "class": inputClasses.value,
        "label": props2.label,
        "prefixIcon": function prefixIcon() {
          return renderTNodeJSX("prefixIcon");
        },
        "suffixIcon": function suffixIcon() {
          return renderTNodeJSX("suffixIcon") || Vue.createVNode(TimeIcon, null, null);
        },
        "popupVisible": !isReadonly.value && isShowPanel.value,
        "placeholder": !innerValue.value ? props2.placeholder || globalConfig.value.placeholder : void 0,
        "value": isShowPanel.value ? currentValue.value : (_innerValue$value3 = innerValue.value) !== null && _innerValue$value3 !== void 0 ? _innerValue$value3 : void 0,
        "inputValue": isShowPanel.value ? currentValue.value : (_innerValue$value4 = innerValue.value) !== null && _innerValue$value4 !== void 0 ? _innerValue$value4 : void 0,
        "inputProps": _objectSpread(_objectSpread({}, props2.inputProps), {}, {
          size: props2.size
        }),
        "popupProps": _objectSpread({
          overlayInnerStyle: {
            width: "auto",
            padding: 0
          }
        }, props2.popupProps),
        "status": props2.status,
        "tips": props2.tips,
        "valueDisplay": function valueDisplay() {
          return renderTNodeJSX("valueDisplay", {
            params: valueDisplayParams.value
          });
        }
      }, selectEvents.value, {
        "panel": function panel() {
          return Vue.createVNode(timePicker_panel_timePickerPanel["default"], {
            "steps": props2.steps,
            "format": format.value,
            "value": currentValue.value,
            "isFooterDisplay": true,
            "isShowPanel": isShowPanel.value,
            "disableTime": props2.disableTime,
            "onChange": handlePanelChange,
            "hideDisabledTime": props2.hideDisabledTime,
            "handleConfirmClick": handleClickConfirm,
            "presets": props2.presets
          }, null);
        }
      }), null)]);
    };
  }
});

exports["default"] = _TimePicker;
