// tdesign-vue-next/lib/date-picker/components/panel/PanelContent.js
/**
 * tdesign v1.20.1
 * (c) 2026 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, createVNode } from 'vue';
import { isFunction, isArray } from 'lodash-es';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/typeof';
import '../../../_chunks/dep-651bdc55.js';
import { u as usePrefixClass } from '../../../_chunks/dep-59f77e78.js';
import '@babel/runtime/helpers/slicedToArray';
import '../../../_chunks/dep-204ae706.js';
import TDateHeader from '../base/Header.js';
import TDateTable from '../base/Table.js';
import { TimePickerPanel } from '../../../time-picker/index.js';
import { g as getDefaultFormat } from '../../../_chunks/dep-f54e73f7.js';
import { parseToDateTime } from '../../utils/index.js';
import '../../../_chunks/dep-76473034.js';
import '../../../config-provider/hooks/useConfig.js';
import '../../../config-provider/utils/context.js';
import '../../../_chunks/dep-5c1cdd0b.js';
import 'dayjs';
import '@babel/runtime/helpers/createClass';
import '@babel/runtime/helpers/classCallCheck';
import '../../../pagination/index.js';
import '../../../pagination/pagination.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import 'tdesign-icons-vue-next';
import '../../../_chunks/dep-eff3e90f.js';
import '../../../_chunks/dep-a7922f61.js';
import '../../../_chunks/dep-69502ff8.js';
import '../../../_chunks/dep-c6879e67.js';
import '../../../_chunks/dep-3d495891.js';
import '../../../_chunks/dep-f32cc39c.js';
import '../../../input-number/index.js';
import '../../../input-number/input-number.js';
import '../../../button/index.js';
import '../../../button/button.js';
import '../../../loading/index.js';
import '../../../loading/directive.js';
import '../../../loading/plugin.js';
import '../../../_chunks/dep-d32d8561.js';
import '../../../loading/icon/gradient.js';
import '../../../_chunks/dep-d15b7e74.js';
import '../../../_chunks/dep-a9edb4d8.js';
import '../../../_chunks/dep-7daa5fe4.js';
import '../../../loading/props.js';
import '../../../_chunks/dep-5dfdb698.js';
import '../../../button/props.js';
import '../../../_chunks/dep-ed0885e7.js';
import '../../../_chunks/dep-16e13fa9.js';
import '../../../input/index.js';
import '../../../input/input.js';
import '../../../input/props.js';
import '../../../_chunks/dep-0d255418.js';
import '../../../input/hooks/useInput.js';
import '../../../form/constants/index.js';
import '../../../input/hooks/useLengthLimit.js';
import '../../../_chunks/dep-f2301d1c.js';
import '../../../input/hooks/useInputEventHandler.js';
import '../../../input/hooks/useInputWidth.js';
import '../../../_chunks/dep-f5c5752f.js';
import '../../../input/input-group.js';
import '../../../input/input-group-props.js';
import '../../../input-number/props.js';
import '../../../_chunks/dep-ff412569.js';
import '../../../input-number/hooks/useInputNumber.js';
import '../../../_chunks/dep-b21bd6d2.js';
import '../../../_chunks/dep-75be08e7.js';
import '../../../select/index.js';
import '../../../select/select.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '../../../common-components/fake-arrow.js';
import '../../../select-input/index.js';
import '../../../select-input/select-input.js';
import '../../../popup/index.js';
import '../../../popup/popup.js';
import '@popperjs/core';
import '../../../popup/container.js';
import '../../../popup/props.js';
import '../../../select-input/props.js';
import '../../../select-input/hooks/useMultiple.js';
import '../../../tag-input/index.js';
import '../../../tag-input/tag-input.js';
import '../../../tag-input/props.js';
import '../../../tag-input/hooks/useDragSorter.js';
import '../../../tag-input/hooks/useHover.js';
import '../../../tag-input/hooks/useTagScroll.js';
import '../../../tag-input/hooks/useTagList.js';
import '../../../tag/index.js';
import '../../../tag/tag.js';
import 'tinycolor2';
import '../../../tag/props.js';
import '../../../tag/check-tag.js';
import '../../../tag/check-tag-props.js';
import '../../../_chunks/dep-8d513880.js';
import '../../../tag/check-tag-group.js';
import '../../../tag/check-tag-group-props.js';
import '../../../select-input/hooks/useOverlayInnerStyle.js';
import '../../../select-input/hooks/useSingle.js';
import '../../../select/components/select-panel.js';
import '../../../select/option.js';
import '../../../select/option-props.js';
import '../../../checkbox/index.js';
import '../../../checkbox/checkbox.js';
import '../../../checkbox/props.js';
import '../../../checkbox/constants/index.js';
import '../../../checkbox/hooks/useCheckboxLazyLoad.js';
import '../../../_chunks/dep-2cabe3fe.js';
import '../../../checkbox/hooks/useKeyboardEvent.js';
import '../../../checkbox/group.js';
import '../../../checkbox/checkbox-group-props.js';
import '../../../_chunks/dep-d4579130.js';
import '../../../select/utils/index.js';
import '../../../select/constants/index.js';
import '../../../select/option-group.js';
import '../../../select/option-group-props.js';
import '../../../select/props.js';
import '../../../select/hooks/usePanelVirtualScroll.js';
import '../../../select/hooks/useKeyboardControl.js';
import '../../../select/hooks/useSelectOptions.js';
import '../../../input-adornment/index.js';
import '../../../input-adornment/input-adornment.js';
import '../../../input-adornment/props.js';
import '../../../pagination/props.js';
import '../../../pagination/hooks/useMoreAction.js';
import '../../../pagination/hooks/usePaginationClasses.js';
import '../../../pagination/pagination-mini.js';
import '../../../pagination/pagination-mini-props.js';
import '../../../tooltip/index.js';
import '../../../tooltip/tooltip.js';
import '../../../tooltip/props.js';
import '../../../tooltip/utils/index.js';
import '../../../_chunks/dep-a7cf6cc7.js';
import '@babel/runtime/helpers/objectDestructuringEmpty';
import '@babel/runtime/helpers/extends';
import '../../../_chunks/dep-3331fba2.js';
import '../../../_chunks/dep-7a2c5955.js';
import '../../hooks/useSelectRange.js';
import '../base/Cell.js';
import '../../../time-picker/time-picker.js';
import '../../../time-picker/panel/time-picker-panel.js';
import '../../../_chunks/dep-ea86b40b.js';
import '../../../time-picker/props.js';
import '../../../_chunks/dep-c817a169.js';
import '../../../time-picker/time-range-picker.js';
import '../../../range-input/index.js';
import '../../../range-input/range-input.js';
import '../../../range-input/props.js';
import '../../../range-input/range-input-popup.js';
import '../../../range-input/range-input-popup-props.js';
import '../../../time-picker/time-range-picker-props.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TPanelContent = defineComponent({
  name: "TPanelContent",
  props: {
    mode: String,
    format: String,
    enableTimePicker: Boolean,
    timePickerProps: {
      type: Object,
      "default": function _default() {
        return {};
      }
    },
    year: Number,
    month: Number,
    range: [Array, Function],
    tableData: Array,
    time: String,
    multiple: Boolean,
    firstDayOfWeek: Number,
    partial: String,
    popupVisible: Boolean,
    onYearChange: Function,
    onMonthChange: Function,
    onJumperClick: Function,
    onCellMouseEnter: Function,
    onCellClick: Function,
    onCellMouseLeave: Function,
    onTimePickerChange: Function,
    value: [String, Number, Array, Date],
    internalYear: Array,
    disableTime: Function,
    defaultTime: [String, Array],
    cell: {
      type: Function
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots;
    var COMPONENT_NAME = usePrefixClass("date-picker__panel");
    var _getDefaultFormat = getDefaultFormat({
        mode: props.mode,
        format: props.format,
        enableTimePicker: props.enableTimePicker
      }),
      timeFormat = _getDefaultFormat.timeFormat;
    var disableTimeOptions = function disableTimeOptions() {
      if (!isFunction(props.disableTime)) {
        return {};
      }
      var startValue = isArray(props.value) ? props.value[0] : props.value;
      var endValue = isArray(props.value) ? props.value[1] : props.value;
      var isRangePicker = props.partial;
      if (!isRangePicker) {
        return props.disableTime(parseToDateTime(startValue, props.format));
      }
      return props.disableTime([parseToDateTime(startValue, props.format), parseToDateTime(endValue, props.format)], {
        partial: props.partial
      });
    };
    var defaultTimeValue = "00:00:00";
    return function () {
      return createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-content")
      }, [createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-").concat(props.mode)
      }, [createVNode(TDateHeader, {
        "mode": props.mode,
        "year": props.year,
        "month": props.month,
        "range": props.range,
        "internalYear": props.internalYear,
        "partial": props.partial,
        "onMonthChange": function onMonthChange(val) {
          var _props$onMonthChange;
          return (_props$onMonthChange = props.onMonthChange) === null || _props$onMonthChange === void 0 ? void 0 : _props$onMonthChange.call(props, val, {
            partial: props.partial
          });
        },
        "onYearChange": function onYearChange(val) {
          var _props$onYearChange;
          return (_props$onYearChange = props.onYearChange) === null || _props$onYearChange === void 0 ? void 0 : _props$onYearChange.call(props, val, {
            partial: props.partial
          });
        },
        "onJumperClick": function onJumperClick(_ref2) {
          var _props$onJumperClick;
          var trigger = _ref2.trigger;
          return (_props$onJumperClick = props.onJumperClick) === null || _props$onJumperClick === void 0 ? void 0 : _props$onJumperClick.call(props, {
            trigger: trigger,
            partial: props.partial
          });
        }
      }, null), createVNode(TDateTable, {
        "mode": props.mode,
        "data": props.tableData,
        "time": props.time,
        "value": props.value,
        "format": props.format,
        "multiple": props.multiple,
        "firstDayOfWeek": props.firstDayOfWeek,
        "onCellClick": function onCellClick(date, _ref3) {
          var _props$onCellClick;
          var e = _ref3.e;
          return (_props$onCellClick = props.onCellClick) === null || _props$onCellClick === void 0 ? void 0 : _props$onCellClick.call(props, date, {
            e: e,
            partial: props.partial
          });
        },
        "onCellMouseEnter": function onCellMouseEnter(date) {
          var _props$onCellMouseEnt;
          return (_props$onCellMouseEnt = props.onCellMouseEnter) === null || _props$onCellMouseEnt === void 0 ? void 0 : _props$onCellMouseEnt.call(props, date, {
            partial: props.partial
          });
        },
        "onCellMouseLeave": props.onCellMouseLeave,
        "cell": props.cell
      }, slots)]), props.enableTimePicker && createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-time")
      }, [createVNode("div", {
        "class": "".concat(COMPONENT_NAME.value, "-time-viewer")
      }, [props.time || defaultTimeValue]), createVNode(TimePickerPanel, _objectSpread({
        key: props.partial,
        isShowPanel: props.popupVisible,
        format: timeFormat,
        value: props.time || defaultTimeValue,
        onChange: props.onTimePickerChange,
        disableTime: disableTimeOptions
      }, props.timePickerProps), null)])]);
    };
  }
});

export { TPanelContent as default };
